"use strict";(self.webpackChunkclient=self.webpackChunkclient||[]).push([[37],{877:(Me,z,m)=>{m.d(z,{M:()=>g});var g=(()=>{return(D=g||(g={}))[D.local=0]="local",D[D.remote=1]="remote",g;var D})()},9037:(Me,z,m)=>{m.r(z),m.d(z,{ImportDataModule:()=>ka});var g=m(6019),D=m(4522),d=m(9133),q=m(9256),x=m(8638),P=m(3856),O=m(7695),C=m(8360),M=m(5658),L=m(8702),H=m(3928),it=m(3667),Ue=m(1144),e=m(3668),X=m(8472),c=m(4099),J=m(2411),nt=m(4102),r=m(8735),R=m(9204),f=m(8053),y=m(5571),at=m(5389),Ce=m(1366),Q=m(2047),K=m(877),Ie=m(1118),pe=m(9744),Je=m(8970),ot=m(1890),Qe=m(7548),Ee=m(7473),rt=m(2277),pt=m(6149),lt=m(839),ct=m(6531),ut=m(661),mt=m(5136),ht=m(6723),dt=m(5814),ft=m(1430),gt=m(1672),_t=m(4205),ee=m(8260);let St=(()=>{class i{constructor(t){this.http=t,this.datasetUrl=ee.N.serverUrl+"restful/dataset/",this.checkStatusUrl=this.datasetUrl+"checkIfExist",this.phenotypeDataUrl=this.datasetUrl+"requestPhenotypeData",this.datasetInfoUrl=this.datasetUrl+"requestDatasetInfo"}checkRemoteDatasetStatus(t,s){return this.http.get(`${this.checkStatusUrl}?database=${t}&datasetID=${s}`).pipe((0,f.U)(n=>({ifDataExist:n.ifDataExist,ifProcessedDataExist:n.ifProcessedDataExist})))}loadRemoteDatasetInfo(t,s){return this.http.get(`${this.datasetInfoUrl}?database=${t}&datasetID=${s}`).pipe((0,f.U)(n=>({expressLocalOrRemote:K.M.remote,database:t,datasetID:s,description:n.description,speciesID:n.speciesID,experimentType:n.experimentType,IDtype:n.IDtype,samples:n.samples})))}loadDatasetPhenotypeData(t,s){return this.http.get(`${this.phenotypeDataUrl}?database=${t}&datasetID=${s}`).pipe(n=>n)}}return i.\u0275fac=function(t){return new(t||i)(e.LFG(D.eN))},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac,providedIn:"root"}),i})();var Y=m(172),le=m(540);let vt=(()=>{class i{constructor(t){this.http=t,this.preprocessUrl=ee.N.serverUrl+"restful/preprocess/preprocessProteomics/"}handleResponse(t){return t.type===D.dt.Response?t.body:Y.E}preprocessProteomicsIntensities(t){const s=new D.aW("POST",this.preprocessUrl,t,{reportProgress:!1,responseType:"json"});return this.http.request(s).pipe((0,f.U)(n=>this.handleResponse(n)),(0,r.h)(n=>n!==Y.E),(0,f.U)(n=>n),(0,Q.K)(n=>(0,le._)(n)))}}return i.\u0275fac=function(t){return new(t||i)(e.LFG(D.eN))},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac,providedIn:"root"}),i})(),yt=(()=>{class i{constructor(t){this.http=t,this.pcaUrl=ee.N.serverUrl+"restful/exploreAnalysis/localDatasetPCA/"}handleResponse(t){return t.type===D.dt.Response?t.body:Y.E}obtainFirstTwoComponents(t){const s=new D.aW("POST",this.pcaUrl,t,{reportProgress:!1,responseType:"json"});return this.http.request(s).pipe((0,f.U)(n=>this.handleResponse(n)),(0,r.h)(n=>n!==Y.E),(0,f.U)(n=>n.pcaResult),(0,Q.K)(n=>(0,le._)(n)))}}return i.\u0275fac=function(t){return new(t||i)(e.LFG(D.eN))},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac,providedIn:"root"}),i})(),At=(()=>{class i{constructor(t){this.http=t,this.removeBatchUrl=ee.N.serverUrl+"restful/preprocess/removebatcheffect/",this.removeNormalizedBatchUrl=ee.N.removeBatchUrl+"removebatcheffectnormalvalues",this.removeCountsBatchUrl=ee.N.removeBatchUrl+"removebatcheffectrnaseq"}handleResponse(t){return t.type===D.dt.Response?t.body:Y.E}removeBatchEffect(t){const s=new D.aW("POST",this.removeBatchUrl,t,{reportProgress:!1,responseType:"json"});return this.http.request(s).pipe((0,f.U)(n=>this.handleResponse(n)),(0,r.h)(n=>n!==Y.E),(0,f.U)(n=>n.adjustedData),(0,Q.K)(n=>(0,le._)(n)))}removeBatchEffectOfNormalizedValues(t){let s;s="counts"===(null==t?void 0:t.valueType)?this.removeCountsBatchUrl:this.removeNormalizedBatchUrl;const n=new D.aW("POST",s,t,{reportProgress:!1,responseType:"json"});return this.http.request(n).pipe((0,f.U)(o=>this.handleResponse(o)),(0,r.h)(o=>o!==Y.E),(0,f.U)(o=>o),(0,Q.K)(o=>(0,le._)(o)))}}return i.\u0275fac=function(t){return new(t||i)(e.LFG(D.eN))},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac,providedIn:"root"}),i})();var Ge=m(4473);let W=(()=>{class i{constructor(t,s,n,o,l,u,h){this.store=t,this.geneAnnotationRequestService=s,this.manageRemoteDatasetService=n,this.preprocessProteomicsService=o,this.pcaService=l,this.removeBatchEffectsService=u,this.downloadSampleDataService=h,this.downloadGEOmatrixDataProgress$=new c.X(null),this.ifDownloadingGEOmatrixData$=new c.X(!1),this.downloadGEOgplDataProgress$=new c.X(null),this.ifDownloadingGEOgplData$=new c.X(!1),this.beforeRemovingBatchPCArequestSubject=new c.X(null),this.beforeRemovingBatchPCAresultSubject=new c.X(null),this.removeBatchRequestSubject=new c.X(null),this.removeBatchResultSubject=new c.X(null),this.removeBatchStatusSubject=new c.X(null),this.afterRemovingBatchPCArequestSubject=new c.X(null),this.afterRemovingBatchPCAresultSubject=new c.X(null),this.selectedSpeciesIDforMicroarraySubject=new c.X(null),this.preprocessProteomicsSubject=new c.X(null),this.preprocessProteomicsResultSubject=new c.X(null),this.preprocessProteomicsStatusSubject=new c.X(null),this.loadRemoteDatasetInfo$=new c.X(null),this.remoteDatasetStatus$=new c.X(null),this.requestGEOmatrixDataSubject=new c.X(null),this.requestDatabaseGeneID2info$=new c.X(null),this.currentDatasetPCAinfo$=new c.X(null),this.currentDatasetSamplesCorInfo$=new c.X(null),this.currentDatasetID$=new c.X(null),this.currentExpressDataPhenotypeDataSubject=new c.X(null),this.requestExpressDataPhenotypeData$=new c.X(null),this.currentGroups$=new c.X(null),this.updateDatasetGroups$=new c.X(null),this.matchedIDsPercentageRequestSubject=new c.X(null),this.allDatasetInfoForDisplay$=this.store.pipe((0,X.Ys)(Ee.yz),(0,nt.d)(2)),this.currentDatasetInfo$=this.currentDatasetID$.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.store.pipe((0,X.Ys)(Ee.iR,{datasetID:p})))),this.currentExpressDataPhenotypeData$=this.currentExpressDataPhenotypeDataSubject.asObservable(),this.theLatestAddedDatasetID$=this.store.pipe((0,X.Ys)(Qe.mV)),this.downloadSampleDataPercentage$=this.downloadSampleDataService.downloadPercentatage$,this.availableMicrorrayIDtypes$=this.selectedSpeciesIDforMicroarraySubject.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.geneAnnotationRequestService.requestArrayIDtypesInOneSpecies(p)),(0,f.U)(p=>p.sort((A,N)=>A.description>N.description?1:-1))),this.requestDatabaseGeneID2info$.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.store.pipe((0,X.Ys)(ot.tw,{ID:`${p.speciesID}-${p.idType}`}))),(0,r.h)(p=>!p),(0,y.M)(this.requestDatabaseGeneID2info$),(0,R.w)(([p,S])=>this.geneAnnotationRequestService.requestGenesAnnoFromServer(S.speciesID,S.idType))).subscribe(p=>{this.store.dispatch(Je.n({geneID2info:p}))}),this.requestExpressDataPhenotypeData$.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>p.expressLocalOrRemote===K.M.remote?this.manageRemoteDatasetService.loadDatasetPhenotypeData(p.database,p.datasetID):this.store.pipe((0,X.Ys)(Qe.nL,{datasetID:p.datasetID})))).subscribe(p=>{this.currentExpressDataPhenotypeDataSubject.next(p)}),this.updateDatasetGroups$.pipe((0,r.h)(p=>null!=p),(0,y.M)(this.currentDatasetID$.pipe((0,r.h)(p=>null!=p)))).subscribe(([p,S])=>{this.store.dispatch(pe.Tl({datasetID:S,groups:p}))}),this.preprocessProteomicsSubject.pipe((0,r.h)(p=>null!=p),(0,at.x)(),(0,Ce.b)(()=>{this.preprocessProteomicsStatusSubject.next({status:"loading",statusValue:"loading"})}),(0,R.w)(p=>this.preprocessProteomicsService.preprocessProteomicsIntensities({proteinIDs:p.proteinIDs,samples2intensities:p.samples2intensities,groups:p.groups,preprocessSettings:p.preprocessSettings}).pipe((0,f.U)(A=>({preprocessID:p.preprocessID,samples:p.samples,groups:p.groups,geneIDs:A.proteinIDs,normalizedValues:A.normalizedValues})))),(0,Q.K)(p=>(0,J.of)({status:"error",statusValue:`Error when preprocessing proteomics data. error:${p.status}, errorText:${p.statusText}`}))).subscribe(p=>{p.hasOwnProperty("status")?this.preprocessProteomicsStatusSubject.next(p):(this.preprocessProteomicsResultSubject.next(p),this.preprocessProteomicsStatusSubject.next({status:"finished",statusValue:"finished"}))}),this.preprocessProteomicsResult$=this.preprocessProteomicsResultSubject.asObservable(),this.preprocessProteomicsStatus$=this.preprocessProteomicsStatusSubject.asObservable(),this.beforeRemovingBatchPCArequestSubject.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.pcaService.obtainFirstTwoComponents({values:p.values,samples:p.samples}).pipe((0,f.U)(A=>({pcaRequestID:p.pcaRequestID,samplesPCAinfos:A})))),(0,Q.K)(p=>(0,J.of)({status:"error",statusValue:`Error when PCA. error:${p.status}, errorText:${p.statusText}`}))).subscribe(p=>{p.hasOwnProperty("status")||this.beforeRemovingBatchPCAresultSubject.next(p)}),this.beforeRemovingBatchPCAresult$=this.beforeRemovingBatchPCAresultSubject.asObservable(),this.removeBatchRequestSubject.pipe((0,r.h)(p=>null!=p),(0,Ce.b)(()=>{this.removeBatchStatusSubject.next({status:"loading",statusValue:"loading"})}),(0,R.w)(p=>this.removeBatchEffectsService.removeBatchEffectOfNormalizedValues(p).pipe((0,f.U)(S=>({removeBatchID:p.removeBatchID,values:S})))),(0,Q.K)(p=>(0,J.of)({status:"error",statusValue:`Error when removing batch effects. error:${p.status}, errorText:${p.statusText}`}))).subscribe(p=>{p.hasOwnProperty("status")?this.removeBatchStatusSubject.next(p):(this.removeBatchResultSubject.next(p),this.removeBatchStatusSubject.next({status:"finished",statusValue:"finished"}))}),this.removeBatchStatus$=this.removeBatchStatusSubject.asObservable(),this.removeBatchResult$=this.removeBatchResultSubject.asObservable(),this.afterRemovingBatchPCArequestSubject.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.pcaService.obtainFirstTwoComponents({values:p.values,samples:p.samples}).pipe((0,f.U)(A=>({pcaRequestID:p.pcaRequestID,samplesPCAinfos:A})))),(0,Q.K)(p=>(0,J.of)({status:"error",statusValue:`Error when PCA. error:${p.status}, errorText:${p.statusText}`}))).subscribe(p=>{p.hasOwnProperty("status")||this.afterRemovingBatchPCAresultSubject.next(p)}),this.afterRemovingBatchPCAresult$=this.afterRemovingBatchPCAresultSubject.asObservable(),this.matchedIDsPercentageResult$=this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(p=>null!=p),(0,R.w)(p=>this.geneAnnotationRequestService.obtainMatchedIDspercentage({speciesID:p.speciesID,IDtype:p.IDtype,IDs:p.IDs}).pipe((0,f.U)(S=>({requestID:p.requestID,matchedPercentage:S})),(0,Q.K)(S=>(0,J.of)({requestID:p.requestID,matchedPercentage:0})))))}requestMicroarrayIDtypesInTheSpecies(t){this.selectedSpeciesIDforMicroarraySubject.next(t)}requestMatchedIDsPercentage(t){this.matchedIDsPercentageRequestSubject.next(t)}addLocalExpressData(t){this.store.dispatch(Ie.LP({expressDataInfo:t})),this.currentExpressDataPhenotypeDataSubject.next(t.phenotypeData),this.currentDatasetID$.next(t.datasetID)}preprocessProteomicsData(t){this.preprocessProteomicsSubject.next(t)}requestBeforeRemovingBatchPCAInfo(t){this.beforeRemovingBatchPCArequestSubject.next(t)}processRemoveBatchEffectRequest(t){this.removeBatchRequestSubject.next(t)}requestAfterRemovingBatchPCAinfo(t){this.afterRemovingBatchPCArequestSubject.next(t)}addDatasetInfo(t){this.store.dispatch(pe.LH({dataset:t}))}deleteThisDataset(t){this.store.dispatch(Ie.Ry({datasetID:t})),this.store.dispatch(pe.VO({datasetID:t}))}requestDatabaseGeneID2info(t){this.requestDatabaseGeneID2info$.next(t)}updateDatasetGroups(t){this.updateDatasetGroups$.next(t)}requestDatasetInfoForReview(t){this.currentDatasetID$.next(t)}requestRemoteDatasetInfo(t){this.loadRemoteDatasetInfo$.next(t)}downloadSampleFile(t){this.downloadSampleDataService.downloadSampleData(t)}prepareLocalProject(t){this.store.dispatch(Ie.gw({newState:t.localExpressDataState})),this.store.dispatch(pe.CF({newState:t.datasetListState})),this.store.dispatch(dt.yx({newState:t.keggPathwayState})),this.store.dispatch(Je.A({newState:t.geneAnnotationsState})),this.store.dispatch(ct.dy({newState:t.geneIDmappingState})),this.store.dispatch(ut.c({newState:t.geneListState})),this.store.dispatch(mt.V({newState:t.geneListEnrichedAnnotationsState})),this.store.dispatch(pt.Uw({newState:t.diffAnalysisResultState})),this.store.dispatch(lt.lO({newState:t.diffExpState})),this.store.dispatch(rt.N8({newState:t.clusterExpState})),this.store.dispatch(ht.Y6({newState:t.keggExpState})),this.store.dispatch(ft.iS({newState:t.PPIexpState})),this.store.dispatch(gt.R({newState:t.survivalExpState}))}}return i.\u0275fac=function(t){return new(t||i)(e.LFG(X.yh),e.LFG(_t.b),e.LFG(St),e.LFG(vt),e.LFG(yt),e.LFG(At),e.LFG(Ge.h))},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac}),i})();var $=m(5043);function bt(i,a){1&i&&(e.TgZ(0,"p"),e._uU(1,"No dataset. Please import one first"),e.qZA())}function xt(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"ul",3),e.NdJ("click",function(){const o=e.CHM(t).$implicit;return e.oxw().outputSelectedDatasetIDForView(o.datasetID)}),e.TgZ(1,"p",4),e._uU(2),e.qZA(),e.TgZ(3,"div",5),e.TgZ(4,"fa-icon",6),e.NdJ("click",function(){const o=e.CHM(t).$implicit;return e.oxw().outputSelectDatasetIDforDelete(o.datasetID)}),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=a.$implicit,s=e.oxw();e.xp6(2),e.Oqu(t.datasetID),e.xp6(2),e.Q6J("icon",s.faMinus)("size","1x")}}let Dt=(()=>{class i{constructor(){this.selectedDatasetIDforView$=new e.vpe,this.selectedDatasetIDforDelete$=new e.vpe,this.faMinus=$.Kl4}ngOnInit(){}outputSelectedDatasetIDForView(t){this.selectedDatasetIDforView$.next(t)}outputSelectDatasetIDforDelete(t){this.selectedDatasetIDforDelete$.next(t)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-present-datasets"]],inputs:{datasets:"datasets"},outputs:{selectedDatasetIDforView$:"selectedDatasetIDforView$",selectedDatasetIDforDelete$:"selectedDatasetIDforDelete$"},decls:4,vars:2,consts:[[1,"mt-3"],[4,"ngIf"],["class","d-flex",3,"click",4,"ngFor","ngForOf"],[1,"d-flex",3,"click"],["href","javascript:void(0)"],[1,"form-inline","ml-1"],["ngbTooltip","Delete this dataset",1,"hoverBigger",3,"icon","size","click"]],template:function(t,s){1&t&&(e.TgZ(0,"h5",0),e._uU(1,"Available Datasets"),e.qZA(),e.YNc(2,bt,2,0,"p",1),e.YNc(3,xt,5,3,"ul",2)),2&t&&(e.xp6(2),e.Q6J("ngIf",!s.datasets||0===(null==s.datasets?null:s.datasets.length)),e.xp6(1),e.Q6J("ngForOf",s.datasets))},directives:[g.O5,g.sg,P.BN,x._L],styles:[""],changeDetection:0}),i})();var B=m(3405),b=m(2668),E=m(6087),T=m(1125),_=m(8805),Tt=m(4168),Ct=m(8178);const je={leading:!0,trailing:!1};function ce(i,a=Ct.P,t=je){return s=>s.lift(new Pt(i,a,t.leading,t.trailing))}class Pt{constructor(a,t,s,n){this.duration=a,this.scheduler=t,this.leading=s,this.trailing=n}call(a,t){return t.subscribe(new Nt(a,this.duration,this.scheduler,this.leading,this.trailing))}}class Nt extends Tt.L{constructor(a,t,s,n,o){super(a),this.duration=t,this.scheduler=s,this.leading=n,this.trailing=o,this._hasTrailingValue=!1,this._trailingValue=null}_next(a){this.throttled?this.trailing&&(this._trailingValue=a,this._hasTrailingValue=!0):(this.add(this.throttled=this.scheduler.schedule(wt,this.duration,{subscriber:this})),this.leading?this.destination.next(a):this.trailing&&(this._trailingValue=a,this._hasTrailingValue=!0))}_complete(){this._hasTrailingValue?(this.destination.next(this._trailingValue),this.destination.complete()):this.destination.complete()}clearThrottle(){const a=this.throttled;a&&(this.trailing&&this._hasTrailingValue&&(this.destination.next(this._trailingValue),this._trailingValue=null,this._hasTrailingValue=!1),a.unsubscribe(),this.remove(a),this.throttled=null)}}function wt(i){const{subscriber:a}=i;a.clearThrottle()}var ue,qt=new Uint8Array(16);function Rt(){if(!ue&&!(ue="undefined"!=typeof crypto&&crypto.getRandomValues&&crypto.getRandomValues.bind(crypto)||"undefined"!=typeof msCrypto&&"function"==typeof msCrypto.getRandomValues&&msCrypto.getRandomValues.bind(msCrypto)))throw new Error("crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported");return ue(qt)}const $t=/^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000)$/i,Ft=function(i){return"string"==typeof i&&$t.test(i)};for(var I=[],Ze=0;Ze<256;++Ze)I.push((Ze+256).toString(16).substr(1));const G=function(i,a,t){var s=(i=i||{}).random||(i.rng||Rt)();if(s[6]=15&s[6]|64,s[8]=63&s[8]|128,a){t=t||0;for(var n=0;n<16;++n)a[t+n]=s[n];return a}return function(i){var a=arguments.length>1&&void 0!==arguments[1]?arguments[1]:0,t=(I[i[a+0]]+I[i[a+1]]+I[i[a+2]]+I[i[a+3]]+"-"+I[i[a+4]]+I[i[a+5]]+"-"+I[i[a+6]]+I[i[a+7]]+"-"+I[i[a+8]]+I[i[a+9]]+"-"+I[i[a+10]]+I[i[a+11]]+I[i[a+12]]+I[i[a+13]]+I[i[a+14]]+I[i[a+15]]).toLowerCase();if(!Ft(t))throw TypeError("Stringified UUID is invalid");return t}(s)};var v=m(9232);const k=function(i,a){const t=[...a.samples],s=(0,v.Wv)(i,t);let n={};for(const[o,l]of Object.entries(a)){const u=s.map(h=>l[h]);n[o]=u}return n},Le=i=>{let t=!1;return i.value.forEach(s=>{(0===s.samples.length||""===s.name)&&(t=!0)}),t?{hasEmptySamples:!0}:null},ke=i=>{const s=i.value.map(o=>o.samples).reduce((o,l)=>o.concat(l));return(0,v.ro)(s)?{groupsOverlapSamples:!0}:null},Ve=i=>{const t=i.value.map(n=>n.name);return(0,v.ro)(t)?{groupNamesDuplicate:!0}:null},me=function(i,a){let t=new Map;return i.forEach(n=>{n.samples.forEach(o=>{t.set(o,n.name)})}),a.map(n=>t.get(n))};var V=m(9220),Et=m(2886);function Gt(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"ol"),e.TgZ(1,"a",2),e.NdJ("click",function(){const o=e.CHM(t).$implicit;return e.oxw().downloadSampleData(o.name)}),e._uU(2),e.qZA(),e.qZA()}if(2&i){const t=a.$implicit;e.xp6(2),e.Oqu(t.description)}}function jt(i,a){if(1&i&&(e.TgZ(0,"div"),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. App is downloading sample data"),e.qZA(),e.TgZ(3,"progressbar",3),e._uU(4),e.qZA(),e.qZA()),2&i){const t=e.oxw();e.xp6(3),e.Q6J("value",t.downloadSampleDataPercentage)("striped",!0)("animate",!0),e.xp6(1),e.hij(" ",t.downloadSampleDataPercentage,"% ")}}let he=(()=>{class i{constructor(t){this.downloadService=t,this.fileNameToBeDownloaded$=new e.vpe}ngOnInit(){}downloadSampleData(t){this.fileNameToBeDownloaded$.next(t)}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(Ge.h))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-download-sample-data"]],inputs:{fileInfos:"fileInfos",downloadSampleDataPercentage:"downloadSampleDataPercentage"},outputs:{fileNameToBeDownloaded$:"fileNameToBeDownloaded$"},decls:4,vars:2,consts:[[4,"ngFor","ngForOf"],[4,"ngIf"],["href","javascript:void(0)",3,"click"],["type","success",3,"value","striped","animate"]],template:function(t,s){1&t&&(e.TgZ(0,"p"),e._uU(1,"If you are the first time using ExpressVis, you can download sample data and follow the steps in readme.txt to import it, or click try sample button in each app (DiffExp, ClusterExp, KeggExp, PPIExp, SurvivalExp) to experience the app without importing data"),e.qZA(),e.YNc(2,Gt,3,1,"ol",0),e.YNc(3,jt,5,4,"div",1)),2&t&&(e.xp6(2),e.Q6J("ngForOf",s.fileInfos),e.xp6(1),e.Q6J("ngIf",s.downloadSampleDataPercentage<100&&s.downloadSampleDataPercentage>0))},directives:[g.sg,g.O5,Et.O1],styles:[""],changeDetection:0}),i})(),de=(()=>{class i{constructor(t){this.ele=t}set appAddRequiredOptionalSign(t){"required"===t?this.ele.nativeElement.insertAdjacentHTML("beforeend","<span style = 'color: red'>*</span>"):"optional"===t&&(this.ele.nativeElement.textContent=this.ele.nativeElement.textContent+" (optional)")}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.SBq))},i.\u0275dir=e.lG2({type:i,selectors:[["","appAddRequiredOptionalSign",""]],inputs:{appAddRequiredOptionalSign:"appAddRequiredOptionalSign"}}),i})();var U=m(5863),fe=m(4863),ie=m(94),w=m(3314),ge=m(3139);let _e=(()=>{class i extends ge.S{constructor(){super(),this.IDcolumnSubject=new c.X(null),this.samplesSubject=new c.X(null),this.IDcolumn$=this.IDcolumnSubject.asObservable(),this.samples$=this.samplesSubject.asObservable()}obtainValuesInfo(t,s){return((i,a,t)=>{let s=[],n=[];const o=i[0];let l=[];o.forEach((h,p)=>{t.indexOf(h)>=0&&l.push(p)});const u=o.indexOf(a);return i.slice(1).forEach(h=>{let p=[];l.forEach(S=>{p.push(parseFloat(h[S]))}),n.push(p),s.push(h[u].split(".")[0])}),{IDs:s,samples:t,values:n}})(this.textMatrix,t,s)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac}),i})();const kt=["displayFigureComponent"];function Vt(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",2),e.TgZ(1,"button",3),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayFigureModalRef?null:n.displayFigureModalRef.hide()}),e.TgZ(2,"span",4),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",5),e._UZ(5,"img",6),e.qZA(),e.TgZ(6,"div",7),e.TgZ(7,"button",8),e.NdJ("click",function(){return e.CHM(t),e.oxw().displayFigureModalRef.hide()}),e._uU(8,"close"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("src",t.figureSrc,e.LSH)}}let ne=(()=>{class i{constructor(t,s){this.modalService=t,this.cdr=s}ngOnInit(){}displayFigure(t){this.displayFigureModalRef=this.modalService.show(this.displayFigureComponent,Object.assign({},{class:"modal-lg"})),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-display-a-figure-in-modal"]],viewQuery:function(t,s){if(1&t&&e.Gf(kt,5,e.Rgc),2&t){let n;e.iGM(n=e.CRH())&&(s.displayFigureComponent=n.first)}},inputs:{figureSrc:"figureSrc",figureLabel:"figureLabel"},decls:4,vars:1,consts:[["href","javascript:void(0)",1,"inline",3,"click"],["displayFigureComponent",""],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"img-fluid","rounded","mx-auto","d-block",3,"src"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"]],template:function(t,s){1&t&&(e.TgZ(0,"a",0),e.NdJ("click",function(o){return s.displayFigure(o)}),e._uU(1),e.qZA(),e.YNc(2,Vt,9,1,"ng-template",null,1,e.W1O)),2&t&&(e.xp6(1),e.hij(" ",s.figureLabel,"\n"))},styles:[""],changeDetection:0}),i})();var te=m(3013);const Ht=["presentFirstLinesComponent"];function Yt(i,a){if(1&i&&(e.TgZ(0,"div"),e.TgZ(1,"p",11),e._uU(2,"The app is parsing the file. Please wait for a moment."),e.qZA(),e._UZ(3,"fa-icon",12),e.qZA()),2&i){const t=e.oxw();e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function Wt(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",13),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentFirstLinesOfTheFile(n)}),e._uU(1," View first lines of the file\n"),e.qZA()}}function zt(i,a){if(1&i&&(e.TgZ(0,"th"),e.TgZ(1,"p",23),e._uU(2),e.qZA(),e.qZA()),2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t," ")}}function Xt(i,a){if(1&i&&(e.TgZ(0,"td"),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.hij(" ",t," ")}}function Kt(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,Xt,2,1,"td",20),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.Q6J("ngForOf",t)}}function es(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",14),e.TgZ(1,"h3"),e._uU(2,"The first five lines of the file"),e.qZA(),e.TgZ(3,"button",15),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentFirstLinesModalRef?null:n.presentFirstLinesModalRef.hide()}),e.TgZ(4,"span",16),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",17),e.TgZ(7,"div",18),e.TgZ(8,"table",19),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.YNc(11,zt,3,1,"th",20),e.ALo(12,"async"),e.qZA(),e.qZA(),e.TgZ(13,"tbody"),e.YNc(14,Kt,2,1,"tr",20),e.ALo(15,"async"),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(16,"div",21),e.TgZ(17,"button",22),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentFirstLinesModalRef.hide()}),e._uU(18," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",e.lcZ(12,2,t.fileHeaders$)),e.xp6(3),e.Q6J("ngForOf",e.lcZ(15,4,t.firstLines$))}}let ts=(()=>{class i extends ie.e{constructor(t,s,n){super(),this.modalService=t,this.importExpressionService=s,this.cdr=n,this.formWidth="15rem",this.expressionInfoInFile$=new e.vpe(null),this.faSpinner=$.LM3,this.fileHeaders$=new w.y(null),this.firstLines$=new w.y(null),this.fileParseStatus$=new w.y(null),this.dropdownSettings={singleSelection:!1,selectAllText:"Select All as Sample Value Columns",unSelectAllText:"UnSelect All",itemsShowLimit:1,enableCheckAll:!0,allowSearchFilter:!0},this.candidateSamples$=new w.y(null),this.selectedSamples$=new c.X(null),this.selectedSamples=[],this.IDcolumn$=new c.X(null)}ngOnInit(){this.firstLines$=this.importExpressionService.firstLines$.pipe((0,_.R)(this.destroyNotifier$)),this.fileHeaders$=this.importExpressionService.fileHeaders$.pipe((0,_.R)(this.destroyNotifier$)),this.fileParseStatus$=this.importExpressionService.fileParseStatus$.pipe((0,_.R)(this.destroyNotifier$))}ngAfterViewInit(){this.candidateSamples$=this.fileHeaders$.pipe((0,r.h)(t=>null!=t)),this.candidateSamples$.subscribe(t=>{(null==t?void 0:t.length)>0&&(this.selectedSamples=t.slice(1),this.selectedSamples$.next(this.selectedSamples))}),(0,b.aj)([this.IDcolumn$.pipe((0,r.h)(t=>null!=t)),this.selectedSamples$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>this.importExpressionService.obtainValuesInfo(t,s))).subscribe(t=>{this.expressionInfoInFile$.next(t)})}processSelectedFile(t){this.importExpressionService.initLoadedFile(t)}presentFirstLinesOfTheFile(t){this.presentFirstLinesModalRef=this.modalService.show(this.presentFirstLinesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}processSelectedIDColumn(t){this.IDcolumn$.next(t)}updateSelectedSamples(t){this.selectedSamples$.next(this.selectedSamples)}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(q.tT),e.Y36(_e),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-counts-or-tpms"]],viewQuery:function(t,s){if(1&t&&e.Gf(Ht,5,e.Rgc),2&t){let n;e.iGM(n=e.CRH())&&(s.presentFirstLinesComponent=n.first)}},inputs:{formWidth:"formWidth",fileType:"fileType",exampleFileFigureSrc:"exampleFileFigureSrc"},outputs:{expressionInfoInFile$:"expressionInfoInFile$"},features:[e._Bn([_e]),e.qOj],decls:21,vars:25,consts:[["role","form",1,"mt-2"],["data-testid","fileType"],[1,"d-flex"],["type","file",1,"form-control","form-control-sm",3,"change"],[1,"ml-2",3,"figureSrc","figureLabel"],[4,"ngIf"],["href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],[1,"mt-2"],[3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","selectedItem$"],["ngDefaultControl","",3,"placeholder","settings","data","ngModel","ngModelChange","onSelect","onSelectAll","onDeSelect"],["presentFirstLinesComponent",""],[1,"text-info"],[1,"hoverBigger",3,"icon","size","spin"],["href","javascript:void(0)",1,"inline",3,"click"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"table-responsive","mr-2","ml-2","mt-2"],[1,"table","table-bordered",2,"max-height","50vh"],[4,"ngFor","ngForOf"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"],[1,"font-weight-bold"]],template:function(t,s){if(1&t&&(e.TgZ(0,"form",0),e.TgZ(1,"div"),e.TgZ(2,"h5",1),e._uU(3),e.qZA(),e.TgZ(4,"div",2),e.TgZ(5,"input",3),e.NdJ("change",function(o){return s.processSelectedFile(o)}),e.qZA(),e._UZ(6,"app-display-a-figure-in-modal",4),e.qZA(),e.qZA(),e.qZA(),e.YNc(7,Yt,4,3,"div",5),e.ALo(8,"async"),e.YNc(9,Wt,2,0,"a",6),e.ALo(10,"async"),e.TgZ(11,"div",7),e.TgZ(12,"app-select-item",8),e.NdJ("selectedItem$",function(o){return s.processSelectedIDColumn(o)}),e.ALo(13,"async"),e.qZA(),e.qZA(),e.TgZ(14,"h5",7),e._uU(15,"Select value Columns:"),e.qZA(),e.TgZ(16,"div"),e.TgZ(17,"ng-multiselect-dropdown",9),e.NdJ("ngModelChange",function(o){return s.selectedSamples=o})("onSelect",function(o){return s.updateSelectedSamples(o)})("onSelectAll",function(o){return s.updateSelectedSamples(o)})("onDeSelect",function(o){return s.updateSelectedSamples(o)}),e.ALo(18,"async"),e.qZA(),e.qZA(),e.YNc(19,es,19,6,"ng-template",null,10,e.W1O)),2&t){let n;e.xp6(3),e.hij("",s.fileType,":"),e.xp6(2),e.Udp("width",s.formWidth),e.xp6(1),e.Q6J("figureSrc",s.exampleFileFigureSrc)("figureLabel","View Example"),e.xp6(1),e.Q6J("ngIf","loading"===(null==(n=e.lcZ(8,17,s.fileParseStatus$))?null:n.status)),e.xp6(2),e.Q6J("ngIf",e.lcZ(10,19,s.firstLines$)),e.xp6(3),e.Q6J("inputItems",e.lcZ(13,21,s.fileHeaders$))("itemLabel","Select Gene ID Column")("ifOutputFirstValue",!0)("formWidth",s.formWidth),e.xp6(4),e.Udp("width",s.formWidth),e.xp6(1),e.Q6J("placeholder","Select sample columns")("settings",s.dropdownSettings)("data",e.lcZ(18,23,s.candidateSamples$))("ngModel",s.selectedSamples)}},directives:[d._Y,d.JL,d.F,ne,g.O5,te.y,C.OP,d.Fj,d.JJ,d.On,P.BN,g.sg],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();function ss(i,a){if(1&i&&(e.TgZ(0,"option",9),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.Q6J("value",t.name),e.xp6(1),e.Oqu(t.description)}}function is(i,a){1&i&&(e.TgZ(0,"div",10),e.TgZ(1,"p",11),e._uU(2," Samples of counts data and TPM/RPKM/FPKM data are not the same. Please change files or change values columns "),e.qZA(),e.qZA())}function ns(i,a){if(1&i&&(e.TgZ(0,"p",12),e._uU(1),e.qZA()),2&i){const t=e.oxw();e.xp6(1),e.AsE(" Only ",100*(null==t.matchedIDsPercentageResult?null:t.matchedIDsPercentageResult.matchedPercentage),"% genes/proteins are mapped to ",t.selectedIDtype,". Please re-select Species, ID column or ID type.\n")}}let as=(()=>{class i extends ie.e{constructor(){super(),this.formWidth="15rem",this.matchedIDsPercentageRequest$=new e.vpe,this.countsTpmsInfo$=new e.vpe,this.countsFileExampleSrc="/static/images/CountSampleFigure.png",this.tpmsFileExampleSrc="/static/images/TPM_FPKM_RPKM_SampleFigure.png",this.countsInfoSubject=new c.X(null),this.tpmsInfoSubject=new c.X(null),this.geneIDtypeSubject=new c.X(null),this.selectedIDtype="",this.matchedIDsCutoff=.5,this.matchedIDsPercentageRequestSubject=new c.X(null),this.matchedIDsPercentageResultSubject=new c.X(null)}ngOnInit(){this.ifShowSamplesNote$=(0,b.aj)([this.countsInfoSubject.pipe((0,r.h)(t=>null!=t)),this.tpmsInfoSubject.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>!(0,v.z7)([...t.samples],[...s.samples])),(0,T.O)(!1)),(0,b.aj)([this.countsInfoSubject.pipe((0,r.h)(t=>null!=t)),this.tpmsInfoSubject.pipe((0,r.h)(t=>null!=t)),this.geneIDtypeSubject.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s,n])=>{if(t&&s&&n){const o=t.IDs.concat(s.IDs),l=Array.from(new Set(o));return{requestID:G(),IDtype:n,IDs:l}}return null}),(0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.matchedIDsPercentageRequest$.next(t),this.matchedIDsPercentageRequestSubject.next(t)}),this.matchedIDsPercentageResultSubject.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>t.matchedPercentage>=this.matchedIDsCutoff),(0,y.M)(this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>{if(t.requestID===s.requestID)return!0}),(0,r.h)(t=>t),(0,y.M)(this.countsInfoSubject.pipe((0,r.h)(t=>null!=t)),this.tpmsInfoSubject.pipe((0,r.h)(t=>null!=t)),this.geneIDtypeSubject.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s,n,o])=>{if((0,v.z7)([...s.samples],[...n.samples])){const u=function(i,a,t){if((0,v.a)(a,t))return i;const s=(0,v.Wv)(t,a),n=i[0].length;let o=[];for(let l of s)o.push(l<0?Array(n).fill(null):i[l]);return o}([...n.values],[...n.IDs],[...s.IDs]),h=function(i,a,t){if((0,v.a)(a,t))return i;const s=(0,v.Wv)(t,a);let n=[];for(let o of i){const l=s.map(u=>o[u]);n.push(l)}return n}([...u],[...n.samples],[...s.samples]);return{IDtype:o,geneIDs:s.IDs,samples:s.samples,expressionValues:h,counts:s.values}}return null})).subscribe(t=>{this.countsTpmsInfo$.next(t)})}ngOnChanges(t){t.matchedIDsPercentageResult&&!t.matchedIDsPercentageResult.isFirstChange()&&this.matchedIDsPercentageResultSubject.next(this.matchedIDsPercentageResult)}processCountsInfo(t){this.countsInfoSubject.next(t)}processTpmsInfo(t){this.tpmsInfoSubject.next(t)}processSelectedGeneIDtype(t){this.geneIDtypeSubject.next(t.target.value),this.selectedIDtype=t.target.value}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-rna-seq-counts-tpms"]],inputs:{formWidth:"formWidth",availableIDtypes:"availableIDtypes",matchedIDsPercentageResult:"matchedIDsPercentageResult"},outputs:{matchedIDsPercentageRequest$:"matchedIDsPercentageRequest$",countsTpmsInfo$:"countsTpmsInfo$"},features:[e.qOj,e.TTD],decls:19,vars:11,consts:[[1,"ml-2"],["data-testid","importCounts",3,"fileType","exampleFileFigureSrc","expressionInfoInFile$"],[1,"mt-4","border-top"],["data-testid","importTpms",3,"fileType","exampleFileFigureSrc","expressionInfoInFile$"],["for","IDtype"],["data-testid","countsTpmsIDtype","name","IDtype",1,"form-control","form-control-sm",3,"change"],[3,"value",4,"ngFor","ngForOf"],["class","mt-3",4,"ngIf"],["data-testid","IDnotMatchNote","class","mt-3 text-danger",4,"ngIf"],[3,"value"],[1,"mt-3"],["data-testid","samplesUnequalNote",1,"text-danger"],["data-testid","IDnotMatchNote",1,"mt-3","text-danger"]],template:function(t,s){1&t&&(e.TgZ(0,"h4"),e._uU(1,"Counts Data"),e.qZA(),e.TgZ(2,"div",0),e.TgZ(3,"app-import-counts-or-tpms",1),e.NdJ("expressionInfoInFile$",function(o){return s.processCountsInfo(o)}),e.qZA(),e.qZA(),e.TgZ(4,"div",2),e.TgZ(5,"h4"),e._uU(6,"TPM/RPKM/FPKM Data"),e.qZA(),e.TgZ(7,"div",0),e.TgZ(8,"app-import-counts-or-tpms",3),e.NdJ("expressionInfoInFile$",function(o){return s.processTpmsInfo(o)}),e.qZA(),e.qZA(),e.qZA(),e.TgZ(9,"div",2),e.TgZ(10,"h4",4),e._uU(11,"Gene ID type:"),e.qZA(),e.TgZ(12,"select",5),e.NdJ("change",function(o){return s.processSelectedGeneIDtype(o)}),e.TgZ(13,"option"),e._uU(14,"---"),e.qZA(),e.YNc(15,ss,2,2,"option",6),e.qZA(),e.qZA(),e.YNc(16,is,3,0,"div",7),e.ALo(17,"async"),e.YNc(18,ns,2,2,"p",8)),2&t&&(e.xp6(3),e.Q6J("fileType","Counts File")("exampleFileFigureSrc",s.countsFileExampleSrc),e.xp6(5),e.Q6J("fileType","TPM/RPKM/FPKM File")("exampleFileFigureSrc",s.tpmsFileExampleSrc),e.xp6(4),e.Udp("width",s.formWidth),e.xp6(3),e.Q6J("ngForOf",s.availableIDtypes),e.xp6(1),e.Q6J("ngIf",e.lcZ(17,9,s.ifShowSamplesNote$)),e.xp6(2),e.Q6J("ngIf",null!=s.matchedIDsPercentageResult&&s.matchedIDsPercentageResult.matchedPercentage<.5))},directives:[ts,d.YN,d.Kr,g.sg,g.O5],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();var j=m(8526);let He=(()=>{class i extends ge.S{constructor(){super(),this.phenotypeDataSubject=new c.X(null),this.phenotypeData$=this.phenotypeDataSubject.asObservable()}obtainPhenotypeObject(t){return((i,a)=>{let t={},s={};return[...i][0].forEach((n,o)=>{t[o]=[]}),[...i].slice(1).forEach(n=>{n.forEach((o,l)=>{t[l].push(o)})}),[...i][0].forEach((n,o)=>{s[n]=[...t[o]]}),s.samples=[...s[a]],s})(this.textMatrix,t)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac}),i})();var Pe=m(6872);const rs=["fileForm"],ps=["presentFirstLinesComponent"];function ls(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",11),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentFirstLinesOfTheFile(n)}),e._uU(1," View the first lines of the file\n"),e.qZA()}}function cs(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",6),e.TgZ(1,"p",12),e._uU(2,"This step is optional. Click skip button to proceed to the next step"),e.qZA(),e.TgZ(3,"button",13),e.NdJ("click",function(n){return e.CHM(t),e.oxw().skipThisStep(n)}),e._uU(4,"Skip"),e.qZA(),e.qZA()}}function us(i,a){if(1&i&&(e.TgZ(0,"th"),e.TgZ(1,"p",24),e._uU(2),e.qZA(),e.qZA()),2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t," ")}}function ms(i,a){if(1&i&&(e.TgZ(0,"td",26),e._uU(1),e.ALo(2,"trimString"),e.qZA()),2&i){const t=a.$implicit;e.s9C("title",t),e.xp6(1),e.hij(" ",e.xi3(2,2,t,20)," ")}}function hs(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,ms,3,5,"td",25),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.Q6J("ngForOf",t)}}function ds(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",14),e.TgZ(1,"h3"),e._uU(2,"The first five lines of the file"),e.qZA(),e.TgZ(3,"button",15),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentFirstLinesModalRef?null:n.presentFirstLinesModalRef.hide()}),e.TgZ(4,"span",16),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",17),e.TgZ(7,"div",18),e.TgZ(8,"table",19),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.YNc(11,us,3,1,"th",20),e.qZA(),e.qZA(),e.TgZ(12,"tbody",21),e.YNc(13,hs,2,1,"tr",20),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(14,"div",22),e.TgZ(15,"button",23),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentFirstLinesModalRef.hide()}),e._uU(16," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",t.phenotypeShowInfo.fileHeaders),e.xp6(2),e.Q6J("ngForOf",t.phenotypeShowInfo.firstLines)}}let Se=(()=>{class i{constructor(t,s,n){this.phenoDataService=t,this.modalService=s,this.cdr=n,this.resetToDefault=!1,this.formWidth="15rem",this.phenotypeData$=new e.vpe,this.skipSignal$=new e.vpe,this.phenotypeFileExampleSrc="/static/images/PhenotypeSampleFigure.png",this.fileHeaders$=new w.y(null),this.fileHeadersFromReset$=new c.X(null),this.firstLines$=new w.y(null),this.firstLinesFromReset$=new c.X(null),this.sampleColumn$=new c.X(null),this.destroyNotifier$=new B.xQ}ngOnInit(){this.firstLines$=(0,j.T)(this.firstLinesFromReset$,this.phenoDataService.firstLines$).pipe((0,_.R)(this.destroyNotifier$)),this.fileHeaders$=(0,j.T)(this.fileHeadersFromReset$,this.phenoDataService.fileHeaders$).pipe((0,_.R)(this.destroyNotifier$)),(0,b.aj)([this.firstLines$,this.fileHeaders$]).subscribe(([t,s])=>{t&&s&&(this.phenotypeShowInfo=Object.assign(Object.assign({},this.phenotypeShowInfo),{firstLines:t,fileHeaders:s}),this.cdr.detectChanges())}),this.sampleColumn$.pipe((0,r.h)(t=>null!=t),(0,f.U)(t=>this.phenoDataService.obtainPhenotypeObject(t))).subscribe(t=>{this.phenotypeData$.next(t),this.phenotypeData=t,this.cdr.detectChanges()})}ngOnChanges(t){t.resetToDefault&&!t.resetToDefault.isFirstChange()&&this.resetToDefault&&this._resetThisComponent()}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processPhenotypeFile(t){this.phenoDataService.initLoadedFile(t)}processSelectedSampleColumn(t){this.sampleColumn$.next(t)}presentFirstLinesOfTheFile(t){this.presentFirstLinesModalRef=this.modalService.show(this.presentFirstLinesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}skipThisStep(t){this.skipSignal$.next(!0)}_resetThisComponent(){this._resetSelectedFile(),this.fileHeadersFromReset$.next(null),this.firstLinesFromReset$.next(null),this.phenotypeData$.next(null)}_resetSelectedFile(){this.fileForm.nativeElement.reset()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(He),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-phenotype-data"]],viewQuery:function(t,s){if(1&t&&(e.Gf(rs,7),e.Gf(ps,5,e.Rgc)),2&t){let n;e.iGM(n=e.CRH())&&(s.fileForm=n.first),e.iGM(n=e.CRH())&&(s.presentFirstLinesComponent=n.first)}},inputs:{resetToDefault:"resetToDefault",formWidth:"formWidth"},outputs:{phenotypeData$:"phenotypeData$",skipSignal$:"skipSignal$"},features:[e._Bn([He]),e.TTD],decls:18,vars:18,consts:[["role","form"],["fileForm",""],[1,"d-flex"],["type","file",1,"form-control","form-control-sm",3,"change"],[1,"ml-2",3,"figureSrc","figureLabel"],["href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],[1,"mt-2"],[3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","ifWidthAuto","selectedItem$"],[1,"align-self-start","ml-3","questionMark",3,"note"],["class","mt-2",4,"ngIf"],["presentFirstLinesComponent",""],["href","javascript:void(0)",1,"inline",3,"click"],[1,"text-muted"],["type","button","data-testid","skipImportPhenotype",1,"btn","btn-expressvis","btn-sm",3,"click"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"table-responsive","mr-2","ml-2"],[1,"table","table-striped","mt-2"],[4,"ngFor","ngForOf"],[1,"border-bottom"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"],[1,"font-weight-bold"],[3,"title",4,"ngFor","ngForOf"],[3,"title"]],template:function(t,s){1&t&&(e.TgZ(0,"form",0,1),e.TgZ(2,"h5"),e._uU(3,"Phenotype File: "),e.qZA(),e.TgZ(4,"div",2),e.TgZ(5,"input",3),e.NdJ("change",function(o){return s.processPhenotypeFile(o)}),e.qZA(),e._UZ(6,"app-display-a-figure-in-modal",4),e.qZA(),e.qZA(),e.YNc(7,ls,2,0,"a",5),e.ALo(8,"async"),e.TgZ(9,"div",6),e.TgZ(10,"div",2),e.TgZ(11,"app-select-item",7),e.NdJ("selectedItem$",function(o){return s.processSelectedSampleColumn(o)}),e.ALo(12,"async"),e.qZA(),e._UZ(13,"app-display-notes-with-question-mark",8),e.qZA(),e.qZA(),e.YNc(14,cs,5,0,"div",9),e.ALo(15,"async"),e.YNc(16,ds,17,2,"ng-template",null,10,e.W1O)),2&t&&(e.xp6(5),e.Udp("width",s.formWidth),e.xp6(1),e.Q6J("figureSrc",s.phenotypeFileExampleSrc)("figureLabel","View Example"),e.xp6(1),e.Q6J("ngIf",e.lcZ(8,12,s.firstLines$)),e.xp6(4),e.Q6J("inputItems",e.lcZ(12,14,s.fileHeaders$))("itemLabel","Select Sample Column")("ifOutputFirstValue",!0)("formWidth",s.formWidth)("ifWidthAuto",!1),e.xp6(2),e.Q6J("note","The default sample column is the first column"),e.xp6(1),e.Q6J("ngIf",!e.lcZ(15,16,s.fileHeaders$)))},directives:[d._Y,d.JL,d.F,ne,g.O5,te.y,U.m,g.sg],pipes:[g.Ov,Pe.K],styles:[""],changeDetection:0}),i})();var ve=m(4813);const fs=["submitButton"];function gs(i,a){1&i&&(e.TgZ(0,"p",6),e._uU(1," Import phenotype data first "),e.qZA())}function _s(i,a){if(1&i&&(e.TgZ(0,"option",11),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.Q6J("value",t),e.xp6(1),e.hij(" ",t," ")}}function Ss(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"form"),e.TgZ(1,"div",7),e.TgZ(2,"h5",8),e._uU(3,"Select Group Attribute In Phenotype Data: "),e.qZA(),e.TgZ(4,"select",9),e.NdJ("change",function(n){return e.CHM(t),e.oxw().defineGroupsAccordingToAttribute(n)}),e.YNc(5,_s,2,2,"option",10),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(4),e.Udp("width",t.formWidth),e.xp6(1),e.Q6J("ngForOf",t.allAttributes)}}function vs(i,a){if(1&i&&(e.TgZ(0,"tr",15),e.TgZ(1,"td"),e._uU(2),e.qZA(),e.TgZ(3,"td"),e._uU(4),e.qZA(),e.TgZ(5,"td"),e._UZ(6,"app-display-limited-items",16),e.qZA(),e.qZA()),2&i){const t=a.$implicit,s=a.index,n=e.oxw(2);e.Q6J("ngbTooltip","Click to drag and move group to re-order the groups"),e.xp6(2),e.hij(" ",t.name," "),e.xp6(2),e.hij(" ",t.samples.length," "),e.xp6(2),e.Q6J("items",n.groups[s].samples)("displayNum",3)}}function ys(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"table",12),e.TgZ(1,"thead"),e.TgZ(2,"tr"),e.TgZ(3,"th"),e._uU(4,"Group Name"),e.qZA(),e.TgZ(5,"th"),e._uU(6,"Samples Num"),e.qZA(),e.TgZ(7,"th"),e._uU(8,"Samples in the Group"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(9,"tbody",13),e.NdJ("cdkDropListDropped",function(n){return e.CHM(t),e.oxw().onDrop(n)}),e.YNc(10,vs,7,5,"tr",14),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(10),e.Q6J("ngForOf",t.groups)}}function As(i,a){if(1&i&&(e.TgZ(0,"button",17,18),e.ALo(2,"async"),e._uU(3,"Submit Groups "),e.qZA()),2&i){const t=e.oxw();e.Q6J("disabled",!e.lcZ(2,1,t.ifSubmitAvailable$))}}let bs=(()=>{class i{constructor(){this.sampleAttribute="samples",this.ifShowSubmitButton=!0,this.formWidth="15rem",this.outGroups=new e.vpe,this.groups=[],this.allAttributes=[],this.groupsChangedSubject=new B.xQ,this.destroyNotifier$=new B.xQ}ngOnInit(){this.phenotypeData&&(this.allAttributes=Object.keys(this.phenotypeData)),this.ifSubmitAvailable$=this.groupsChangedSubject.asObservable(),this.groupsChangedSubject.next(!1)}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&(this.allAttributes=Object.keys(this.phenotypeData))}ngAfterViewInit(){this.ifShowSubmitButton&&(this.submitStream$=(0,E.R)(this.submitButton.nativeElement,"click"),this.submitStream$.pipe((0,_.R)(this.destroyNotifier$),(0,r.h)(t=>null!=t)).subscribe(t=>{this.outGroups.next([...this.groups])}))}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}defineGroupsAccordingToAttribute(t){const s=[...this.phenotypeData[t.target.value]],n=[...this.phenotypeData[this.sampleAttribute]];let o=new Object;new Set(s).forEach(h=>{o[h]=[]}),s.forEach((h,p)=>{o[h].push(n[p])});const u=Object.keys(o).map(h=>({name:h,samples:o[h]}));this.groups=u,this.groupsChangedSubject.next(!0),this.ifShowSubmitButton||this.outGroups.next(u)}onDrop(t){(0,O.bA)(this.groups,t.previousIndex,t.currentIndex),this.ifShowSubmitButton||this.outGroups.next(this.groups)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-define-groups-using-phenotype-data"]],viewQuery:function(t,s){if(1&t&&e.Gf(fs,5),2&t){let n;e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},inputs:{phenotypeData:"phenotypeData",sampleAttribute:"sampleAttribute",ifShowSubmitButton:"ifShowSubmitButton",formWidth:"formWidth"},outputs:{outGroups:"outGroups"},features:[e.TTD],decls:6,vars:4,consts:[[1,"d-flex","justify-content-center"],["class","",4,"ngIf"],[4,"ngIf"],["class","table",4,"ngIf"],[1,"mt-3"],["class","btn btn-expressvis","data-testid","submitGroups",3,"disabled",4,"ngIf"],[1,""],[1,"form-group","form-group-sm"],["for","defineGroups",2,"margin-bottom","5px"],[1,"form-control","form-control-sm",3,"change"],[3,"value",4,"ngFor","ngForOf"],[3,"value"],[1,"table"],["cdkDropList","",3,"cdkDropListDropped"],["cdkDrag","","cdkDragLockAxis","y",3,"ngbTooltip",4,"ngFor","ngForOf"],["cdkDrag","","cdkDragLockAxis","y",3,"ngbTooltip"],[3,"items","displayNum"],["data-testid","submitGroups",1,"btn","btn-expressvis",3,"disabled"],["submitButton",""]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.YNc(1,gs,2,0,"p",1),e.qZA(),e.YNc(2,Ss,6,3,"form",2),e.YNc(3,ys,11,1,"table",3),e.TgZ(4,"div",4),e.YNc(5,As,4,3,"button",5),e.qZA()),2&t&&(e.xp6(1),e.Q6J("ngIf",!s.phenotypeData),e.xp6(1),e.Q6J("ngIf",s.phenotypeData),e.xp6(1),e.Q6J("ngIf",s.groups.length>=1),e.xp6(2),e.Q6J("ngIf",s.ifShowSubmitButton))},directives:[g.O5,d._Y,d.JL,d.F,g.sg,d.YN,d.Kr,O.Wj,O.Zt,x._L,ve.w],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();var xs=m(471),Ne=m(1789),we=m(9879),qe=m(3128),Ds=m(4250),Re=m(5903),$e=m(3991);const Ts=["submitButton"];function Cs(i,a){1&i&&(e.TgZ(0,"p"),e._uU(1," Import phenotype data first "),e.qZA())}function Is(i,a){if(1&i&&e._uU(0),2&i){const t=e.oxw().$implicit;e.hij(" ",t.get("name").value," ")}}function Zs(i,a){if(1&i&&e._UZ(0,"input",16),2&i){const t=e.oxw().index,s=e.oxw(2);e.Q6J("formControl",s.getControl(t,"name"))}}function Ps(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"tr"),e.TgZ(1,"td"),e.TgZ(2,"div",8),e.TgZ(3,"fa-icon",10),e.NdJ("click",function(){const o=e.CHM(t).index;return e.oxw(2).removeGroupFormGroup(o)}),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"td"),e.TgZ(5,"app-editable",11),e.YNc(6,Is,1,1,"ng-template",12),e.YNc(7,Zs,1,1,"ng-template",13),e.qZA(),e.qZA(),e.TgZ(8,"td"),e.TgZ(9,"app-define-filter-steps",14),e.NdJ("filterStepsInfo$",function(n){const l=e.CHM(t).index;return e.oxw(2).obtainGroupSamples(n,l)}),e.qZA(),e.qZA(),e.TgZ(10,"td"),e._UZ(11,"app-display-limited-items",15),e.qZA(),e.qZA()}if(2&i){const t=a.index,s=e.oxw(2);e.xp6(3),e.Q6J("icon",s.faMinus)("size","1x"),e.xp6(6),e.Q6J("dataAttributes",s.allAttributes),e.xp6(2),e.Q6J("items",s.getControlValue(t,"samples"))("displayNum",3)}}function Ns(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"table",6),e.TgZ(1,"thead"),e.TgZ(2,"tr"),e._UZ(3,"th"),e.TgZ(4,"th"),e._uU(5,"Group Name"),e.qZA(),e.TgZ(6,"th"),e._uU(7,"Filter"),e.qZA(),e.TgZ(8,"th"),e._uU(9,"Samples In the Group"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(10,"tbody"),e.YNc(11,Ps,12,5,"tr",7),e.TgZ(12,"tr"),e.TgZ(13,"td"),e.TgZ(14,"div",8),e.TgZ(15,"fa-icon",9),e.NdJ("click",function(){return e.CHM(t),e.oxw().addGroupFormGroup()}),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",t.groupsForm.get("groups").controls),e.xp6(4),e.Q6J("icon",t.faPlus)("size","1x")}}function ws(i,a){if(1&i&&(e.TgZ(0,"p",18),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij("Sample ",t.samplesString," are assigned to multiple groups")}}function qs(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,ws,2,1,"p",17),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf",t.ifOverlap)}}function Rs(i,a){1&i&&(e.TgZ(0,"p",19),e._uU(1,"Group names should be unique"),e.qZA())}function $s(i,a){1&i&&(e.TgZ(0,"p",19),e._uU(1,"Group names or samples should not be empty"),e.qZA())}function Bs(i,a){if(1&i&&(e.TgZ(0,"button",20,21),e.ALo(2,"async"),e._uU(3,"Submit Groups"),e.qZA()),2&i){const t=e.oxw();e.Q6J("disabled",!e.lcZ(2,1,t.ifSubmitAvailable$))}}let Fs=(()=>{class i{constructor(t,s){this.formBuilder=t,this.cdr=s,this.sampleAttribute="samples",this.ifShowSubmitButton=!0,this.outGroups=new e.vpe,this.faPlus=$.r8p,this.faMinus=$.Kl4,this.groups=[],this.destroyNotifier$=new B.xQ,this.groupsForm=this.formBuilder.group({groups:this.formBuilder.array([this._createGroupFormGroup()],[d.kI.required,Le,ke,Ve])})}ngOnInit(){this.phenotypeData&&this._processPhenotypeData(),this.ifSubmitAvailable$=this.groupsForm.get("groups").statusChanges.pipe((0,_.R)(this.destroyNotifier$),(0,f.U)(t=>"VALID"===t)),this.groupsForm.get("groups").statusChanges.pipe((0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.ifShowSubmitButton||this.outGroups.next("VALID"===t?this.groupsForm.get("groups").value:null)}),this.ifGroupsOverlapSamplesNote$=this.groupsForm.get("groups").valueChanges.pipe((0,_.R)(this.destroyNotifier$),(0,f.U)(t=>{const n=t.map(l=>l.samples).reduce((l,u)=>l.concat(u)),o=(0,v.ro)(n);if(o){const l=(0,v.kX)(n);return{ifOverlap:o,samplesString:(0,v.w6)(l)}}return{ifOverlap:o}}),(0,T.O)({ifOverlap:!1}))}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&this._processPhenotypeData()}ngAfterViewInit(){this.ifShowSubmitButton&&(this.submitStream$=(0,E.R)(this.submitButton.nativeElement,"click"),this.submitStream$.pipe((0,_.R)(this.destroyNotifier$),(0,r.h)(t=>null!=t)).subscribe(t=>{this.outGroups.next(this.groupsForm.get("groups").value)}))}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}_processPhenotypeData(){this.allAttributes=[...Object.keys(this.phenotypeData)],this.allSamples=[...this.phenotypeData[this.sampleAttribute]]}_createGroupFormGroup(){return new d.cw({name:new d.NI("New Group",d.kI.required),samples:new d.NI([],[d.kI.required,d.kI.minLength(1)])})}getControl(t,s){return this.groupsForm.get("groups").at(t).get(s)}getControlValue(t,s){return this.groupsForm.get("groups").at(t).get(s).value}addGroupFormGroup(){this.groupsForm.get("groups").push(this._createGroupFormGroup())}removeGroupFormGroup(t){this.groupsForm.get("groups").removeAt(t)}obtainGroupSamples(t,s){const n=t.matchType;let o=[],l=[];if(t.filterSteps.length>=1){for(let p of t.filterSteps){const S=p.attribute,A=p.filterType,N=p.filterValue.toLowerCase(),re=v.qP[A](this.phenotypeData[S],N);o.push(new Set(re))}l=[...o.reduce(xs.JN[n])].map(p=>this.allSamples[p])}this.groupsForm.get("groups").at(s).patchValue({samples:l})}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(d.qu),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-define-groups-using-multiple-attributes"]],viewQuery:function(t,s){if(1&t&&e.Gf(Ts,5),2&t){let n;e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},inputs:{phenotypeData:"phenotypeData",sampleAttribute:"sampleAttribute",ifShowSubmitButton:"ifShowSubmitButton"},outputs:{outGroups:"outGroups"},features:[e.TTD],decls:9,vars:8,consts:[[1,"d-flex","justify-content-center"],[4,"ngIf"],["class","table",4,"ngIf"],["class","alertP",4,"ngIf"],[1,"mt-3"],["class","btn btn-expressvis","data-testid","submitGroups",3,"disabled",4,"ngIf"],[1,"table"],[4,"ngFor","ngForOf"],[1,"form-inline"],["ngbTooltip","Add a new group","data-testid","addGroup",1,"hoverBigger",3,"icon","size","click"],["ngbTooltip","Delete this group",1,"hoverBigger",3,"icon","size","click"],["initialMode","edit"],["appEditableView",""],["appEditableEdit",""],[3,"dataAttributes","filterStepsInfo$"],[3,"items","displayNum"],["appEditableOnEnter","","appFocusable","","ngDefaultControl","","placeholder","Group Name","data-testid","inputGroupName","required","",1,"form-control","form-control-sm",3,"formControl"],["data-testid","samplesOverlapNote",4,"ngIf"],["data-testid","samplesOverlapNote"],[1,"alertP"],["data-testid","submitGroups",1,"btn","btn-expressvis",3,"disabled"],["submitButton",""]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.YNc(1,Cs,2,0,"p",1),e.qZA(),e.YNc(2,Ns,16,3,"table",2),e.YNc(3,qs,2,1,"div",1),e.ALo(4,"async"),e.YNc(5,Rs,2,0,"p",3),e.YNc(6,$s,2,0,"p",3),e.TgZ(7,"div",4),e.YNc(8,Bs,4,3,"button",5),e.qZA()),2&t&&(e.xp6(1),e.Q6J("ngIf",!s.phenotypeData),e.xp6(1),e.Q6J("ngIf",s.phenotypeData),e.xp6(1),e.Q6J("ngIf",e.lcZ(4,6,s.ifGroupsOverlapSamplesNote$)),e.xp6(2),e.Q6J("ngIf",null==s.groupsForm.get("groups").errors?null:s.groupsForm.get("groups").errors.groupNamesDuplicate),e.xp6(1),e.Q6J("ngIf",null==s.groupsForm.get("groups").errors?null:s.groupsForm.get("groups").errors.hasEmptySamples),e.xp6(2),e.Q6J("ngIf",s.ifShowSubmitButton))},directives:[g.O5,g.sg,P.BN,x._L,Ne.y,we.i,qe.O,Ds.u,ve.w,d.Fj,Re._,$e.x,d.Q7,d.JJ,d.oH],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();const Os=["submitButton"];function Ms(i,a){if(1&i&&e._uU(0),2&i){const t=e.oxw().$implicit;e.hij(" ",t.get("name").value," ")}}function Us(i,a){if(1&i&&e._UZ(0,"input",14),2&i){const t=e.oxw().index,s=e.oxw();e.Q6J("formControl",s.getControl(t,"name"))}}function Js(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"tr"),e.TgZ(1,"td"),e.TgZ(2,"div",2),e.TgZ(3,"fa-icon",8),e.NdJ("click",function(){const o=e.CHM(t).index;return e.oxw().removeGroupFormGroup(o)}),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"td"),e.TgZ(5,"app-editable",9),e.NdJ("update",function(){const o=e.CHM(t).index;return e.oxw().updateGroupName(o,"name")}),e.YNc(6,Ms,1,1,"ng-template",10),e.YNc(7,Us,1,1,"ng-template",11),e.qZA(),e.qZA(),e.TgZ(8,"td"),e.TgZ(9,"ng-multiselect-dropdown",12),e.NdJ("ngModelChange",function(n){const l=e.CHM(t).index;return e.oxw().groups[l].samples=n})("onSelect",function(n){const l=e.CHM(t).index;return e.oxw().updateGroupSamples(n,l)})("onDeSelect",function(n){const l=e.CHM(t).index;return e.oxw().updateGroupSamples(n,l)}),e.qZA(),e.qZA(),e.TgZ(10,"td"),e._UZ(11,"app-display-limited-items",13),e.qZA(),e.qZA()}if(2&i){const t=a.index,s=e.oxw();e.xp6(3),e.Q6J("icon",s.faMinus)("size","1x"),e.xp6(6),e.Q6J("placeholder","Select samples")("settings",s.dropdownSettings)("data",s.samples)("ngModel",s.groups[t].samples),e.xp6(2),e.Q6J("items",s.getControlValue(t,"samples"))("displayNum",3)}}function Qs(i,a){if(1&i&&(e.TgZ(0,"p",15),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij("Sample ",t.samplesString," are assigned to multiple groups")}}function Es(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,Qs,2,1,"p",5),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf",t.ifOverlap)}}function Gs(i,a){1&i&&(e.TgZ(0,"p",15),e._uU(1,"Group names should be unique"),e.qZA())}function js(i,a){if(1&i&&(e.TgZ(0,"button",16,17),e.ALo(2,"async"),e._uU(3,"Submit Groups"),e.qZA()),2&i){const t=e.oxw();e.Q6J("disabled",!e.lcZ(2,1,t.ifSubmitAvailable$))}}let Ls=(()=>{class i{constructor(t){this.formBuilder=t,this.ifShowSubmitButton=!0,this.outGroups=new e.vpe,this.faPlus=$.r8p,this.faMinus=$.Kl4,this.groups=[],this.dropdownSettings={singleSelection:!1,selectAllText:"Select All",unSelectAllText:"UnSelect All",itemsShowLimit:1,enableCheckAll:!1,allowSearchFilter:!0},this.destroyNotifier$=new B.xQ,this.groupsForm=this.formBuilder.group({groups:this.formBuilder.array([this._createGroupFormGroup()],[Le,ke,Ve])}),this.groups.push({name:"New Group",samples:[]})}ngOnInit(){this.groupsForm.get("groups").controls[0].get("name"),this.ifSubmitAvailable$=this.groupsForm.get("groups").statusChanges.pipe((0,_.R)(this.destroyNotifier$),(0,f.U)(t=>"VALID"===t)),this.groupsForm.get("groups").statusChanges.pipe((0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.ifShowSubmitButton||"VALID"===t&&this.outGroups.next(this.groupsForm.get("groups").value)}),this.ifGroupsOverlapSamplesNote$=this.groupsForm.get("groups").valueChanges.pipe((0,_.R)(this.destroyNotifier$),(0,f.U)(t=>{const n=t.map(l=>l.samples).reduce((l,u)=>l.concat(u)),o=(0,v.ro)(n);if(o){const l=(0,v.kX)(n);return{ifOverlap:o,samplesString:(0,v.w6)(l)}}return{ifOverlap:o}}),(0,T.O)({ifOverlap:!1}))}ngAfterViewInit(){this.ifShowSubmitButton&&(this.submitStream$=(0,E.R)(this.submitButton.nativeElement,"click"),this.submitStream$.pipe((0,_.R)(this.destroyNotifier$),(0,r.h)(t=>null!=t)).subscribe(t=>{this.outGroups.next(this.groupsForm.get("groups").value)}))}ngOnChanges(t){}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}_createGroupFormGroup(){return new d.cw({name:new d.NI("New Group",d.kI.required),samples:new d.NI([],d.kI.required)})}getControl(t,s){return this.groupsForm.get("groups").at(t).get(s)}getControlValue(t,s){return this.groupsForm.get("groups").at(t).get(s).value}addGroupFormGroup(){this.groupsForm.get("groups").push(this._createGroupFormGroup()),this.groups.push({name:"New Group",samples:[]})}removeGroupFormGroup(t){this.groupsForm.get("groups").removeAt(t),this.groups=this.groups.filter((n,o)=>o!=t)}updateGroupName(t,s){const n=this.groupsForm.get("groups").controls[t].get("name").value,o=[...this.groups];this.groups=o.map((l,u)=>u===t?Object.assign(Object.assign({},l),{name:n}):Object.assign({},l))}updateGroupSamples(t,s){this.groupsForm.get("groups").controls[s].patchValue({samples:this.groups[s].samples})}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(d.qu))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-define-groups-manually"]],viewQuery:function(t,s){if(1&t&&e.Gf(Os,5),2&t){let n;e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},inputs:{samples:"samples",ifShowSubmitButton:"ifShowSubmitButton"},outputs:{outGroups:"outGroups"},features:[e.TTD],decls:21,vars:8,consts:[[1,"table"],[4,"ngFor","ngForOf"],[1,"form-inline"],["ngbTooltip","Add a new group",1,"hoverBigger",3,"icon","size","click"],[4,"ngIf"],["class","alertP",4,"ngIf"],[1,"mt-3"],["class","btn btn-expressvis","data-testid","submitGroups",3,"disabled",4,"ngIf"],["ngbTooltip","Delete this group",1,"hoverBigger",3,"icon","size","click"],["initialMode","edit",3,"update"],["appEditableView",""],["appEditableEdit",""],["ngDefaultControl","",3,"placeholder","settings","data","ngModel","ngModelChange","onSelect","onDeSelect"],[3,"items","displayNum"],["appEditableOnEnter","","appFocusable","","ngDefaultControl","","placeholder","Group Name","required","",1,"form-control","form-control-sm",3,"formControl"],[1,"alertP"],["data-testid","submitGroups",1,"btn","btn-expressvis",3,"disabled"],["submitButton",""]],template:function(t,s){1&t&&(e.TgZ(0,"table",0),e.TgZ(1,"thead"),e.TgZ(2,"tr"),e._UZ(3,"th"),e.TgZ(4,"th"),e._uU(5,"Group Name"),e.qZA(),e.TgZ(6,"th"),e._uU(7,"Manually Select"),e.qZA(),e.TgZ(8,"th"),e._uU(9,"Samples In the Group"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(10,"tbody"),e.YNc(11,Js,12,8,"tr",1),e.TgZ(12,"tr"),e.TgZ(13,"td"),e.TgZ(14,"div",2),e.TgZ(15,"fa-icon",3),e.NdJ("click",function(){return s.addGroupFormGroup()}),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.YNc(16,Es,2,1,"div",4),e.ALo(17,"async"),e.YNc(18,Gs,2,0,"p",5),e.TgZ(19,"div",6),e.YNc(20,js,4,3,"button",7),e.qZA()),2&t&&(e.xp6(11),e.Q6J("ngForOf",s.groupsForm.get("groups").controls),e.xp6(4),e.Q6J("icon",s.faPlus)("size","1x"),e.xp6(1),e.Q6J("ngIf",e.lcZ(17,6,s.ifGroupsOverlapSamplesNote$)),e.xp6(2),e.Q6J("ngIf",null==s.groupsForm.get("groups").errors?null:s.groupsForm.get("groups").errors.groupNamesDuplicate),e.xp6(2),e.Q6J("ngIf",s.ifShowSubmitButton))},directives:[g.sg,P.BN,x._L,g.O5,Ne.y,we.i,qe.O,C.OP,d.Fj,d.JJ,d.On,ve.w,Re._,$e.x,d.Q7,d.oH],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();function ks(i,a){if(1&i&&(e.TgZ(0,"th"),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.hij("",t," ")}}function Vs(i,a){if(1&i&&(e.TgZ(0,"td",5),e._uU(1),e.ALo(2,"trimString"),e.qZA()),2&i){const t=a.$implicit,s=e.oxw().$implicit;e.s9C("title",s[t]),e.xp6(1),e.hij("",e.xi3(2,2,s[t],20)," ")}}function Hs(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,Vs,3,5,"td",4),e.qZA()),2&i){const t=e.oxw(2);e.xp6(1),e.Q6J("ngForOf",t.allAttributes)}}function Ys(i,a){if(1&i&&(e.TgZ(0,"table",2),e.TgZ(1,"thead"),e.TgZ(2,"tr"),e.YNc(3,ks,2,1,"th",3),e.qZA(),e.qZA(),e.TgZ(4,"tbody"),e.YNc(5,Hs,2,1,"tr",3),e.qZA(),e.qZA()),2&i){const t=e.oxw();e.xp6(3),e.Q6J("ngForOf",t.allAttributes),e.xp6(2),e.Q6J("ngForOf",t.phenotypeDataArray)}}let Ws=(()=>{class i{constructor(){this.allAttributes=[],this.phenotypeDataArray=[]}ngOnInit(){this.phenotypeData&&this._processPhenotypeData()}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&this._processPhenotypeData()}_processPhenotypeData(){this.allAttributes=Object.keys(this.phenotypeData),this.phenotypeDataArray=(0,v.gG)(this.phenotypeData)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-present-phenotype-data"]],inputs:{phenotypeData:"phenotypeData"},features:[e.TTD],decls:2,vars:1,consts:[[1,"table-responsive","border"],["class","table table-striped",4,"ngIf"],[1,"table","table-striped"],[4,"ngFor","ngForOf"],[3,"title",4,"ngFor","ngForOf"],[3,"title"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.YNc(1,Ys,6,2,"table",1),e.qZA()),2&t&&(e.xp6(1),e.Q6J("ngIf",s.phenotypeDataArray.length>=1))},directives:[g.O5,g.sg],pipes:[Pe.K],styles:["table[_ngcontent-%COMP%]{max-height:80vh}tbody[_ngcontent-%COMP%]{text-align:start;overflow-y:scroll;overflow-x:hidden}td[_ngcontent-%COMP%], th[_ngcontent-%COMP%]{text-align:left;height:2em;white-space:nowrap}th[sortable][_ngcontent-%COMP%]{cursor:pointer}"],changeDetection:0}),i})();const zs=["phenotypeContent"];function Xs(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",8),e.TgZ(1,"input",9),e.NdJ("change",function(){const o=e.CHM(t).$implicit;return e.oxw().updateCheckbox(o)}),e.qZA(),e.TgZ(2,"label"),e._uU(3),e.qZA(),e.qZA()}if(2&i){const t=a.$implicit,s=e.oxw();e.xp6(1),e.Q6J("checked",t===s.currentMethod),e.xp6(2),e.Oqu(t)}}function Ks(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-define-groups-using-phenotype-data",10),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("ifShowSubmitButton",t.ifShowSubmitButton)("phenotypeData",t.phenotypeData)("formWidth",t.formWidth)}}function ei(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-define-groups-using-multiple-attributes",11),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("phenotypeData",t.phenotypeData)("ifShowSubmitButton",t.ifShowSubmitButton)}}function ti(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-define-groups-manually",12),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("ifShowSubmitButton",t.ifShowSubmitButton)("samples",t.samples)}}function si(i,a){1&i&&(e.TgZ(0,"p",13),e._uU(1," Select one Method\n"),e.qZA())}function ii(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",14),e.TgZ(1,"h4",15),e._uU(2," Phenotype data"),e.qZA(),e.TgZ(3,"button",16),e.NdJ("click",function(){return e.CHM(t).$implicit.dismiss("Cross click")}),e.TgZ(4,"span",17),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",18),e._UZ(7,"app-present-phenotype-data",19),e.qZA()}if(2&i){const t=e.oxw();e.xp6(7),e.Q6J("phenotypeData",t.phenotypeData)}}let ye=(()=>{class i{constructor(t){this.modalService=t,this.ifShowSubmitButton=!0,this.resetToDefault=!1,this.formWidth="15rem",this.outGroups=new e.vpe,this.DEFINE_GROUP_METHODS=["Assign Using Group Attribute","Filter Using Multiple Attributes","Manually Select"],this.currentMethod="",this.samplesAttribte="samples"}ngOnInit(){this.phenotypeData&&(this.samples=this._obtainSamples())}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&this.phenotypeData&&(this.samples=this._obtainSamples()),t.resetToDefault&&!t.resetToDefault.isFirstChange()&&this.resetToDefault&&(this.currentMethod="",this.outGroups.next(null))}_obtainSamples(){return[...this.phenotypeData[this.samplesAttribte]]}updateCheckbox(t){this.currentMethod=this.currentMethod!=t?t:""}processGroups(t){this.outGroups.next((null==t?void 0:t.length)>0?[...t]:[])}displayTableToDisplayPhenotype(t){this.modalService.open(this.phenotypeContent,{size:"xl",backdrop:"static"})}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(x.FF))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-define-groups"]],viewQuery:function(t,s){if(1&t&&e.Gf(zs,7),2&t){let n;e.iGM(n=e.CRH())&&(s.phenotypeContent=n.first)}},inputs:{phenotypeData:"phenotypeData",ifShowSubmitButton:"ifShowSubmitButton",resetToDefault:"resetToDefault",formWidth:"formWidth"},outputs:{outGroups:"outGroups"},features:[e.TTD],decls:23,vars:5,consts:[["href","javascript:void(0)","data-testid","displayPhenotype",1,"inline",3,"click"],[1,"d-flex","flex-wrap","justify-content-around"],["data-testid","selectGroupMethod",4,"ngFor","ngForOf"],[3,"ifShowSubmitButton","phenotypeData","formWidth","outGroups",4,"ngIf"],[3,"phenotypeData","ifShowSubmitButton","outGroups",4,"ngIf"],[3,"ifShowSubmitButton","samples","outGroups",4,"ngIf"],["class","text-danger",4,"ngIf"],["phenotypeContent",""],["data-testid","selectGroupMethod"],["type","checkbox",3,"checked","change"],[3,"ifShowSubmitButton","phenotypeData","formWidth","outGroups"],[3,"phenotypeData","ifShowSubmitButton","outGroups"],[3,"ifShowSubmitButton","samples","outGroups"],[1,"text-danger"],["calss","modal-header"],[1,"modal-title"],["type","button","aria-label","Close",1,"close",3,"click"],["aria-hidden","true"],[1,"modal-body"],[3,"phenotypeData"]],template:function(t,s){1&t&&(e.TgZ(0,"p"),e._uU(1," Defining samples in each group is the basis of differential analysis and clustering analysis. Three methods are supported:"),e.qZA(),e.TgZ(2,"ul"),e.TgZ(3,"li"),e._uU(4,"Assign by Group Attribute: A phenotype file containing group attribute need to be uploaded. Group mapping will be complete based on values of the group attribute (For hundreds of samples or more). "),e.TgZ(5,"a",0),e.NdJ("click",function(o){return s.displayTableToDisplayPhenotype(o)}),e._uU(6," View Phenotype Data "),e.qZA(),e.qZA(),e.TgZ(7,"li"),e._uU(8," Filter by Multiple Attributes: A phenotype file need to be upload. Select samples based on sample annotations. "),e.TgZ(9,"a",0),e.NdJ("click",function(o){return s.displayTableToDisplayPhenotype(o)}),e._uU(10," View Phenotype Data "),e.qZA(),e.qZA(),e.TgZ(11,"li"),e._uU(12," Manually Select: Specify the groups and select the corresponding samples (For small-scale datasets, phenotype file is not required). "),e.qZA(),e.qZA(),e.TgZ(13,"h5"),e._uU(14,"Select one method:"),e.qZA(),e.TgZ(15,"div",1),e.YNc(16,Xs,4,2,"div",2),e.qZA(),e.YNc(17,Ks,1,3,"app-define-groups-using-phenotype-data",3),e.YNc(18,ei,1,2,"app-define-groups-using-multiple-attributes",4),e.YNc(19,ti,1,2,"app-define-groups-manually",5),e.YNc(20,si,2,0,"p",6),e.YNc(21,ii,8,1,"ng-template",null,7,e.W1O)),2&t&&(e.xp6(16),e.Q6J("ngForOf",s.DEFINE_GROUP_METHODS),e.xp6(1),e.Q6J("ngIf","Assign Using Group Attribute"===s.currentMethod),e.xp6(1),e.Q6J("ngIf","Filter Using Multiple Attributes"===s.currentMethod),e.xp6(1),e.Q6J("ngIf","Manually Select"===s.currentMethod),e.xp6(1),e.Q6J("ngIf",""===s.currentMethod))},directives:[g.sg,g.O5,bs,Fs,Ls,Ws],styles:[""],changeDetection:0}),i})();const ni=["submitButton"];function ai(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",11),e.NdJ("click",function(n){return e.CHM(t),e.oxw().showPCAplotAfterRemovingBatch(n)}),e._uU(1," PCA plot "),e.qZA()}}let Ae=(()=>{class i extends ie.e{constructor(){super(),this.ifBatchEffectHasBeenRemoved=!1,this.displayNormalizedPCAsignal$=new e.vpe,this.displayPCAofValuesAfterRemovingBatchSignal$=new e.vpe,this.removeBatchInfo$=new e.vpe,this.yesOrNo=["Yes","No"],this.batchAttributeSubject=new c.X(null),this.ifUseGroupVariableSubject=new c.X(null)}ngOnInit(){}ngAfterViewInit(){this.ifSubmitAvailable$=(0,b.aj)([this.batchAttributeSubject,this.ifUseGroupVariableSubject]).pipe((0,f.U)(([t,s])=>null!=t&&null!=s),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$)),(0,E.R)(this.submitButton.nativeElement,"click").pipe((0,r.h)(t=>null!=t),(0,y.M)(this.batchAttributeSubject,this.ifUseGroupVariableSubject),(0,f.U)(([t,s,n])=>s&&n?"Yes"===n?{ifRemoveBatch:!0,batchAttribute:s,ifUseGroupVariable:!0}:{ifRemoveBatch:!0,batchAttribute:s,ifUseGroupVariable:!1}:null),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.removeBatchInfo$.next(t)})}showPCAplotBeforeRemovingBatch(t){this.displayNormalizedPCAsignal$.next(!0)}processSelectedAttribute(t){this.batchAttributeSubject.next(t)}processYesOrNo(t){this.ifUseGroupVariableSubject.next(t)}showPCAplotAfterRemovingBatch(t){this.displayPCAofValuesAfterRemovingBatchSignal$.next(!0)}skipThisStep(t){this.removeBatchInfo$.next({ifRemoveBatch:!1})}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-remove-batch-settings"]],viewQuery:function(t,s){if(1&t&&e.Gf(ni,7),2&t){let n;e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},inputs:{phenotypeDataAttributes:"phenotypeDataAttributes",ifBatchEffectHasBeenRemoved:"ifBatchEffectHasBeenRemoved"},outputs:{displayNormalizedPCAsignal$:"displayNormalizedPCAsignal$",displayPCAofValuesAfterRemovingBatchSignal$:"displayPCAofValuesAfterRemovingBatchSignal$",removeBatchInfo$:"removeBatchInfo$"},features:[e.qOj],decls:34,vars:14,consts:[["href","https://bioconductor.org/packages/release/bioc/html/sva.html"],["type","button","data-testid","skipBatchRemoveStep",1,"inline","btn","btn-expressvis","btn-sm",3,"ngbTooltip","click"],["data-testid","displayPCAbeforeRemovingBatch","href","javascript:void(0)",1,"inline",3,"click"],[1,"mt-2"],["data-testid","batchColumn",3,"inputItems","itemLabel","ifOutputFirstValue","selectedItem$"],["data-testid","ifUseGroups",3,"inputItems","itemLabel","ifOutputFirstValue","ifShowQuestionNote","questionNote","selectedItem$"],["type","button","data-testid","submitRemoveBatchSettings",1,"btn","btn-expressvis","btn-sm",3,"disabled"],["submitButton",""],["data-testid","displayPCAafterRemovingBatch","href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],[1,"text-muted","mt-3"],["type","button","data-testid","skipBatchRemoveStep",1,"btn","btn-expressvis","btn-sm",3,"ngbTooltip","click"],["data-testid","displayPCAafterRemovingBatch","href","javascript:void(0)",1,"inline",3,"click"]],template:function(t,s){1&t&&(e.TgZ(0,"p"),e._uU(1," This step is used to remove batch effects caused by laboratory conditions, choice of reagent, instruments used, etc. For more details, please see "),e.TgZ(2,"a",0),e._uU(3,"SVA"),e.qZA(),e._uU(4,". For most datasets, this step can be skipped. "),e.TgZ(5,"button",1),e.NdJ("click",function(o){return s.skipThisStep(o)}),e._uU(6," Skip\n"),e.qZA(),e.qZA(),e.TgZ(7,"p"),e._uU(8,"A phenotype data file containing "),e.TgZ(9,"strong"),e._uU(10,"batch"),e.qZA(),e._uU(11," attribute need to be uploaded."),e.qZA(),e.TgZ(12,"p"),e._uU(13,"Before removing batch effects, view "),e.TgZ(14,"a",2),e.NdJ("click",function(o){return s.showPCAplotBeforeRemovingBatch(o)}),e._uU(15," PCA plot "),e.qZA(),e._uU(16," to check if batch effects exist\n"),e.qZA(),e.TgZ(17,"div",3),e.TgZ(18,"app-select-item",4),e.NdJ("selectedItem$",function(o){return s.processSelectedAttribute(o)}),e.qZA(),e.qZA(),e.TgZ(19,"div",3),e.TgZ(20,"app-select-item",5),e.NdJ("selectedItem$",function(o){return s.processYesOrNo(o)}),e.qZA(),e.qZA(),e.TgZ(21,"div",3),e.TgZ(22,"button",6,7),e.ALo(24,"async"),e._uU(25," Submit to Remove Batch Effects "),e.qZA(),e.qZA(),e.TgZ(26,"p",3),e._uU(27," View "),e.YNc(28,ai,2,0,"a",8),e._uU(29," to check if removing batch effects works "),e.qZA(),e.TgZ(30,"p",9),e._uU(31,"If remove batch effects does not work, you can click Skip button to use values without removing batch effects"),e.qZA(),e.TgZ(32,"button",10),e.NdJ("click",function(o){return s.skipThisStep(o)}),e._uU(33," Skip\n"),e.qZA()),2&t&&(e.xp6(5),e.Q6J("ngbTooltip","Click to skip this step"),e.xp6(13),e.Q6J("inputItems",s.phenotypeDataAttributes)("itemLabel","Select Batch Attribute in Phenotype Data")("ifOutputFirstValue",!1),e.xp6(2),e.Q6J("inputItems",s.yesOrNo)("itemLabel","If Use Group Variable")("ifOutputFirstValue",!0)("ifShowQuestionNote",!0)("questionNote","If Yes, then the group factor will be specified as a biological covariate, their signals will be preserved in the adjusted data. Default is Yes. For more details, see SVA package documentation"),e.xp6(2),e.Q6J("disabled",!e.lcZ(24,12,s.ifSubmitAvailable$)),e.xp6(6),e.Q6J("ngIf",s.ifBatchEffectHasBeenRemoved),e.xp6(4),e.Q6J("ngbTooltip","Click to skip this step"))},directives:[x._L,te.y,g.O5],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();var be=m(9769);let xe=(()=>{class i{constructor(){}ngOnInit(){}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-express-data-success-import-note"]],decls:2,vars:0,consts:[[1,"text-success"]],template:function(t,s){1&t&&(e.TgZ(0,"p",0),e._uU(1," Data is uploaded successfully, click one of the apps for later analysis. Importing data may take a long time, you should download the project to local storage before closing the browser and then re-load it to restore your analysis later.\n"),e.qZA())},styles:[""],changeDetection:0}),i})();var Z=m(3203),oi=m(8105),Be=m(4080);const ri=["svgElement"],pi=["dragRectangle"];let Ye=(()=>{class i{constructor(t){this.elRef=t,this.selectedSamples=new e.vpe,this.svgWidth=430,this.svgHeight=330,this.marginTop=10,this.marginBottom=80,this.marginLeft=80,this.marginRight=80,this.circleRadius=3,this.circleRadiusBigger=5,this.annotationBarWidth=15,this.annotationLableHeight=15,this.traceDragEvent={startX:0,startY:0},this.pointsPosition=[],this.changeDragRect$=new c.X(null),this.hostElement=this.elRef.nativeElement}ngOnInit(){this.changeDragRect$.pipe((0,r.h)(t=>null!=t),(0,oi.e)(200),(0,r.h)(t=>t.width>=0&&t.height>=0)).subscribe(t=>{this._modifyDragRectPosWidthHeight(t),this._toggleDragRectVisibility("visible");const s=this._obtainSelectedSamplesOfUsingDragRect(t);this._highlightSelectedSamples(s)})}ngAfterViewInit(){var t;(null===(t=this.PCAinfoArray)||void 0===t?void 0:t.length)>0&&this._initMap()}ngOnChanges(t){var s;t.PCAinfoArray&&!t.PCAinfoArray.isFirstChange()&&(null===(s=this.PCAinfoArray)||void 0===s?void 0:s.length)>0&&(this._removeGraphic(),this._initMap()),t.highlightSamplesInfo&&!t.highlightSamplesInfo.isFirstChange()&&(this._removeGraphic(),this._initMap())}ngOnDestroy(){this.toolTipDiv&&this.toolTipDiv.remove()}_initMap(){this._prepareSettings(),this._parpareForSamplesAnnotations(),this._addGraphic(),this._addDragRect(),this._addDragRectEvent()}_prepareSettings(){const t=this;t.graphicWidth=t.svgWidth-t.marginLeft-t.marginRight,t.graphicHeight=t.svgHeight-t.marginTop-t.marginBottom;const s=this.PCAinfoArray.map(A=>A.firstCompValue),n=(0,Z.VV$)(s),o=(0,Z.Fp7)(s),l=(o-n)/10;t.xScale=(0,Z.BYU)().domain([n-l,o+l]).range([0,t.graphicWidth]);const u=this.PCAinfoArray.map(A=>A.secondCompValue),h=(0,Z.VV$)(u),p=(0,Z.Fp7)(u),S=(p-h)/10;t.yScale=(0,Z.BYU)().domain([h-S,p+S]).range([t.graphicHeight,0])}_removeGraphic(){(0,Z.Ys)(this.svgElement.nativeElement).selectAll("*").remove()}_addGraphic(){var t=this;this.d3Svg=(0,Z.Ys)(this.svgElement.nativeElement),this.d3Svg.attr("width",t.svgWidth).attr("height",t.svgHeight),t.toolTipDiv=(0,Z.Ys)("body").append("div").attr("class","d3Tooltip").style("opacity",0),this._addAxises(),this._addPoints(),this._addSamplesAnnotationsBars(),this._addPointsBorder()}_addAxises(){const t=this,s=t.marginTop+t.graphicHeight,o=s+this.d3Svg.append("g").attr("transform","translate("+t.marginLeft+","+s+")").call((0,Z.LLu)(t.xScale).tickFormat(N=>N.toString())).node().getBBox().height+20,l=t.marginLeft+.5*t.graphicWidth;this.d3Svg.append("text").attr("transform","translate("+l+","+o+")").style("text-anchor","middle").style("font-size",10).text("PC1");let h=this.d3Svg.append("g").attr("transform","translate("+t.marginLeft+","+t.marginTop+")").call((0,Z.y4O)(t.yScale).tickFormat(N=>N.toString()));const p=t.marginLeft-h.node().getBBox().width-10,S=t.marginTop+.5*t.graphicWidth;this.d3Svg.append("text").attr("transform","translate("+p+","+S+")rotate(-90)").style("text-anchor","middle").style("font-size",10).text("PC2")}_addPoints(){const t=this;this.d3Svg.append("g").selectAll("samplePoints").data(t.PCAinfoArray).enter().append("circle").attr("transform","translate("+t.marginLeft+","+t.marginTop+")").attr("id",function(n){return n.sampleName}).attr("cx",function(n){return t.xScale(n.firstCompValue)}).attr("cy",function(n){return t.yScale(n.secondCompValue)}).attr("r",t.circleRadius).attr("stroke",function(n){return t.attributes2colors?t.attributes2colors.get(t.samples2attributes.get(n.sampleName)):(0,Be.y)(1)[0]}).attr("class","pointer").style("fill",function(n){return t.attributes2colors?t.attributes2colors.get(t.samples2attributes.get(n.sampleName)):(0,Be.y)(1)[0]}).on("mousemove",function(n,o){t.toolTipDiv.transition().duration(100).style("opacity",1),t.toolTipDiv.html(o.sampleName).style("left",n.pageX+"px").style("top",n.pageY-20+"px")}).on("mouseout",function(n){t.toolTipDiv.transition().duration(200).style("opacity",0)}),t.pointsPosition=t.PCAinfoArray.map(n=>({sample:n.sampleName,x:t.xScale(n.firstCompValue)+t.marginLeft,y:t.yScale(n.secondCompValue)+t.marginTop}))}_addSamplesAnnotationsBars(){const t=this;if(this.highlightSamplesInfo){const s=20,n=t.marginLeft+t.graphicWidth+5,o=t.d3Svg.append("g").attr("transform","translate("+n+","+(t.marginTop+s)+")");o.append("text").text(this.highlightSamplesInfo.attribute);const l=[...this.highlightSamplesInfo.attributes2samples.keys()];o.append("g").attr("transform","translate(0,2)").selectAll(".sampleAttrRect").data(l).enter().append("rect").attr("class","grid").attr("width",t.annotationBarWidth).attr("height",t.annotationLableHeight).attr("transform",function(u,h){return"translate(0,"+h*t.annotationLableHeight+")"}).style("fill",function(u,h){return t.attributes2colors.get(u)}),o.append("g").attr("transform","translate("+(t.annotationBarWidth+2)+", 2)").selectAll(".sampleAttrValues").data(l).enter().append("text").attr("transform",function(u,h){return"translate(2,"+(h+.5)*t.annotationLableHeight+")"}).text(function(u){return u}).attr("dy","0.32em").style("font-size","0.7em").style("text-anchor","start")}}_addPointsBorder(){const t=this;t.d3Svg.append("rect").attr("transform","translate("+t.marginLeft+","+t.marginTop+")").attr("x",0).attr("y",0).attr("height",t.graphicHeight).attr("width",t.graphicWidth).attr("fill","none").attr("stroke","black")}_parpareForSamplesAnnotations(){var t;if(null===(t=this.highlightSamplesInfo)||void 0===t?void 0:t.attributes2samples){const s=this._obtainSamples2attributes(),n=[...this.highlightSamplesInfo.attributes2samples.keys()],o=(0,Be.y)(n.length),l=new Map;for(const[u,h]of(0,v.$R)(n,o))l.set(u,h);this.attributes2colors=l,this.samples2attributes=s}}_addDragRectEvent(){var t=this,s=(0,Z.ohM)().on("start",function(n){t.traceDragEvent={startX:n.x,startY:n.y},t.changeDragRect$.next({x:t.traceDragEvent.startX,y:t.traceDragEvent.startY,width:0,height:0})}).on("drag",function(n){t.changeDragRect$.next({x:t.traceDragEvent.startX,y:t.traceDragEvent.startY,width:n.x-t.traceDragEvent.startX,height:n.y-t.traceDragEvent.startY})}).on("end",function(n){t._toggleDragRectVisibility("hidden")});this.d3Svg.call(s)}_addDragRect(){(0,Z.Ys)(this.dragRectangle.nativeElement).append("rect").attr("id","dragRect").attr("height",20).attr("width",20).attr("x",1).attr("y",1).attr("visibility","hidden").attr("fill","#f2f2f2")}_toggleDragRectVisibility(t){(0,Z.Ys)(this.dragRectangle.nativeElement).select("rect").attr("visibility",t)}_modifyDragRectPosWidthHeight(t){const{x:s,y:n,width:o,height:l}=t;(0,Z.Ys)("#dragRect").attr("x",s).attr("y",n).attr("width",o).attr("height",l)}_obtainSelectedSamplesOfUsingDragRect(t){let s=[];return this.pointsPosition.forEach(n=>{this._determineIfPointerInAnRectArea({x:n.x,y:n.y},t)&&s.push(n.sample)}),s}_determineIfPointerInAnRectArea(t,s){const h=t.x,p=t.y;return h>=s.x&&h<=s.x+s.width&&p>=s.y&&p<=s.y+s.height}_highlightSelectedSamples(t){if((0,Z.td_)(".pointer").attr("r",this.circleRadius),t.length>=1){const n=t.map(o=>"#"+o).join(",");(0,Z.td_)(n).attr("r",this.circleRadiusBigger)}}_obtainSamples2attributes(){let t=new Map;return this.highlightSamplesInfo.attributes2samples.forEach((s,n)=>{s.forEach(o=>{t.set(o,n)})}),t}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.SBq))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-samples-pca-plot"]],viewQuery:function(t,s){if(1&t&&(e.Gf(ri,5),e.Gf(pi,5)),2&t){let n;e.iGM(n=e.CRH())&&(s.svgElement=n.first),e.iGM(n=e.CRH())&&(s.dragRectangle=n.first)}},inputs:{PCAinfoArray:"PCAinfoArray",highlightSamplesInfo:"highlightSamplesInfo"},outputs:{selectedSamples:"selectedSamples"},features:[e.TTD],decls:4,vars:0,consts:[["svgElement",""],["dragRectangle",""]],template:function(t,s){1&t&&(e.O4$(),e.TgZ(0,"svg",null,0),e._UZ(2,"g",null,1),e.qZA())},styles:[".tooltip[_ngcontent-%COMP%]{position:absolute;text-align:center;background:lightsteelblue}"],changeDetection:0}),i})();var We=m(2326);function li(i,a){if(1&i&&(e.TgZ(0,"option",6),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.Q6J("value",t),e.xp6(1),e.hij(" ",t," ")}}let De=(()=>{class i{constructor(t){this.cdr=t}ngOnInit(){this.phenotypeData&&(this.annotationNames=Object.keys(this.phenotypeData))}ngAfterViewInit(){this.cdr.detectChanges()}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&this.phenotypeData&&(this.annotationNames=Object.keys(this.phenotypeData))}changeSampleAnnotationType(t){const s=t.target.value,n=this.phenotypeData.samples,o=this.phenotypeData[s];let l=new Map;for(let[u,h]of(0,v.$R)(n,o))if(l.has(h)){const p=l.get(h);p.push(u),l.set(h,p)}else l.set(h,[u]);this.highlightSamplesInfo={attribute:s,attributes2samples:l},this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-display-pca-plot-before-removing-batch"]],inputs:{PCAinfoArray:"PCAinfoArray",phenotypeData:"phenotypeData"},features:[e.TTD],decls:10,vars:7,consts:[["data-testid","selectAttributeForPCA",1,"form-control","form-control-sm","ml-1","form-inline",2,"width","10rem",3,"change"],[3,"value",4,"ngFor","ngForOf"],[1,"d-flex","justify-content-end","align-items-center"],[1,"align-self-center","ml-3","questionMark",3,"note","placement"],[1,"d-flex","justify-content-center"],["appD3ExportMap","",3,"pdfName","pngName","PCAinfoArray","highlightSamplesInfo"],[3,"value"]],template:function(t,s){1&t&&(e.TgZ(0,"h5"),e._uU(1,"Select Batch Column in Phenotype Data:"),e.qZA(),e.TgZ(2,"select",0),e.NdJ("change",function(o){return s.changeSampleAnnotationType(o)}),e.TgZ(3,"option"),e._uU(4,"---"),e.qZA(),e.YNc(5,li,2,2,"option",1),e.qZA(),e.TgZ(6,"div",2),e._UZ(7,"app-display-notes-with-question-mark",3),e.qZA(),e.TgZ(8,"div",4),e._UZ(9,"app-samples-pca-plot",5),e.qZA()),2&t&&(e.xp6(5),e.Q6J("ngForOf",s.annotationNames),e.xp6(2),e.Q6J("note","The horizontal and vertical coordinates represent PC1 and PC2 respectively. Each dot represents one sample. Samples from different batches are marked with different colors")("placement","bottom"),e.xp6(2),e.Q6J("pdfName","before removing batch effects.pdf")("pngName","before removing batch effects.png")("PCAinfoArray",s.PCAinfoArray)("highlightSamplesInfo",s.highlightSamplesInfo))},directives:[d.YN,d.Kr,g.sg,U.m,Ye,We.I],styles:[""],changeDetection:0}),i})();function ci(i,a){if(1&i&&(e.TgZ(0,"option",11),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.Q6J("value",t),e.xp6(1),e.hij(" ",t," ")}}let Te=(()=>{class i{constructor(t){this.cdr=t}ngOnInit(){this.phenotypeData&&(this.annotationNames=Object.keys(this.phenotypeData))}ngAfterViewInit(){this.cdr.detectChanges()}ngOnChanges(t){t.phenotypeData&&!t.phenotypeData.isFirstChange()&&this.phenotypeData&&(this.annotationNames=Object.keys(this.phenotypeData))}changeSampleAnnotationType(t){const s=t.target.value,n=this.phenotypeData.samples,o=this.phenotypeData[s];let l=new Map;for(let[u,h]of(0,v.$R)(n,o))if(l.has(h)){const p=l.get(h);p.push(u),l.set(h,p)}else l.set(h,[u]);this.highlightSamplesInfo={attribute:s,attributes2samples:l},this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-display-pcas-before-and-after-removing-batch"]],inputs:{PCAinfoArrayBefore:"PCAinfoArrayBefore",PCAinfoArrayAfter:"PCAinfoArrayAfter",phenotypeData:"phenotypeData"},features:[e.TTD],decls:22,vars:13,consts:[[1,"row"],["data-testid","selectAttributeForPCA",1,"form-control","form-control-sm","ml-1","form-inline",2,"width","10rem",3,"change"],[3,"value",4,"ngFor","ngForOf"],[1,"row","mt-3"],[1,"col-12","col-md-6"],[1,"text-center"],[1,"d-flex","justify-content-end","align-items-center"],[1,"align-self-center","ml-3","questionMark",3,"note","placement"],[1,"d-flex","justify-content-center","border-right"],["appD3ExportMap","",3,"pdfName","pngName","PCAinfoArray","highlightSamplesInfo"],[1,"d-flex","justify-content-center"],[3,"value"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"h5"),e._uU(2,"Select Batch Column in Phenotype Data:"),e.qZA(),e.TgZ(3,"select",1),e.NdJ("change",function(o){return s.changeSampleAnnotationType(o)}),e.TgZ(4,"option"),e._uU(5,"---"),e.qZA(),e.YNc(6,ci,2,2,"option",2),e.qZA(),e.qZA(),e.TgZ(7,"div",3),e.TgZ(8,"div",4),e.TgZ(9,"h5",5),e._uU(10,"Before removing batch effects"),e.qZA(),e.TgZ(11,"div",6),e._UZ(12,"app-display-notes-with-question-mark",7),e.qZA(),e.TgZ(13,"div",8),e._UZ(14,"app-samples-pca-plot",9),e.qZA(),e.qZA(),e.TgZ(15,"div",4),e.TgZ(16,"h5",5),e._uU(17,"After removing batch effects"),e.qZA(),e.TgZ(18,"div",6),e._UZ(19,"app-display-notes-with-question-mark",7),e.qZA(),e.TgZ(20,"div",10),e._UZ(21,"app-samples-pca-plot",9),e.qZA(),e.qZA(),e.qZA()),2&t&&(e.xp6(6),e.Q6J("ngForOf",s.annotationNames),e.xp6(6),e.Q6J("note","The horizontal and vertical coordinates represent PC1 and PC2 respectively. Each dot represents one sample. Samples from different batches are marked with different colors")("placement","bottom"),e.xp6(2),e.Q6J("pdfName","before removing batch effects.pdf")("pngName","before removing batch effects.png")("PCAinfoArray",s.PCAinfoArrayBefore)("highlightSamplesInfo",s.highlightSamplesInfo),e.xp6(5),e.Q6J("note","The horizontal and vertical coordinates represent PC1 and PC2 respectively. Each dot represents one sample. Samples from different batches are marked with different colors")("placement","auto"),e.xp6(2),e.Q6J("pdfName","after removing batch effects.pdf")("pngName","after removing batch effects.png")("PCAinfoArray",s.PCAinfoArrayAfter)("highlightSamplesInfo",s.highlightSamplesInfo))},directives:[d.YN,d.Kr,g.sg,U.m,Ye,We.I],styles:[""],changeDetection:0}),i})();const ui=["displayPCAbefore"],mi=["displayPCAafter"],hi=["submitButton"];function di(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 2. Import Counts Data and TPM/RPKM/FPKM Data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-import-rna-seq-counts-tpms",16),e.NdJ("matchedIDsPercentageRequest$",function(n){return e.CHM(t),e.oxw().processMatchedIDsPercentageRequest(n)})("countsTpmsInfo$",function(n){return e.CHM(t),e.oxw().processCountsTpmsInfo(n)}),e.ALo(6,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","Raw counts data and normalized TPM/RPKM/FPKM data are required. Raw counts data is used in differential analysis and normalized data is used in clustering analysis"),e.xp6(1),e.Q6J("availableIDtypes",t.availableRNAseqIDtypes)("matchedIDsPercentageResult",e.lcZ(6,4,t.matchedIDsPercentageResult$))}}function fi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",10),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 3. Import Phenotype data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-import-phenotype-data",17),e.NdJ("skipSignal$",function(n){return e.CHM(t),e.oxw().skipImportPhenotypeDataStep(n)})("phenotypeData$",function(n){return e.CHM(t),e.oxw().processPhenotypeData(n)}),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Phenotype Data contains the annotations of the samples. We can use the sample annotations to define groups"),e.xp6(1),e.Q6J("formWidth",t.formWidth)}}function gi(i,a){1&i&&(e.TgZ(0,"p",18),e._uU(1," Samples from normalized data and phenotype data should be the same. Please change the phenotype data file or select another column as sample column "),e.qZA())}function _i(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",10),e.TgZ(1,"h3",3),e._uU(2," Step 4. Define Groups"),e.qZA(),e.TgZ(3,"app-define-groups",19),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.ALo(4,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(1),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("phenotypeData",e.lcZ(4,3,t.inputPhenotypeData$))("ifShowSubmitButton",!1)}}function Si(i,a){if(1&i&&(e.TgZ(0,"p"),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij(" ",t.statusValue," ")}}function vi(i,a){if(1&i&&(e.TgZ(0,"div",23),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. The app is removing batch effects"),e.qZA(),e._UZ(3,"fa-icon",24),e.qZA()),2&i){const t=e.oxw(3);e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function yi(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,Si,2,1,"p",21),e.YNc(2,vi,4,3,"div",22),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf","error"===(null==t?null:t.status)),e.xp6(1),e.Q6J("ngIf","loading"===(null==t?null:t.status))}}function Ai(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",10),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 5. Remove Batch Effects "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-remove-batch-settings",20),e.NdJ("displayNormalizedPCAsignal$",function(n){return e.CHM(t),e.oxw().displayPCAplotBeforeRemovingBatchEffect(n)})("displayPCAofValuesAfterRemovingBatchSignal$",function(n){return e.CHM(t),e.oxw().displayPCAofValuesAfterRemovingBatchEffect(n)})("removeBatchInfo$",function(n){return e.CHM(t),e.oxw().processRemoveBatchInfo(n)}),e.qZA(),e.YNc(6,yi,3,2,"div",21),e.ALo(7,"async"),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Remove batch effects caused by laboratory conditions, choice of reagent, instruments used etc"),e.xp6(1),e.Q6J("phenotypeDataAttributes",t.phenotypeDataAttributes)("ifBatchEffectHasBeenRemoved",t.ifBatchEffectHasBeenRemoved),e.xp6(1),e.Q6J("ngIf",e.lcZ(7,5,t.removeBatchStatus$))}}function bi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",10),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 6. Input Dataset name "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-click-focus-input",25),e.NdJ("inputString$",function(n){return e.CHM(t),e.oxw().processInputDatasetName(n)}),e.qZA(),e.qZA()}2&i&&(e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","We can import multiple datasets. To distinguish them, each should has a unique name. We can select one in later analsyis"),e.xp6(1),e.Q6J("placeholder","Dataset Name"))}function xi(i,a){1&i&&e._UZ(0,"app-express-data-success-import-note",26)}function Di(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pca-plot-before-removing-batch",31),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",33),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("phenotypeData",e.lcZ(6,2,t.inputPhenotypeData$))("PCAinfoArray",t.beforeRemovingBatchPCAresult)}}function Ti(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pcas-before-and-after-removing-batch",34),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",35),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("PCAinfoArrayBefore",t.beforeRemovingBatchPCAresult)("PCAinfoArrayAfter",t.afterRemovingBatchPCAresult)("phenotypeData",e.lcZ(6,3,t.inputPhenotypeData$))}}let Ci=(()=>{class i{constructor(t,s,n){this.facade=t,this.modalService=s,this.cdr=n,this.faSpinner=$.LM3,this.formWidth="15rem",this.availableSpecies=V.it,this.availableRNAseqIDtypes=V.v$,this.samplesFilesInfo=[{name:"RNA-seq-sample-data1.zip",description:"RNA-seq sample data1"}],this.downloadSampleDataPercentage$=this.facade.downloadSampleDataPercentage$,this.steps=[{stepName:"defineSpecies",upStep:"",nextStep:"importCountsTpms",resetChildren:!1},{stepName:"importCountsTpms",upStep:"defineSpecies",nextStep:"importPhenotypeData",resetChildren:!1},{stepName:"importPhenotypeData",upStep:"renameSamples",nextStep:"defineGroups",resetChildren:!1},{stepName:"defineGroups",upStep:"importPhenotypeData",nextStep:"inputDatasetName",resetChildren:!1},{stepName:"removeBatch",upStep:"defineGroups",nextStep:"inputDatasetName",resetChildren:!1},{stepName:"inputDatasetName",upStep:"removeBatch",nextStep:"clickSubmitButton",resetChildren:!1},{stepName:"clickSubmitButton",upStep:"inputDatasetName",nextStep:"",resetChildren:!1}],this.ifShowSteps={defineSpecies:!0,importCountsTpms:!1,importPhenotypeData:!1,defineGroups:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.selectedSpecies$=new c.X(null),this.countsTpmsInfo$=new c.X(null),this.inputPhenotypeData$=new c.X(null),this.datasetName$=new c.X(null),this.groups$=new c.X(null),this.removeBatchSetting$=new c.X(null),this.matchedIDsPercentageRequestSubject=new c.X(null),this.ifBeforeRemoveBatchPCAexist=!1,this.requestBeforeRemovingBatchPCAinfoSubject=new c.X(null),this.ifBatchEffectHasBeenRemoved=!1,this.ifAfterRemoveBatchPCAexist=!1,this.requestAfterRemovingBatchPCAinfoSubject=new c.X(null),this.destroyNotifier$=new B.xQ}ngOnInit(){this.ifShowSamplesNotSameNote$=(0,b.aj)([this.countsTpmsInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>!(0,v.z7)([...t.samples],[...s.samples])),(0,T.O)(!1)),this.sampleExpressionValues$=this.countsTpmsInfo$.pipe((0,r.h)(t=>null!=t),(0,f.U)(t=>({samples:t.samples,valuesMatrix:t.counts}))),this.matchedIDsPercentageResult$=this.facade.matchedIDsPercentageResult$,this.facade.beforeRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.importDataUniqueID)).subscribe(t=>{this.beforeRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifBeforeRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.RNAseqRemoveBatchResult$=this.facade.removeBatchResult$.pipe((0,r.h)(t=>null===t||(null==t?void 0:t.removeBatchID)===this.importDataUniqueID)),this.removeBatchStatus$=this.facade.removeBatchStatus$,this.RNAseqRemoveBatchResult$.pipe((0,r.h)(t=>null!=t)).subscribe(t=>{this.ifBatchEffectHasBeenRemoved=!0,this.cdr.detectChanges()}),this.facade.afterRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.importDataUniqueID)).subscribe(t=>{this.afterRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifAfterRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.ifSubmitAvailable$=(0,b.aj)([this.selectedSpecies$,this.countsTpmsInfo$,this.inputPhenotypeData$,this.removeBatchSetting$,this.groups$,this.datasetName$]).pipe((0,f.U)(([t,s,n,o,l,u])=>!!(t&&s&&n&&o&&l&&u)),(0,T.O)(!1)),this.ifShowUploadNote$=this.facade.theLatestAddedDatasetID$.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.datasetName$.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>t===s),(0,_.R)(this.destroyNotifier$),(0,T.O)(!1))}ngAfterViewInit(){(0,b.aj)([this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t)),this.selectedSpecies$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>Object.assign(Object.assign({},t),{speciesID:s.taxID}))).subscribe(t=>{this.facade.requestMatchedIDsPercentage(t)}),this.requestBeforeRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.countsTpmsInfo$),(0,f.U)(([t,s])=>{const n=G();return this.importDataUniqueID=n,{pcaRequestID:this.importDataUniqueID,values:s.counts,samples:s.samples}})).subscribe(t=>{this.beforeRemovingBatchPCAresult=[],this.ifBeforeRemoveBatchPCAexist=!1,this.cdr.detectChanges(),this.facade.requestBeforeRemovingBatchPCAInfo(t)}),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>!0===(null==t?void 0:t.ifRemoveBatch)),(0,y.M)(this.countsTpmsInfo$,this.inputPhenotypeData$,this.groups$),(0,f.U)(([t,s,n,o])=>{this.importDataUniqueID||(this.importDataUniqueID=G());const u=k(s.samples,n)[t.batchAttribute];if(t.ifUseGroupVariable){const h=me(o,s.samples);return{removeBatchID:this.importDataUniqueID,values:s.counts,batch:u,group:h,valueType:"counts"}}return{removeBatchID:this.importDataUniqueID,values:s.counts,batch:u,group:null,valueType:"counts"}})).subscribe(t=>{this.facade.processRemoveBatchEffectRequest(t)}),this.requestAfterRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>t),(0,y.M)(this.countsTpmsInfo$,this.RNAseqRemoveBatchResult$),(0,f.U)(([t,s,n])=>({pcaRequestID:this.importDataUniqueID,values:n.values,samples:s.samples}))).subscribe(t=>{this.afterRemovingBatchPCAresult=[],this.ifAfterRemoveBatchPCAexist=!1,this.facade.requestAfterRemovingBatchPCAinfo(t)}),(0,b.aj)([this.selectedSpecies$,this.countsTpmsInfo$,this.inputPhenotypeData$,this.groups$,this.removeBatchSetting$,this.datasetName$]).subscribe(([t,s,n,o,l,u])=>{t&&this._displayNextStepOfCurrentStep("defineSpecies"),t&&s&&this._displayNextStepOfCurrentStep("importCountsTpms"),t&&s&&n&&Object.keys(n).length>1&&(0,v.z7)([...s.samples],[...n.samples])&&this._displayNextStepOfCurrentStep("importPhenotypeData"),t&&s&&n&&o&&this._displayNextStepOfCurrentStep("defineGroups"),t&&s&&n&&o&&l&&this._displayNextStepOfCurrentStep("removeBatch")}),this.submitStream$=(0,E.R)(this.submitButton.nativeElement,"click"),this.submitStream$.pipe((0,_.R)(this.destroyNotifier$),(0,r.h)(t=>null!=t),ce(500),(0,y.M)(this.selectedSpecies$.pipe((0,r.h)(t=>null!=t)),this.countsTpmsInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t)),this.groups$.pipe((0,r.h)(t=>null!=t)),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t)),this.RNAseqRemoveBatchResult$,this.datasetName$.pipe((0,r.h)(t=>null!=t)))).subscribe(([t,s,n,o,l,u,h,p])=>{const S=k([...n.samples],o);let A;A=u.ifRemoveBatch?{geneIDs:n.geneIDs,samples:n.samples,counts:h.counts,datasetID:p,phenotypeData:S,expressionValues:n.expressionValues,removeBatchInfo:u,defaultGroups:l}:{geneIDs:n.geneIDs,samples:n.samples,counts:n.counts,datasetID:p,phenotypeData:S,expressionValues:n.expressionValues,removeBatchInfo:u,defaultGroups:l},this.facade.addDatasetInfo({expressLocalOrRemote:K.M.local,datasetID:p,description:p,speciesID:s.taxID,speciesName:s.name,IDtype:n.IDtype,ifLogTransformed:!1,experimentType:"RNA-seq of coding RNA",samples:n.samples,groups:l}),this.facade.addLocalExpressData(A),this.facade.requestDatabaseGeneID2info({speciesID:s.taxID,idType:n.IDtype})})}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSampleFileNameToBeDownloaded(t){this.facade.downloadSampleFile(t)}processSelectedSpecies(t){t&&this.selectedSpecies$.next(t)}processMatchedIDsPercentageRequest(t){this.matchedIDsPercentageRequestSubject.next(t)}processCountsTpmsInfo(t){this.countsTpmsInfo$.next(t),this.inputPhenotypeData$.next({samples:t.samples})}processPhenotypeData(t){this.phenotypeDataAttributes=Object.keys(t),this.inputPhenotypeData$.next(t)}skipImportPhenotypeDataStep(t){this._displayNextStepOfCurrentStep("importPhenotypeData")}processGroups(t){this.groups$.next(t)}displayPCAplotBeforeRemovingBatchEffect(t){this.displayPCAbeforeModalRef=this.modalService.show(this.displayPCAbefore),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}displayPCAofValuesAfterRemovingBatchEffect(t){this.displayPCAafterModalRef=this.modalService.show(this.displayPCAafter,Object.assign({},{class:"modal-xl"})),this.ifAfterRemoveBatchPCAexist||this.requestAfterRemovingBatchPCAinfoSubject.next(!0),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}processRemoveBatchInfo(t){this.removeBatchSetting$.next(t)}processInputDatasetName(t){this.datasetName$.next(t)}_displayNextStepOfCurrentStep(t){let s;this.steps.forEach((o,l)=>{o.stepName===t&&(s=l+1)}),this.ifShowSteps[this.steps[s].stepName]=!0,this.ifShowSteps=Object.assign({},this.ifShowSteps),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-rna-seq"]],viewQuery:function(t,s){if(1&t&&(e.Gf(ui,5,e.Rgc),e.Gf(mi,5,e.Rgc),e.Gf(hi,5)),2&t){let n;e.iGM(n=e.CRH())&&(s.displayPCAbefore=n.first),e.iGM(n=e.CRH())&&(s.displayPCAafter=n.first),e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},decls:28,vars:21,consts:[[1,"mt-4","p-2","border"],[3,"fileInfos","downloadSampleDataPercentage","fileNameToBeDownloaded$"],[1,"d-flex"],[3,"appAddRequiredOptionalSign"],[1,"align-self-center","ml-3","questionMark",3,"note"],[3,"availableSpecies","selectedSpecies$"],["class","mt-4 p-2 border",4,"ngIf"],["class","mt-4 p-1 border",4,"ngIf"],[1,"mt-3"],["data-testid","samplesUnSampeNote","class","alert alert-warning",4,"ngIf"],[1,"mt-4","p-1","border"],["data-testid","submitRnaseq",1,"btn","btn-expressvis",3,"disabled"],["submitButton",""],["data-testid","uploadNote",4,"ngIf"],["displayPCAbefore",""],["displayPCAafter",""],[3,"availableIDtypes","matchedIDsPercentageResult","matchedIDsPercentageRequest$","countsTpmsInfo$"],["data-testid","loadPhenotype",3,"formWidth","skipSignal$","phenotypeData$"],["data-testid","samplesUnSampeNote",1,"alert","alert-warning"],[3,"phenotypeData","ifShowSubmitButton","outGroups"],["data-testid","removeBatch",3,"phenotypeDataAttributes","ifBatchEffectHasBeenRemoved","displayNormalizedPCAsignal$","displayPCAofValuesAfterRemovingBatchSignal$","removeBatchInfo$"],[4,"ngIf"],["class","form-inline",4,"ngIf"],[1,"form-inline"],[1,"hoverBigger",3,"icon","size","spin"],["data-testid","inputDatasetName",3,"placeholder","inputString$"],["data-testid","uploadNote"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[3,"phenotypeData","PCAinfoArray"],[1,"modal-footer"],["type","button","data-testid","closePCAbeforeRemoveBatch",1,"btn","btn-default",3,"click"],[3,"PCAinfoArrayBefore","PCAinfoArrayAfter","phenotypeData"],["type","button","data-testid","closePCAafterRemoveBatch",1,"btn","btn-default",3,"click"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"app-download-sample-data",1),e.NdJ("fileNameToBeDownloaded$",function(o){return s.processSampleFileNameToBeDownloaded(o)}),e.ALo(2,"async"),e.qZA(),e.qZA(),e.TgZ(3,"div",0),e.TgZ(4,"div",2),e.TgZ(5,"h3",3),e._uU(6," Step 1. Select Species "),e.qZA(),e._UZ(7,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(8,"app-select-species",5),e.NdJ("selectedSpecies$",function(o){return s.processSelectedSpecies(o)}),e.qZA(),e.qZA(),e.YNc(9,di,7,6,"div",6),e.YNc(10,fi,6,3,"div",7),e.TgZ(11,"div",8),e.YNc(12,gi,2,0,"p",9),e.ALo(13,"async"),e.qZA(),e.YNc(14,_i,5,5,"div",7),e.YNc(15,Ai,8,7,"div",7),e.YNc(16,bi,6,3,"div",7),e.TgZ(17,"div",10),e.TgZ(18,"button",11,12),e.ALo(20,"async"),e._uU(21," Upload RNA-seq Data "),e.qZA(),e.qZA(),e.YNc(22,xi,1,0,"app-express-data-success-import-note",13),e.ALo(23,"async"),e.YNc(24,Di,10,4,"ng-template",null,14,e.W1O),e.YNc(26,Ti,10,5,"ng-template",null,15,e.W1O)),2&t&&(e.xp6(1),e.Q6J("fileInfos",s.samplesFilesInfo)("downloadSampleDataPercentage",e.lcZ(2,13,s.downloadSampleDataPercentage$)),e.xp6(4),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","With the species information, the app can load annotations of the genes in the dataset"),e.xp6(1),e.Q6J("availableSpecies",s.availableSpecies),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importCountsTpms),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importPhenotypeData),e.xp6(2),e.Q6J("ngIf",e.lcZ(13,15,s.ifShowSamplesNotSameNote$)),e.xp6(2),e.Q6J("ngIf",s.ifShowSteps.defineGroups),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.removeBatch),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.inputDatasetName),e.xp6(2),e.Q6J("disabled",!e.lcZ(20,17,s.ifSubmitAvailable$)),e.xp6(4),e.Q6J("ngIf",!!e.lcZ(23,19,s.ifShowUploadNote$)))},directives:[he,de,U.m,fe.a,g.O5,as,Se,ye,Ae,P.BN,be.y,xe,De,Te],pipes:[g.Ov],styles:[".bd-content[_ngcontent-%COMP%]{min-height:100vh}"]}),i})();const Ii=["presentFirstLinesComponent"],Zi=["fileForm"];function Pi(i,a){if(1&i&&(e.TgZ(0,"div"),e.TgZ(1,"p",19),e._uU(2,"The app is parsing the file. Please wait for a moment."),e.qZA(),e._UZ(3,"fa-icon",20),e.qZA()),2&i){const t=e.oxw();e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function Ni(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",21),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentFirstLinesOfTheFile(n)}),e._uU(1," View first lines of the file\n"),e.qZA()}}function wi(i,a){if(1&i&&(e.TgZ(0,"option",22),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.Q6J("value",t.name),e.xp6(1),e.Oqu(t.description)}}function qi(i,a){if(1&i&&(e.TgZ(0,"p",23),e._uU(1),e.qZA()),2&i){const t=e.oxw();e.xp6(1),e.AsE(" Only ",100*(null==t.matchedIDsPercentageResult?null:t.matchedIDsPercentageResult.matchedPercentage),"% genes/proteins are mapped to ",t.selectedIDtype,". Please re-select ID column or ID type.\n")}}function Ri(i,a){if(1&i&&(e.TgZ(0,"div",24),e.TgZ(1,"h5"),e._uU(2,"Check if value columns and probe ID are correct:"),e.qZA(),e.TgZ(3,"p"),e._uU(4),e.ALo(5,"async"),e.qZA(),e._UZ(6,"app-display-limited-items",25),e.ALo(7,"async"),e.TgZ(8,"p"),e._uU(9),e.ALo(10,"async"),e.qZA(),e.TgZ(11,"p",19),e._uU(12,"If values columns and probe ID are correct, proceed to the next step"),e.qZA(),e.qZA()),2&i){const t=e.oxw();e.xp6(4),e.hij("There are ",e.lcZ(5,4,t.selectedSamples$).length," samples. Their corresponding value columns are: "),e.xp6(2),e.Q6J("items",e.lcZ(7,6,t.selectedSamples$))("displayNum",5),e.xp6(3),e.hij("The probe ID column is: ",e.lcZ(10,8,t.IDcolumn$),"")}}function $i(i,a){if(1&i&&(e.TgZ(0,"th"),e.TgZ(1,"p",35),e._uU(2),e.qZA(),e.qZA()),2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t," ")}}function Bi(i,a){if(1&i&&(e.TgZ(0,"td"),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.hij(" ",t," ")}}function Fi(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,Bi,2,1,"td",32),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.Q6J("ngForOf",t)}}function Oi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",26),e.TgZ(1,"h3"),e._uU(2,"The first five lines of the file"),e.qZA(),e.TgZ(3,"button",27),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentFirstLinesModalRef?null:n.presentFirstLinesModalRef.hide()}),e.TgZ(4,"span",28),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",29),e.TgZ(7,"div",30),e.TgZ(8,"table",31),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.YNc(11,$i,3,1,"th",32),e.ALo(12,"async"),e.qZA(),e.qZA(),e.TgZ(13,"tbody"),e.YNc(14,Fi,2,1,"tr",32),e.ALo(15,"async"),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(16,"div",33),e.TgZ(17,"button",34),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentFirstLinesModalRef.hide()}),e._uU(18," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",e.lcZ(12,2,t.fileHeaders$)),e.xp6(3),e.Q6J("ngForOf",e.lcZ(15,4,t.firstLines$))}}let ze=(()=>{class i extends ie.e{constructor(t,s,n){super(),this.importExpressionService=t,this.modalService=s,this.cdr=n,this.formWidth="15rem",this.expressionImported$=new e.vpe(null),this.matchedIDsPercentageRequest$=new e.vpe,this.faSpinner=$.LM3,this.logTransformedItems=V.Os,this.microarrayFileExampleSrc="/static/images/MicroarrayNormalizedValuesSampleFigure.png",this.fileHeaders$=new w.y(null),this.fileHeadersFromReset$=new c.X(null),this.firstLines$=new w.y(null),this.firstLinesFromReset$=new c.X(null),this.fileParseStatus$=new w.y(null),this.dropdownSettings={singleSelection:!1,selectAllText:"Select All as Sample Value Columns",unSelectAllText:"UnSelect All",itemsShowLimit:1,enableCheckAll:!0,allowSearchFilter:!0},this.candidateSamples$=new w.y(null),this.selectedSamples$=new c.X(null),this.selectedSamples=[],this.IDcolumn$=new c.X(null),this.expressionInfoInFile$=new c.X(null),this.selectedIDtype$=new c.X(null),this.selectedLogTransformMethod$=new c.X(null),this.matchedIDsCutoff=.5,this.matchedIDsPercentageRequestSubject=new c.X(null),this.matchedIDsPercentageResultSubject=new c.X(null)}ngOnInit(){this.fileParseStatus$=this.importExpressionService.fileParseStatus$,this.firstLines$=(0,j.T)(this.firstLinesFromReset$,this.importExpressionService.firstLines$).pipe((0,_.R)(this.destroyNotifier$)),this.fileHeaders$=(0,j.T)(this.fileHeadersFromReset$,this.importExpressionService.fileHeaders$).pipe((0,_.R)(this.destroyNotifier$))}ngAfterViewInit(){this.candidateSamples$=this.fileHeaders$.pipe((0,r.h)(t=>null!=t)),this.candidateSamples$.subscribe(t=>{(null==t?void 0:t.length)>0&&(this.selectedSamples=t.slice(1),this.selectedSamples$.next(this.selectedSamples))}),(0,b.aj)([this.IDcolumn$,this.selectedSamples$]).pipe((0,f.U)(([t,s])=>t&&(null==s?void 0:s.length)>0?this.importExpressionService.obtainValuesInfo(t,s):null),(0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.expressionInfoInFile$.next(t)}),(0,b.aj)([this.expressionInfoInFile$,this.selectedIDtype$]).pipe((0,f.U)(([t,s])=>t&&s?{requestID:G(),IDtype:s,IDs:t.IDs}:null),(0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.matchedIDsPercentageRequest$.next(t),this.matchedIDsPercentageRequestSubject.next(t)}),this.matchedIDsPercentageResultSubject.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>t.matchedPercentage>=this.matchedIDsCutoff),(0,y.M)(this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>{if(t.requestID===s.requestID)return!0}),(0,r.h)(t=>t),(0,y.M)(this.expressionInfoInFile$,this.selectedIDtype$,this.selectedLogTransformMethod$),(0,f.U)(([t,s,n,o])=>s&&n&&o?Object.assign(Object.assign({},s),{IDtype:n,logTransformedMethod:o}):null),(0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.expressionImported$.next(t)})}ngOnChanges(t){t.matchedIDsPercentageResult&&!t.matchedIDsPercentageResult.isFirstChange()&&this.matchedIDsPercentageResultSubject.next(this.matchedIDsPercentageResult),t.resetSelection&&!t.resetSelection.isFirstChange()&&this.resetSelection&&this._resetThisComponent()}processSelectedFile(t){this._resetIntermediateData(),this.importExpressionService.initLoadedFile(t)}presentFirstLinesOfTheFile(t){this.presentFirstLinesModalRef=this.modalService.show(this.presentFirstLinesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}outputIDtype(t){this.selectedIDtype$.next(t.target.value),this.selectedIDtype=t.target.value}processSelectedProbeColumn(t){this.IDcolumn$.next(t)}processLogTransformedMethod(t){this.selectedLogTransformMethod$.next(t)}updateSelectedSamples(t){this.selectedSamples$.next(this.selectedSamples)}_resetThisComponent(){this._resetSelectedFile(),this.fileHeadersFromReset$.next(null),this.firstLinesFromReset$.next(null),this.IDcolumn$.next(null),this.expressionInfoInFile$.next(null),this.expressionImported$.next(null),this.selectedSamples$.next(null),this.selectedSamples=[],this.matchedIDsPercentageResult=null,this.cdr.detectChanges()}_resetSelectedFile(){this.fileForm.nativeElement.reset()}_resetIntermediateData(){this.fileHeadersFromReset$.next(null),this.firstLinesFromReset$.next(null),this.expressionInfoInFile$.next(null),this.IDcolumn$.next(null),this.selectedSamples$.next(null),this.selectedSamples=[],this.matchedIDsPercentageResult=null,this.expressionImported$.next(null),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(_e),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-normalized-values"]],viewQuery:function(t,s){if(1&t&&(e.Gf(Ii,5,e.Rgc),e.Gf(Zi,7)),2&t){let n;e.iGM(n=e.CRH())&&(s.presentFirstLinesComponent=n.first),e.iGM(n=e.CRH())&&(s.fileForm=n.first)}},inputs:{formWidth:"formWidth",availableIDtypes:"availableIDtypes",fileType:"fileType",IDtypeLabel:"IDtypeLabel",IDcolumnLabel:"IDcolumnLabel",matchedIDsPercentageResult:"matchedIDsPercentageResult",resetSelection:"resetSelection"},outputs:{expressionImported$:"expressionImported$",matchedIDsPercentageRequest$:"matchedIDsPercentageRequest$"},features:[e._Bn([_e]),e.qOj,e.TTD],decls:36,vars:38,consts:[["data-testid","fileType"],[1,"d-flex"],["role","form"],["fileForm",""],["type","file",1,"form-control","form-control-sm",3,"change"],[1,"m-2",3,"figureSrc","figureLabel"],[4,"ngIf"],["data-testid","firstLinesIcon","href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],["for","IDtype",1,"mt-2"],["name","IDtype",1,"form-control","form-control-sm",3,"change"],[3,"value",4,"ngFor","ngForOf"],[1,"mt-2"],["data-testid","selectIDcolumn",3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","selectedItem$"],["ngDefaultControl","",3,"placeholder","settings","data","ngModel","ngModelChange","onSelect","onSelectAll","onDeSelect"],["name","logTransformMethod",3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","selectedItem$"],[1,"align-self-start","ml-3","questionMark",3,"note"],["data-testid","IDnotMatchNote","class","mt-3 text-danger",4,"ngIf"],["class","mt-2","data-testid","showSamples",4,"ngIf"],["presentFirstLinesComponent",""],[1,"text-info"],[1,"hoverBigger",3,"icon","size","spin"],["data-testid","firstLinesIcon","href","javascript:void(0)",1,"inline",3,"click"],[3,"value"],["data-testid","IDnotMatchNote",1,"mt-3","text-danger"],["data-testid","showSamples",1,"mt-2"],[3,"items","displayNum"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"table-responsive","mr-2","ml-2","mt-2"],[1,"table","table-bordered",2,"max-height","50vh"],[4,"ngFor","ngForOf"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"],[1,"font-weight-bold"]],template:function(t,s){if(1&t&&(e.TgZ(0,"div"),e.TgZ(1,"h5",0),e._uU(2),e.qZA(),e.TgZ(3,"div",1),e.TgZ(4,"form",2,3),e.TgZ(6,"div"),e.TgZ(7,"input",4),e.NdJ("change",function(o){return s.processSelectedFile(o)}),e.qZA(),e.qZA(),e.qZA(),e._UZ(8,"app-display-a-figure-in-modal",5),e.qZA(),e.qZA(),e.YNc(9,Pi,4,3,"div",6),e.ALo(10,"async"),e.YNc(11,Ni,2,0,"a",7),e.ALo(12,"async"),e.TgZ(13,"h5",8),e._uU(14),e.qZA(),e.TgZ(15,"select",9),e.NdJ("change",function(o){return s.outputIDtype(o)}),e.TgZ(16,"option"),e._uU(17,"---"),e.qZA(),e.YNc(18,wi,2,2,"option",10),e.qZA(),e.TgZ(19,"div",11),e.TgZ(20,"app-select-item",12),e.NdJ("selectedItem$",function(o){return s.processSelectedProbeColumn(o)}),e.ALo(21,"async"),e.qZA(),e.qZA(),e.TgZ(22,"h5",11),e._uU(23,"Select Value Columns:"),e.qZA(),e.TgZ(24,"div"),e.TgZ(25,"ng-multiselect-dropdown",13),e.NdJ("ngModelChange",function(o){return s.selectedSamples=o})("onSelect",function(o){return s.updateSelectedSamples(o)})("onSelectAll",function(o){return s.updateSelectedSamples(o)})("onDeSelect",function(o){return s.updateSelectedSamples(o)}),e.ALo(26,"async"),e.qZA(),e.qZA(),e.TgZ(27,"div",11),e.TgZ(28,"div",1),e.TgZ(29,"app-select-item",14),e.NdJ("selectedItem$",function(o){return s.processLogTransformedMethod(o)}),e.qZA(),e._UZ(30,"app-display-notes-with-question-mark",15),e.qZA(),e.qZA(),e.YNc(31,qi,2,2,"p",16),e.YNc(32,Ri,13,10,"div",17),e.ALo(33,"async"),e.YNc(34,Oi,19,6,"ng-template",null,18,e.W1O)),2&t){let n;e.xp6(2),e.hij("",s.fileType,":"),e.xp6(5),e.Udp("width",s.formWidth),e.xp6(1),e.Q6J("figureSrc",s.microarrayFileExampleSrc)("figureLabel","View Example"),e.xp6(1),e.Q6J("ngIf","loading"===(null==(n=e.lcZ(10,28,s.fileParseStatus$))?null:n.status)),e.xp6(2),e.Q6J("ngIf",e.lcZ(12,30,s.firstLines$)),e.xp6(3),e.hij("",s.IDtypeLabel,":"),e.xp6(1),e.Udp("width",s.formWidth),e.xp6(3),e.Q6J("ngForOf",s.availableIDtypes),e.xp6(2),e.Q6J("inputItems",e.lcZ(21,32,s.fileHeaders$))("itemLabel",s.IDcolumnLabel)("ifOutputFirstValue",!0)("formWidth",s.formWidth),e.xp6(4),e.Udp("width",s.formWidth),e.xp6(1),e.Q6J("placeholder","Select sample columns")("settings",s.dropdownSettings)("data",e.lcZ(26,34,s.candidateSamples$))("ngModel",s.selectedSamples),e.xp6(4),e.Q6J("inputItems",s.logTransformedItems)("itemLabel","Values log Transformed Method")("ifOutputFirstValue",!0)("formWidth",s.formWidth),e.xp6(1),e.Q6J("note","Normalization values from some softwares are log transformed. For example, those from RMA algorithm of R affy package are log2 transformed"),e.xp6(1),e.Q6J("ngIf",null!=s.matchedIDsPercentageResult&&s.matchedIDsPercentageResult.matchedPercentage<.5),e.xp6(1),e.Q6J("ngIf",e.lcZ(33,36,s.selectedSamples$))}},directives:[d._Y,d.JL,d.F,ne,g.O5,d.YN,d.Kr,g.sg,te.y,C.OP,d.Fj,d.JJ,d.On,U.m,P.BN,ve.w],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();var Fe=m(7682);const Mi=["displayPCAbefore"],Ui=["displayPCAafter"],Ji=["submitbutton"];function Qi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 2. Import normalized data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-import-normalized-values",15),e.NdJ("matchedIDsPercentageRequest$",function(n){return e.CHM(t),e.oxw().processMatchedIDsPercentageRequest(n)})("expressionImported$",function(n){return e.CHM(t),e.oxw().processExpressionValues(n)}),e.ALo(6,"async"),e.ALo(7,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","The expression data should be normalized values from a normalization software, like oligo and Affy"),e.xp6(1),e.Q6J("availableIDtypes",e.lcZ(6,8,t.availableIDtypes$))("formWidth",t.formWidth)("fileType","Microarray Normalized Values File")("IDtypeLabel","Probe ID Type in the File")("IDcolumnLabel","Select Probe Column")("matchedIDsPercentageResult",e.lcZ(7,10,t.matchedIDsPercentageResult$))}}function Ei(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 3. Import Phenotype data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-import-phenotype-data",16),e.NdJ("phenotypeData$",function(n){return e.CHM(t),e.oxw().processPhenotypeData(n)})("skipSignal$",function(n){return e.CHM(t),e.oxw().skipImportPhenotypeData(n)}),e.qZA(),e.qZA()}2&i&&(e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Phenotype Data contains the annotations of the samples. We can use the sample annotations to define groups"))}function Gi(i,a){1&i&&(e.TgZ(0,"p",17),e._uU(1," Samples from normalized data and phenotype data should be the same. Please change the phenotype data file or select another column as sample column "),e.qZA())}function ji(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"h3",3),e._uU(2," Step 4. Define Groups "),e.qZA(),e.TgZ(3,"app-define-groups",18),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.ALo(4,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(1),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("phenotypeData",e.lcZ(4,3,t.inputPhenotypeData$))("ifShowSubmitButton",!1)}}function Li(i,a){if(1&i&&(e.TgZ(0,"p"),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij(" ",t.statusValue," ")}}function ki(i,a){if(1&i&&(e.TgZ(0,"div",23),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. The app is removing batch effects"),e.qZA(),e._UZ(3,"fa-icon",24),e.qZA()),2&i){const t=e.oxw(3);e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function Vi(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,Li,2,1,"p",21),e.YNc(2,ki,4,3,"div",22),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf","error"===(null==t?null:t.status)),e.xp6(1),e.Q6J("ngIf","loading"===(null==t?null:t.status))}}function Hi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",19),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 5. Remove Batch Effects "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-remove-batch-settings",20),e.NdJ("displayNormalizedPCAsignal$",function(n){return e.CHM(t),e.oxw().displayPCAplotBeforeRemovingBatchEffect(n)})("displayPCAofValuesAfterRemovingBatchSignal$",function(n){return e.CHM(t),e.oxw().displayPCAofValuesAfterRemovingBatchEffect(n)})("removeBatchInfo$",function(n){return e.CHM(t),e.oxw().processRemoveBatchInfo(n)}),e.qZA(),e.YNc(6,Vi,3,2,"div",21),e.ALo(7,"async"),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Remove batch effects caused by laboratory conditions, choice of reagent, instruments used etc"),e.xp6(1),e.Q6J("phenotypeDataAttributes",t.phenotypeDataAttributes)("ifBatchEffectHasBeenRemoved",t.ifBatchEffectHasBeenRemoved),e.xp6(1),e.Q6J("ngIf",e.lcZ(7,5,t.removeBatchStatus$))}}function Yi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 6. Input Dataset name "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"div",7),e.TgZ(6,"app-click-focus-input",25),e.NdJ("inputString$",function(n){return e.CHM(t),e.oxw().processInputDatasetName(n)}),e.qZA(),e.qZA(),e.qZA()}2&i&&(e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","We can import multiple datasets. To distinguish them, each should has a unique name. We can select one in later analsyis"),e.xp6(2),e.Q6J("placeholder","Dataset Name"))}function Wi(i,a){1&i&&e._UZ(0,"app-express-data-success-import-note",26)}function zi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pca-plot-before-removing-batch",31),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",33),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("phenotypeData",e.lcZ(6,2,t.inputPhenotypeData$))("PCAinfoArray",t.beforeRemovingBatchPCAresult)}}function Xi(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pcas-before-and-after-removing-batch",34),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",35),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("PCAinfoArrayBefore",t.beforeRemovingBatchPCAresult)("PCAinfoArrayAfter",t.afterRemovingBatchPCAresult)("phenotypeData",e.lcZ(6,3,t.inputPhenotypeData$))}}let Ki=(()=>{class i{constructor(t,s,n){this.facade=t,this.modalService=s,this.cdr=n,this.faSpinner=$.LM3,this.availableSpecies=V.it,this.availableIDtypes$=this.facade.availableMicrorrayIDtypes$,this.formWidth="15rem",this.samplesFilesInfo=[{name:"Microarray-sample-data1.zip",description:"Microarray sample data 1"}],this.downloadSampleDataPercentage$=this.facade.downloadSampleDataPercentage$,this.matchedIDsPercentageRequestSubject=new c.X(null),this.selectedSpecies$=new c.X(null),this.expressionValuesInfo$=new c.X(null),this.inputPhenotypeData$=new c.X(null),this.datasetName$=new c.X(null),this.groups$=new c.X(null),this.removeBatchSetting$=new c.X(null),this.datasetName="",this.steps=[{stepName:"defineSpecies",upStep:"",nextStep:"importNormalizedData",resetChildren:!1},{stepName:"importNormalizedData",upStep:"defineSpecies",nextStep:"importPhenotypeData",resetChildren:!1},{stepName:"importPhenotypeData",upStep:"importNormalizedData",nextStep:"defineGroups",resetChildren:!1},{stepName:"defineGroups",upStep:"importPhenotypeData",nextStep:"removeBatch",resetChildren:!1},{stepName:"removeBatch",upStep:"defineGroups",nextStep:"inputDatasetName",resetChildren:!1},{stepName:"inputDatasetName",upStep:"removeBatch",nextStep:"clickSubmitButton",resetChildren:!1},{stepName:"clickSubmitButton",upStep:"inputDatasetName",nextStep:"",resetChildren:!1}],this.ifResetChildrenSteps={defineSpecies:!1,importNormalizedData:!1,importPhenotypeData:!1,defineGroups:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.ifShowSteps={defineSpecies:!0,importNormalizedData:!1,importPhenotypeData:!1,defineGroups:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.ifSkipImportPhenotypeData=!1,this.ifBeforeRemoveBatchPCAexist=!1,this.requestBeforeRemovingBatchPCAinfoSubject=new c.X(null),this.ifBatchEffectHasBeenRemoved=!1,this.ifAfterRemoveBatchPCAexist=!1,this.requestAfterRemovingBatchPCAinfoSubject=new c.X(null),this.destroyNotifier$=new B.xQ}ngOnInit(){this.matchedIDsPercentageResult$=this.facade.matchedIDsPercentageResult$,this.ifShowSamplesNotSameNote$=(0,b.aj)([this.expressionValuesInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>!(0,v.z7)([...t.samples],[...s.samples])),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$)),this.facade.beforeRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.importDataUniqueID)).subscribe(t=>{this.beforeRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifBeforeRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.removeBatchResult$=this.facade.removeBatchResult$.pipe((0,r.h)(t=>null===t||(null==t?void 0:t.removeBatchID)===this.importDataUniqueID)),this.removeBatchStatus$=this.facade.removeBatchStatus$,this.removeBatchResult$.pipe((0,r.h)(t=>null!=t)).subscribe(t=>{this.ifBatchEffectHasBeenRemoved=!0,this.cdr.detectChanges()}),this.facade.afterRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.importDataUniqueID)).subscribe(t=>{this.afterRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifAfterRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.ifSubmitAvailable$=(0,b.aj)([this.selectedSpecies$,this.expressionValuesInfo$,this.inputPhenotypeData$,this.groups$,this.removeBatchSetting$,this.datasetName$]).pipe((0,f.U)(([t,s,n,o,l,u])=>!!(t&&s&&n&&o&&l&&u)&&!!(0,v.z7)(s.samples,n.samples)),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$))}ngAfterViewInit(){(0,b.aj)([this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t)),this.selectedSpecies$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>Object.assign(Object.assign({},t),{speciesID:s.taxID}))).subscribe(t=>{this.facade.requestMatchedIDsPercentage(t)}),this.requestBeforeRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.expressionValuesInfo$),(0,f.U)(([t,s])=>{if(!this.importDataUniqueID){const n=G();this.importDataUniqueID=n}return{pcaRequestID:this.importDataUniqueID,values:s.values,samples:s.samples}})).subscribe(t=>{this.beforeRemovingBatchPCAresult=[],this.ifBeforeRemoveBatchPCAexist=!1,this.facade.requestBeforeRemovingBatchPCAInfo(t),this.cdr.detectChanges()}),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>!0===(null==t?void 0:t.ifRemoveBatch)),(0,y.M)(this.expressionValuesInfo$,this.inputPhenotypeData$,this.groups$),(0,f.U)(([t,s,n,o])=>{const l=s.samples,h=k(l,n)[t.batchAttribute];if(t.ifUseGroupVariable){const p=me(o,l);return{removeBatchID:this.importDataUniqueID,values:s.values,batch:h,group:p,valueType:"normalized"}}return{removeBatchID:this.importDataUniqueID,values:s.values,batch:h,group:null,valueType:"normalized"}})).subscribe(t=>{this.facade.processRemoveBatchEffectRequest(t)}),this.requestAfterRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>t),(0,y.M)(this.expressionValuesInfo$,this.removeBatchResult$),(0,f.U)(([t,s,n])=>({pcaRequestID:this.importDataUniqueID,values:n.values,samples:s.samples}))).subscribe(t=>{this.afterRemovingBatchPCAresult=[],this.ifAfterRemoveBatchPCAexist=!1,this.facade.requestAfterRemovingBatchPCAinfo(t)}),this.submitStream$=(0,E.R)(this.button.nativeElement,"click"),this.submitStream$.pipe((0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$),ce(500),(0,y.M)(this.selectedSpecies$.pipe((0,r.h)(t=>null!=t)),this.expressionValuesInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t)),this.groups$.pipe((0,r.h)(t=>null!=t)),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t)),this.removeBatchResult$,this.datasetName$.pipe((0,r.h)(t=>null!=t)))).subscribe(([t,s,n,o,l,u,h,p])=>{const A=k(n.samples,o);let F;F=u.ifRemoveBatch?{geneIDs:n.IDs,samples:n.samples,expressionValues:h.values,datasetID:p,phenotypeData:A,defaultGroups:l}:{geneIDs:n.IDs,samples:n.samples,expressionValues:n.values,datasetID:p,phenotypeData:A,defaultGroups:l},this.facade.addDatasetInfo({expressLocalOrRemote:K.M.local,datasetID:p,description:p,speciesID:s.taxID,speciesName:s.name,IDtype:n.IDtype,ifLogTransformed:!0,experimentType:"Expression profiling by array",samples:n.samples,groups:l}),this.facade.addLocalExpressData(F),this.facade.requestDatabaseGeneID2info({speciesID:s.taxID,idType:n.IDtype})}),(0,b.aj)([this.selectedSpecies$,this.expressionValuesInfo$,this.inputPhenotypeData$,this.groups$,this.removeBatchSetting$,this.datasetName$]).subscribe(([t,s,n,o,l,u])=>{t&&this._displayNextStepOfCurrentStep("defineSpecies"),t&&s&&this._displayNextStepOfCurrentStep("importNormalizedData"),t&&s&&n&&Object.keys(n).length>1&&(0,v.z7)([...s.samples],[...n.samples])&&this._displayNextStepOfCurrentStep("importPhenotypeData"),t&&s&&n&&o&&this._displayNextStepOfCurrentStep("defineGroups"),t&&s&&n&&o&&l&&this._displayNextStepOfCurrentStep("removeBatch")}),this.ifShowUploadNote$=this.facade.theLatestAddedDatasetID$.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.datasetName$.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>t===s),(0,_.R)(this.destroyNotifier$),(0,T.O)(!1))}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSampleFileNameToBeDownloaded(t){this.facade.downloadSampleFile(t)}processSelectedSpecies(t){t&&(this.selectedSpecies$.next(t),this.facade.requestMicroarrayIDtypesInTheSpecies(t.taxID))}processMatchedIDsPercentageRequest(t){this.matchedIDsPercentageRequestSubject.next(t)}processExpressionValues(t){this.expressionValuesInfo$.next(t),this.inputPhenotypeData$.next({samples:t.samples})}processPhenotypeData(t){this.phenotypeDataAttributes=Object.keys(t),this.inputPhenotypeData$.next(t)}skipImportPhenotypeData(t){this._displayNextStepOfCurrentStep("importPhenotypeData")}processGroups(t){this.groups$.next(t)}displayPCAplotBeforeRemovingBatchEffect(t){this.displayPCAbeforeModalRef=this.modalService.show(this.displayPCAbefore),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}processRemoveBatchInfo(t){this.removeBatchSetting$.next(t)}displayPCAofValuesAfterRemovingBatchEffect(t){this.displayPCAafterModalRef=this.modalService.show(this.displayPCAafter,Object.assign({},{class:"modal-xl"})),this.ifAfterRemoveBatchPCAexist||this.requestAfterRemovingBatchPCAinfoSubject.next(!0),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}processInputDatasetName(t){this.datasetName$.next(t)}_displayNextStepOfCurrentStep(t){let s;this.steps.forEach((o,l)=>{o.stepName===t&&(s=l+1)}),this.ifShowSteps[this.steps[s].stepName]=!0,this.ifShowSteps=Object.assign({},this.ifShowSteps),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-microarray"]],viewQuery:function(t,s){if(1&t&&(e.Gf(Mi,5,e.Rgc),e.Gf(Ui,5,e.Rgc),e.Gf(Ji,5)),2&t){let n;e.iGM(n=e.CRH())&&(s.displayPCAbefore=n.first),e.iGM(n=e.CRH())&&(s.displayPCAafter=n.first),e.iGM(n=e.CRH())&&(s.button=n.first)}},decls:28,vars:22,consts:[[1,"mt-4","p-2","border"],[3,"fileInfos","downloadSampleDataPercentage","fileNameToBeDownloaded$"],[1,"d-flex"],[3,"appAddRequiredOptionalSign"],[1,"align-self-center","ml-3","questionMark",3,"note"],[3,"availableSpecies","ifOutputFirstValue","selectedSpecies$"],["class","mt-4 p-2 border",4,"ngIf"],[1,"mt-3"],["class","alert alert-warning",4,"ngIf"],["class","mt-4 p-1 border",4,"ngIf"],["data-testid","submitMicroarray",1,"btn","btn-expressvis",3,"disabled"],["submitbutton",""],["data-testid","uploadNote",4,"ngIf"],["displayPCAbefore",""],["displayPCAafter",""],[3,"availableIDtypes","formWidth","fileType","IDtypeLabel","IDcolumnLabel","matchedIDsPercentageResult","matchedIDsPercentageRequest$","expressionImported$"],["appTableHeaderBodySameWidth","",3,"phenotypeData$","skipSignal$"],[1,"alert","alert-warning"],[3,"phenotypeData","ifShowSubmitButton","outGroups"],[1,"mt-4","p-1","border"],["data-testid","removeBatch",3,"phenotypeDataAttributes","ifBatchEffectHasBeenRemoved","displayNormalizedPCAsignal$","displayPCAofValuesAfterRemovingBatchSignal$","removeBatchInfo$"],[4,"ngIf"],["class","form-inline",4,"ngIf"],[1,"form-inline"],[1,"hoverBigger",3,"icon","size","spin"],[3,"placeholder","inputString$"],["data-testid","uploadNote"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[3,"phenotypeData","PCAinfoArray"],[1,"modal-footer"],["type","button","data-testid","closePCAbeforeRemoveBatch",1,"btn","btn-default",3,"click"],[3,"PCAinfoArrayBefore","PCAinfoArrayAfter","phenotypeData"],["type","button","data-testid","closePCAafterRemoveBatch",1,"btn","btn-default",3,"click"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"app-download-sample-data",1),e.NdJ("fileNameToBeDownloaded$",function(o){return s.processSampleFileNameToBeDownloaded(o)}),e.ALo(2,"async"),e.qZA(),e.qZA(),e.TgZ(3,"div",0),e.TgZ(4,"div",2),e.TgZ(5,"h3",3),e._uU(6," Step 1. Select Species "),e.qZA(),e._UZ(7,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(8,"app-select-species",5),e.NdJ("selectedSpecies$",function(o){return s.processSelectedSpecies(o)}),e.qZA(),e.qZA(),e.YNc(9,Qi,8,12,"div",6),e.YNc(10,Ei,6,2,"div",6),e.TgZ(11,"div",7),e.YNc(12,Gi,2,0,"p",8),e.ALo(13,"async"),e.qZA(),e.YNc(14,ji,5,5,"div",6),e.YNc(15,Hi,8,7,"div",9),e.YNc(16,Yi,7,3,"div",6),e.TgZ(17,"div",0),e.TgZ(18,"button",10,11),e.ALo(20,"async"),e._uU(21," Upload Microarray Data "),e.qZA(),e.qZA(),e.YNc(22,Wi,1,0,"app-express-data-success-import-note",12),e.ALo(23,"async"),e.YNc(24,zi,10,4,"ng-template",null,13,e.W1O),e.YNc(26,Xi,10,5,"ng-template",null,14,e.W1O)),2&t&&(e.xp6(1),e.Q6J("fileInfos",s.samplesFilesInfo)("downloadSampleDataPercentage",e.lcZ(2,14,s.downloadSampleDataPercentage$)),e.xp6(4),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","With the species information, the app can load annotations of the genes in the dataset"),e.xp6(1),e.Q6J("availableSpecies",s.availableSpecies)("ifOutputFirstValue",!1),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importNormalizedData),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importPhenotypeData),e.xp6(2),e.Q6J("ngIf",e.lcZ(13,16,s.ifShowSamplesNotSameNote$)),e.xp6(2),e.Q6J("ngIf",s.ifShowSteps.defineGroups),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.removeBatch),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.inputDatasetName),e.xp6(2),e.Q6J("disabled",!e.lcZ(20,18,s.ifSubmitAvailable$)),e.xp6(4),e.Q6J("ngIf",!!e.lcZ(23,20,s.ifShowUploadNote$)))},directives:[he,de,U.m,fe.a,g.O5,ze,Se,Fe.n,ye,Ae,P.BN,be.y,xe,De,Te],pipes:[g.Ov],styles:[""]}),i})();var en=m(5488);const Xe=new Map([["PANDA",["Protein Group accession"]],["MaxQuant",["Majority protein IDs","Gene names","Unique peptides"]],["Proteome Discoverer",["Accession"]],["Spectronaut",["PG.ProteinGroups","PG.Genes"]]]),tn=new Map([["PANDA","Proteins.txt"],["MaxQuant","proteinGroups.txt"],["Proteome Discoverer","xxx.txt"],["Spectronaut","Report_Protein Quant.txt"]]),sn=new Map([["PANDA","/static/images/PANDAsampleFigure.png"],["MaxQuant","/static/images/MaxQuantSampleFigure.png"],["Proteome Discoverer","/static/images/PDsampleFigure.png"],["Spectronaut","/static/images/SpecSampleFigure.png"]]),Ke=function(i,a){const t=i.annotationHeaders.indexOf(a);return i.values.map(n=>n[t])},rn={Equal:function(i,a){return i===a},Unequal:function(i,a){return i!==a},Down:function(i,a){return parseFloat(i)<=parseFloat(a)}},se=function(i,a,t,s){return{columnIndex:i.indexOf(a),columnValue:t,filterType:s}},ae=function(i,a){return i.filter(s=>rn[a.filterType](s[a.columnIndex],a.columnValue))};let et=(()=>{class i extends ge.S{constructor(){super(),this.obtainProteomicsData=(t,s)=>{const n=[...this.textMatrix[0]],l=function(i,a,t){return"MaxQuant"===i?function(i,a){const t=se(a,"Only identified by site","+","Unequal"),s=se(a,"Reverse","+","Unequal"),n=se(a,"Potential contaminant","+","Unequal"),o=ae(i,t),l=ae(o,s);return ae(l,n)}(a,t):"PANDA"===i?a:"Spectronaut"===i?function(i,a){if(a.includes("PG.Qvalue")){const t=se(a,"PG.Qvalue","0.01","Down");return ae(i,t)}return i}(a,t):"Proteome Discoverer"===i?function(i,a){let t;return t=a.includes("Protein FDR Confidence: Combined")?se(a,"Protein FDR Confidence: Combined","High","Equal"):se(a,"Protein FDR Confidence","High","Equal"),ae(i,t)}(a,t):void 0}(t,this.textMatrix.slice(1),n),u=function(i,a,t,s){const n=function(i,a,t){let s=[];switch(i){case"PANDA":s=a.filter(n=>n.startsWith("Intensity"));break;case"MaxQuant":s=a.filter(n=>!!n.match(t));break;case"Proteome Discoverer":s=a.filter(n=>n.startsWith("Area")),0===s.length&&(s=a.filter(n=>n.startsWith("Abundance")));break;case"Spectronaut":s=a.filter(n=>!!n.match(/^\[\d+\]/i))}return s}(i,t,s),o=(0,v.yF)(n,t);return{intensityHeaders:n,values:(0,v.ip)(a,o).map(h=>h.map(p=>parseFloat(p)))}}(t,l,n,s),h=function(i,a,t){const s=Xe.get(i),n=(0,v.yF)(s,t);return{annotationHeaders:s,values:(0,v.ip)(a,n)}}(t,l,n);return(0,J.of)({intensityInfo:u,annotationInfo:h})}}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac}),i})();const mn=["fileForm"],hn=["presentFirstLinesComponent"];function dn(i,a){if(1&i&&(e.TgZ(0,"div"),e.TgZ(1,"p",16),e._uU(2,"The app is parsing the file. Please wait for a moment."),e.qZA(),e._UZ(3,"fa-icon",17),e.qZA()),2&i){const t=e.oxw();e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function fn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",18),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentFirstLinesOfTheFile(n)}),e._uU(1," View the first lines of the file\n"),e.qZA()}}function gn(i,a){1&i&&(e.TgZ(0,"p",19),e._uU(1," The selected file is not from the selected software. Please select another software or select another file.\n"),e.qZA())}function _n(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-select-item",20),e.NdJ("selectedItem$",function(n){return e.CHM(t),e.oxw().selectMaxquantAlgorithm(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("inputItems",t.maxquantAlgorithms)("itemLabel","Quantitative Algorithm")("ifOutputFirstValue",!0)("resetSelection",t.resetSelection)}}function Sn(i,a){if(1&i&&(e.TgZ(0,"p",21),e._uU(1),e.qZA()),2&i){const t=e.oxw();e.xp6(1),e.AsE(" Only ",100*(null==t.matchedIDsPercentageResult?null:t.matchedIDsPercentageResult.matchedPercentage),"% proteins are mapped to ",t.proteinIDtype,". Please re-select Species, ID column or ID type.\n")}}function vn(i,a){if(1&i&&(e.TgZ(0,"th"),e.TgZ(1,"p",32),e._uU(2),e.qZA(),e.qZA()),2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t," ")}}function yn(i,a){if(1&i&&(e.TgZ(0,"td",34),e._uU(1),e.ALo(2,"trimString"),e.qZA()),2&i){const t=a.$implicit;e.s9C("title",t),e.xp6(1),e.hij(" ",e.xi3(2,2,t,20)," ")}}function An(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,yn,3,5,"td",33),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.Q6J("ngForOf",t)}}function bn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",22),e.TgZ(1,"h3"),e._uU(2,"The first five lines of the file"),e.qZA(),e.TgZ(3,"button",23),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentFirstLinesModalRef?null:n.presentFirstLinesModalRef.hide()}),e.TgZ(4,"span",24),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",25),e.TgZ(7,"div",26),e.TgZ(8,"table",27),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.YNc(11,vn,3,1,"th",28),e.ALo(12,"async"),e.qZA(),e.qZA(),e.TgZ(13,"tbody",29),e.YNc(14,An,2,1,"tr",28),e.ALo(15,"async"),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(16,"div",30),e.TgZ(17,"button",31),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentFirstLinesModalRef.hide()}),e._uU(18," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",e.lcZ(12,2,t.fileHeaders$)),e.xp6(3),e.Q6J("ngForOf",e.lcZ(15,4,t.firstLines$))}}let xn=(()=>{class i{constructor(t,s,n){this.importProteomicsDataService=t,this.modalService=s,this.cdr=n,this.resetSelection=!1,this.formWidth="15rem",this.matchedIDsPercentageRequest$=new e.vpe,this.quantificationInfo$=new e.vpe,this.fileHeaders$=new w.y(null),this.fileHeadersFromReset$=new c.X(null),this.firstLines$=new w.y(null),this.firstLinesFromReset$=new c.X(null),this.fileParseStatus$=new w.y(null),this.quantitativeSoftwares=["MaxQuant","PANDA","Spectronaut","Proteome Discoverer"],this.availableProteomicsIDtypes=V.jB.map(o=>o.name),this.maxquantAlgorithms=["Intensity","iBAQ","LFQ"],this.faQuestion=$.Psp,this.faSpinner=$.LM3,this.defaultQuantitativeFileName="niho",this.quantitativeFileSrc="/static/images/FromQuantitativeSoftwareSampleFigure.png",this.fileEventSubject=new B.xQ,this.quantitativeSoftware$=new c.X(null),this.algorithm$=new c.X(null),this.infoFromProteimicsFile$=new c.X(null),this.proteinIDcolumn$=new c.X(null),this.proteinIDtype$=new c.X(null),this.matchedIDsCutoff=.5,this.matchedIDsPercentageRequestSubject=new c.X(null),this.matchedIDsPercentageResultSubject=new c.X(null),this.destroyNotifier$=new B.xQ}ngOnInit(){this.fileParseStatus$=this.importProteomicsDataService.fileParseStatus$,this.firstLines$=(0,j.T)(this.firstLinesFromReset$,this.importProteomicsDataService.firstLines$).pipe((0,_.R)(this.destroyNotifier$)),this.fileHeaders$=(0,j.T)(this.fileHeadersFromReset$,this.importProteomicsDataService.fileHeaders$).pipe((0,_.R)(this.destroyNotifier$)),this.fileEventSubject.pipe((0,r.h)(t=>null!==t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.importProteomicsDataService.initLoadedFile(t)}),this.ifSelectedSoftwareCorrect$=this.fileHeaders$.pipe((0,r.h)(t=>null!=t),(0,f.U)(t=>{let s=!1;return this.possibleIDcolumns.forEach(n=>{t.includes(n)&&(s=!0)}),s}),(0,T.O)(!0)),(0,b.aj)([this.algorithm$,this.quantitativeSoftware$,this.firstLines$]).pipe((0,R.w)(([t,s,n])=>{if(t&&s&&n){if("MaxQuant"===s&&"Not Required"!==t){const o=new RegExp(`^${t}.+`,"i");return this.importProteomicsDataService.obtainProteomicsData(s,o)}return this.importProteomicsDataService.obtainProteomicsData(s,null)}return(0,J.of)(null)}),(0,r.h)(t=>null!=t)).subscribe(t=>{this.infoFromProteimicsFile$.next(t)}),this.quantitativeSoftware$.pipe((0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this._resetSelectedFile(),this.defaultQuantitativeFileName=`The default file name from ${t} is: ${tn.get(t)}`,this.quantitativeFileSrc=sn.get(t),this.possibleIDcolumns=Xe.get(t)}),(0,b.aj)([this.quantitativeSoftware$,this.infoFromProteimicsFile$,this.proteinIDcolumn$,this.proteinIDtype$,this.ifSelectedSoftwareCorrect$.pipe((0,en.T)(1))]).pipe((0,f.U)(([t,s,n,o,l])=>{if(t&&s&&n&&o&&l){const u=Ke(s.annotationInfo,n);let h=[];return u.forEach(S=>{h=h.concat(S.split(";"))}),{requestID:G(),IDtype:o,IDs:u}}return null}),(0,r.h)(t=>null!=t)).subscribe(t=>{this.matchedIDsPercentageRequest$.next(t),this.matchedIDsPercentageRequestSubject.next(t)}),this.matchedIDsPercentageResultSubject.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>t.matchedPercentage>=this.matchedIDsCutoff),(0,y.M)(this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>t.requestID===s.requestID),(0,r.h)(t=>t),(0,y.M)(this.infoFromProteimicsFile$,this.proteinIDcolumn$,this.proteinIDtype$),(0,f.U)(([t,s,n,o])=>{const l=Ke(s.annotationInfo,n);return Object.assign(Object.assign({},s),{proteinIDtype:o,proteinIDs:l})})).subscribe(t=>{this.quantificationInfo$.next(t)})}ngOnChanges(t){t.resetSelection&&!t.resetSelection.isFirstChange()&&this.resetSelection&&this._resetThisComponent(),t.matchedIDsPercentageResult&&!t.matchedIDsPercentageResult.isFirstChange()&&this.matchedIDsPercentageResultSubject.next(this.matchedIDsPercentageResult)}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSelectedQuantitativeSoftware(t){this.quantitativeSoftware$.next(t),"MaxQuant"!==t&&this.algorithm$.next("Not Required"),this._resetThisComponent()}processProteomicsFile(t){this.fileEventSubject.next(t)}presentFirstLinesOfTheFile(t){this.presentFirstLinesModalRef=this.modalService.show(this.presentFirstLinesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}selectMaxquantAlgorithm(t){this.algorithm$.next(t)}processSelectedProteinIDColumn(t){this.proteinIDcolumn$.next(t)}processSelectedProteinIDtype(t){this.proteinIDtype=t,this.proteinIDtype$.next(t)}_resetThisComponent(){this._resetSelectedFile(),this.fileHeadersFromReset$.next(null),this.firstLinesFromReset$.next(null),this.infoFromProteimicsFile$.next(null),this.quantificationInfo$.next(null)}_resetSelectedFile(){this.fileForm.nativeElement.reset()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(et),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-proteomics-quantitative-data"]],viewQuery:function(t,s){if(1&t&&(e.Gf(mn,7),e.Gf(hn,5,e.Rgc)),2&t){let n;e.iGM(n=e.CRH())&&(s.fileForm=n.first),e.iGM(n=e.CRH())&&(s.presentFirstLinesComponent=n.first)}},inputs:{resetSelection:"resetSelection",formWidth:"formWidth",matchedIDsPercentageResult:"matchedIDsPercentageResult"},outputs:{matchedIDsPercentageRequest$:"matchedIDsPercentageRequest$",quantificationInfo$:"quantificationInfo$"},features:[e._Bn([et]),e.TTD],decls:25,vars:30,consts:[["data-testid","selectProSource",3,"inputItems","itemLabel","ifOutputFirstValue","resetSelection","selectedItem$"],[1,"mt-3"],[1,"d-flex"],["role","form"],["fileForm",""],["data-testid","quantitativeFile","type","file",1,"form-control","form-control-sm",3,"ngbTooltip","change"],[1,"ml-2",3,"figureSrc","figureLabel"],[4,"ngIf"],["href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],["data-testid","softwareErrorNote","class","mt-3 text-danger",4,"ngIf"],[1,"mt-2"],["data-testid","selectAlgorithm",3,"inputItems","itemLabel","ifOutputFirstValue","resetSelection","selectedItem$",4,"ngIf"],["data-testid","proteomicsIDcolumn",3,"inputItems","itemLabel","ifOutputFirstValue","resetSelection","selectedItem$"],["data-testid","proteomicsIDtype",3,"inputItems","itemLabel","ifOutputFirstValue","resetSelection","selectedItem$"],["data-testid","IDnotMatchNote","class","mt-3 text-danger",4,"ngIf"],["presentFirstLinesComponent",""],[1,"text-info"],[1,"hoverBigger",3,"icon","size","spin"],["href","javascript:void(0)",1,"inline",3,"click"],["data-testid","softwareErrorNote",1,"mt-3","text-danger"],["data-testid","selectAlgorithm",3,"inputItems","itemLabel","ifOutputFirstValue","resetSelection","selectedItem$"],["data-testid","IDnotMatchNote",1,"mt-3","text-danger"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"table-responsive","mr-2","ml-2"],[1,"table","table-bordered",2,"max-height","50vh"],[4,"ngFor","ngForOf"],[2,"max-height","40vh"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"],[1,"font-weight-bold"],[3,"title",4,"ngFor","ngForOf"],[3,"title"]],template:function(t,s){if(1&t&&(e.TgZ(0,"app-select-item",0),e.NdJ("selectedItem$",function(o){return s.processSelectedQuantitativeSoftware(o)}),e.qZA(),e.TgZ(1,"h5",1),e._uU(2,"Select File from the Quantitative Software:"),e.qZA(),e.TgZ(3,"div",2),e.TgZ(4,"form",3,4),e.TgZ(6,"div"),e.TgZ(7,"input",5),e.NdJ("change",function(o){return s.processProteomicsFile(o)}),e.qZA(),e.qZA(),e.qZA(),e._UZ(8,"app-display-a-figure-in-modal",6),e.qZA(),e.YNc(9,dn,4,3,"div",7),e.ALo(10,"async"),e.YNc(11,fn,2,0,"a",8),e.ALo(12,"async"),e.YNc(13,gn,2,0,"p",9),e.ALo(14,"async"),e.TgZ(15,"div",10),e.YNc(16,_n,1,4,"app-select-item",11),e.ALo(17,"async"),e.qZA(),e.TgZ(18,"div",10),e.TgZ(19,"app-select-item",12),e.NdJ("selectedItem$",function(o){return s.processSelectedProteinIDColumn(o)}),e.qZA(),e.qZA(),e.TgZ(20,"div",10),e.TgZ(21,"app-select-item",13),e.NdJ("selectedItem$",function(o){return s.processSelectedProteinIDtype(o)}),e.qZA(),e.qZA(),e.YNc(22,Sn,2,2,"p",14),e.YNc(23,bn,19,6,"ng-template",null,15,e.W1O)),2&t){let n;e.Q6J("inputItems",s.quantitativeSoftwares)("itemLabel","Quantitative Software")("ifOutputFirstValue",!0)("resetSelection",s.resetSelection),e.xp6(7),e.Udp("width",s.formWidth),e.Q6J("ngbTooltip",s.defaultQuantitativeFileName),e.xp6(1),e.Q6J("figureSrc",s.quantitativeFileSrc)("figureLabel","View Example"),e.xp6(1),e.Q6J("ngIf","loading"===(null==(n=e.lcZ(10,22,s.fileParseStatus$))?null:n.status)),e.xp6(2),e.Q6J("ngIf",e.lcZ(12,24,s.firstLines$)),e.xp6(2),e.Q6J("ngIf",!e.lcZ(14,26,s.ifSelectedSoftwareCorrect$)),e.xp6(3),e.Q6J("ngIf","MaxQuant"===e.lcZ(17,28,s.quantitativeSoftware$)),e.xp6(3),e.Q6J("inputItems",s.possibleIDcolumns)("itemLabel","Protein ID Column")("ifOutputFirstValue",!0)("resetSelection",s.resetSelection),e.xp6(2),e.Q6J("inputItems",s.availableProteomicsIDtypes)("itemLabel","Protein ID Type")("ifOutputFirstValue",!1)("resetSelection",s.resetSelection),e.xp6(1),e.Q6J("ngIf",null!=s.matchedIDsPercentageResult&&s.matchedIDsPercentageResult.matchedPercentage<.5)}},directives:[te.y,d._Y,d.JL,d.F,x._L,ne,g.O5,P.BN,g.sg],pipes:[g.Ov,Pe.K],styles:[""],changeDetection:0}),i})();var Dn=m(1954);function Tn(i,a){if(1&i&&e._uU(0),2&i){const t=e.oxw().$implicit;e.hij(" ",t.controls.newName.value," ")}}function Cn(i,a){1&i&&(e.TgZ(0,"p",11),e._uU(1," Sample name should not be empty "),e.qZA())}function In(i,a){if(1&i&&(e._UZ(0,"input",9),e.YNc(1,Cn,2,0,"p",10)),2&i){const t=e.oxw().index,s=e.oxw(2);e.Q6J("formControl",s.getControl(t,"newName")),e.xp6(1),e.Q6J("ngIf",null==s.getControl(t,"newName").errors?null:s.getControl(t,"newName").errors.required)}}function Zn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"tr"),e.TgZ(1,"td"),e._uU(2),e.qZA(),e.TgZ(3,"td"),e.TgZ(4,"app-editable",6),e.NdJ("update",function(){const o=e.CHM(t).index;return e.oxw(2).updateSampleName(o,"newName")}),e.YNc(5,Tn,1,1,"ng-template",7),e.YNc(6,In,2,2,"ng-template",8),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t.controls.rawName.value," ")}}function Pn(i,a){if(1&i&&(e.TgZ(0,"table",3),e.TgZ(1,"thead"),e.TgZ(2,"tr"),e.TgZ(3,"th"),e._uU(4,"Raw Sample Name"),e.qZA(),e.TgZ(5,"th"),e._uU(6,"New Sample Name"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(7,"tbody",4),e.YNc(8,Zn,7,1,"tr",5),e.qZA(),e.qZA()),2&i){const t=e.oxw();e.xp6(8),e.Q6J("ngForOf",t.sampleNameFormArray.controls)}}let Nn=(()=>{class i{constructor(){this.resetToDefault=!1,this.rawNames2newNames$=new e.vpe}ngOnInit(){if(this.rawSampleNames){const t=this.rawSampleNames.map(s=>new d.cw({rawName:new d.NI(s,[d.kI.required]),newName:new d.NI(s,[d.kI.required])}));this.sampleNameFormArray=new d.Oe(t)}}ngOnChanges(t){if(t.rawSampleNames&&!t.rawSampleNames.isFirstChange()){const s=this.rawSampleNames.map(n=>new d.cw({rawName:new d.NI(n,[d.kI.required]),newName:new d.NI(n,[d.kI.required])}));this.sampleNameFormArray=new d.Oe(s)}t.resetToDefault&&!t.resetToDefault.isFirstChange()&&this.rawNames2newNames$.next(null)}getControl(t,s){return this.sampleNameFormArray.at(t).get(s)}updateSampleName(t,s){const n=new Map;this.sampleNameFormArray.value.forEach(o=>{n.set(o.rawName,o.newName)}),this.rawNames2newNames$.next(n)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-rename-samples-manually"]],inputs:{rawSampleNames:"rawSampleNames",resetToDefault:"resetToDefault"},outputs:{rawNames2newNames$:"rawNames2newNames$"},features:[e.TTD],decls:4,vars:1,consts:[[1,"table-responsive",2,"max-height","40vh"],[1,"text-info"],["class","table",4,"ngIf"],[1,"table"],[2,"max-height","500px"],[4,"ngFor","ngForOf"],["initialMode","view",3,"update"],["appEditableView",""],["appEditableEdit",""],["appEditableOnEnter","","appFocusable","","ngDefaultControl","","required","",3,"formControl"],["style","color: red",4,"ngIf"],[2,"color","red"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"p",1),e._uU(2,"Double click to input new sample name"),e.qZA(),e.YNc(3,Pn,9,1,"table",2),e.qZA()),2&t&&(e.xp6(3),e.Q6J("ngIf",s.sampleNameFormArray))},directives:[g.O5,g.sg,Ne.y,we.i,qe.O,d.Fj,Re._,$e.x,d.Q7,d.JJ,d.oH],styles:[""],changeDetection:0}),i})(),tt=(()=>{class i extends ge.S{constructor(){super()}obtainRawtoNew(t,s){const n=[...this.textMatrix[0]].map(h=>h.toString()),o=n.indexOf(t.toString().trim()),l=n.indexOf(s.toString().trim());let u=new Map;return this.textMatrix.slice(1).forEach(h=>{u.set(h[o],h[l])}),(0,J.of)(u)}determineIfSamplesAreEqual(t){const s=[...this.textLines].slice(1).map(n=>n.split("\t")[0]);return(0,J.of)((0,v.z7)(t,s))}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275prov=e.Yz7({token:i,factory:i.\u0275fac}),i})();var wn=m(5346);const qn=["fileForm"],Rn=["presentFirstLinesComponent"],$n=["presentOldAndNewNamesComponent"];function Bn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",15),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentFirstLinesOfTheFile(n)}),e._uU(1," View the first lines of the file\n"),e.qZA()}}function Fn(i,a){1&i&&(e.TgZ(0,"p",16),e._uU(1,"Samples in the mapping file are not the same as those in the quantification file"),e.qZA())}function On(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"a",17),e.NdJ("click",function(n){return e.CHM(t),e.oxw().presentOldNamesAndNewNames(n)}),e._uU(1," View Old Names and New Names "),e.qZA()}}function Mn(i,a){if(1&i&&(e.TgZ(0,"th"),e.TgZ(1,"p",27),e._uU(2),e.qZA(),e.qZA()),2&i){const t=a.$implicit;e.xp6(2),e.hij(" ",t," ")}}function Un(i,a){if(1&i&&(e.TgZ(0,"td"),e._uU(1),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.hij(" ",t," ")}}function Jn(i,a){if(1&i&&(e.TgZ(0,"tr"),e.YNc(1,Un,2,1,"td",24),e.qZA()),2&i){const t=a.$implicit;e.xp6(1),e.Q6J("ngForOf",t)}}function Qn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",18),e.TgZ(1,"h3"),e._uU(2,"The first five lines of the file"),e.qZA(),e.TgZ(3,"button",19),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentFirstLinesModalRef?null:n.presentFirstLinesModalRef.hide()}),e.TgZ(4,"span",20),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",21),e.TgZ(7,"div",22),e.TgZ(8,"table",23),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.YNc(11,Mn,3,1,"th",24),e.ALo(12,"async"),e.qZA(),e.qZA(),e.TgZ(13,"tbody"),e.YNc(14,Jn,2,1,"tr",24),e.ALo(15,"async"),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(16,"div",25),e.TgZ(17,"button",26),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentFirstLinesModalRef.hide()}),e._uU(18," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(11),e.Q6J("ngForOf",e.lcZ(12,2,t.fileHeaders$)),e.xp6(3),e.Q6J("ngForOf",e.lcZ(15,4,t.firstLines$))}}function En(i,a){if(1&i&&(e.TgZ(0,"tr"),e.TgZ(1,"td"),e._uU(2),e.qZA(),e.TgZ(3,"td"),e._uU(4),e.qZA(),e.qZA()),2&i){const t=a.$implicit,s=e.oxw(2);e.xp6(2),e.hij(" ",t," "),e.xp6(2),e.hij(" ",s.old2new.get(t)," ")}}function Gn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",18),e.TgZ(1,"h3"),e._uU(2,"Old names and new names"),e.qZA(),e.TgZ(3,"button",19),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.presentOldandNewNamesModalRef?null:n.presentOldandNewNamesModalRef.hide()}),e.TgZ(4,"span",20),e._uU(5,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(6,"div",21),e.TgZ(7,"div",22),e.TgZ(8,"table",23),e.TgZ(9,"thead"),e.TgZ(10,"tr"),e.TgZ(11,"th"),e._uU(12,"Old Name"),e.qZA(),e.TgZ(13,"th"),e._uU(14,"New Name"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(15,"tbody"),e.YNc(16,En,5,2,"tr",24),e.qZA(),e.qZA(),e.qZA(),e.qZA(),e.TgZ(17,"div",25),e.TgZ(18,"button",26),e.NdJ("click",function(){return e.CHM(t),e.oxw().presentOldandNewNamesModalRef.hide()}),e._uU(19," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(16),e.Q6J("ngForOf",t.old2new.keys())}}let jn=(()=>{class i{constructor(t,s,n){this.mappingFileService=t,this.modalService=s,this.cdr=n,this.rawSampleNames=[],this.resetToDefault=!1,this.formWidth="15rem",this.rawNames2newNames$=new e.vpe,this.renameFileExampleSrc="/static/images/SamplesRenameSampleFigure.png",this.fileHeaders$=new w.y(null),this.fileHeadersFromReset$=new c.X(null),this.firstLines$=new w.y(null),this.firstLinesFromReset$=new c.X(null),this.oldNameColumn$=new c.X(null),this.newNameColumn$=new c.X(null),this.destroyNotifier$=new B.xQ}ngOnInit(){this.firstLines$=(0,j.T)(this.firstLinesFromReset$,this.mappingFileService.firstLines$).pipe((0,_.R)(this.destroyNotifier$)),this.fileHeaders$=(0,j.T)(this.fileHeadersFromReset$,this.mappingFileService.fileHeaders$).pipe((0,_.R)(this.destroyNotifier$)),this.ifShowSamplesNote$=this.firstLines$.pipe((0,r.h)(t=>null!=t),(0,R.w)(t=>this.mappingFileService.determineIfSamplesAreEqual(this.rawSampleNames)),(0,f.U)(t=>!t),(0,T.O)(!1)),(0,b.aj)([this.ifShowSamplesNote$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>!t)),this.oldNameColumn$.pipe((0,r.h)(t=>null!=t)),this.newNameColumn$.pipe((0,r.h)(t=>null!=t))]).pipe((0,R.w)(([t,s,n])=>this.mappingFileService.obtainRawtoNew(s,n))).subscribe(t=>{this.old2new=t,this.rawNames2newNames$.next(t)})}ngOnChanges(t){t.resetToDefault&&!t.resetToDefault.isFirstChange()&&this.resetToDefault&&this._resetThisComponent()}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSelectedFile(t){this.mappingFileService.initLoadedFile(t)}processSelectedOldNameColumn(t){this.oldNameColumn$.next(t)}processSelectedNewNameColumn(t){this.newNameColumn$.next(t)}presentFirstLinesOfTheFile(t){this.presentFirstLinesModalRef=this.modalService.show(this.presentFirstLinesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}presentOldNamesAndNewNames(t){this.presentOldandNewNamesModalRef=this.modalService.show(this.presentOldAndNewNamesComponent,Object.assign({},{class:"modal-xl"})),this.cdr.detectChanges()}_resetThisComponent(){this._resetSelectedFile(),this.fileHeadersFromReset$.next(null),this.firstLinesFromReset$.next(null),this.rawNames2newNames$.next(null)}_resetSelectedFile(){this.fileForm.nativeElement.reset()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(tt),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-rename-samples-with-mapping-file"]],viewQuery:function(t,s){if(1&t&&(e.Gf(qn,7),e.Gf(Rn,5,e.Rgc),e.Gf($n,5,e.Rgc)),2&t){let n;e.iGM(n=e.CRH())&&(s.fileForm=n.first),e.iGM(n=e.CRH())&&(s.presentFirstLinesComponent=n.first),e.iGM(n=e.CRH())&&(s.presentOldAndNewNamesComponent=n.first)}},inputs:{rawSampleNames:"rawSampleNames",resetToDefault:"resetToDefault",formWidth:"formWidth"},outputs:{rawNames2newNames$:"rawNames2newNames$"},features:[e._Bn([tt]),e.TTD],decls:28,vars:30,consts:[["role","form"],["fileForm",""],[1,"d-flex"],["data-testid","selectRenameFile","type","file",1,"form-control","form-control-sm",3,"ngbTooltip","change"],[1,"ml-2",3,"figureSrc","figureLabel"],["href","javascript:void(0)","class","inline",3,"click",4,"ngIf"],[1,"mt-2","d-flex"],["data-testid","oldNameColumn",3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","selectedItem$"],[1,"align-self-start","ml-3","questionMark",3,"note"],["data-testid","newNameColumn",3,"inputItems","itemLabel","ifOutputFirstValue","formWidth","selectedItem$"],["data-testid","notEqualNote","class","text-danger",4,"ngIf"],[1,"mt-2"],["href","javascript:void(0)",3,"click",4,"ngIf"],["presentFirstLinesComponent",""],["presentOldAndNewNamesComponent",""],["href","javascript:void(0)",1,"inline",3,"click"],["data-testid","notEqualNote",1,"text-danger"],["href","javascript:void(0)",3,"click"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[1,"table-responsive","mr-2","ml-2"],[1,"table","table-bordered","mt-2",2,"max-height","50vh"],[4,"ngFor","ngForOf"],[1,"modal-footer"],["data-testid","closeGeneListInput","type","button",1,"btn","btn-default","btn-sm",3,"click"],[1,"font-weight-bold"]],template:function(t,s){1&t&&(e.TgZ(0,"form",0,1),e.TgZ(2,"div"),e.TgZ(3,"h5"),e._uU(4,"Samples Rename Mapping File:"),e.qZA(),e.TgZ(5,"div",2),e.TgZ(6,"input",3),e.NdJ("change",function(o){return s.processSelectedFile(o)}),e.qZA(),e._UZ(7,"app-display-a-figure-in-modal",4),e.qZA(),e.qZA(),e.qZA(),e.YNc(8,Bn,2,0,"a",5),e.ALo(9,"async"),e.TgZ(10,"div",6),e.TgZ(11,"app-select-item",7),e.NdJ("selectedItem$",function(o){return s.processSelectedOldNameColumn(o)}),e.ALo(12,"async"),e.qZA(),e._UZ(13,"app-display-notes-with-question-mark",8),e.qZA(),e.TgZ(14,"div",6),e.TgZ(15,"app-select-item",9),e.NdJ("selectedItem$",function(o){return s.processSelectedNewNameColumn(o)}),e.ALo(16,"reverse"),e.ALo(17,"async"),e.qZA(),e._UZ(18,"app-display-notes-with-question-mark",8),e.qZA(),e.YNc(19,Fn,2,0,"p",10),e.ALo(20,"async"),e.TgZ(21,"div",11),e.YNc(22,On,2,0,"a",12),e.ALo(23,"async"),e.qZA(),e.YNc(24,Qn,19,6,"ng-template",null,13,e.W1O),e.YNc(26,Gn,20,1,"ng-template",null,14,e.W1O)),2&t&&(e.xp6(6),e.Udp("width",s.formWidth),e.Q6J("ngbTooltip","The file should contains one column for old names, another one for new names"),e.xp6(1),e.Q6J("figureSrc",s.renameFileExampleSrc)("figureLabel","View Example"),e.xp6(1),e.Q6J("ngIf",e.lcZ(9,18,s.firstLines$)),e.xp6(3),e.Q6J("inputItems",e.lcZ(12,20,s.fileHeaders$))("itemLabel","Select Old Name Column")("ifOutputFirstValue",!0)("formWidth",s.formWidth),e.xp6(2),e.Q6J("note","The default column for old names is the first column"),e.xp6(2),e.Q6J("inputItems",e.lcZ(16,22,e.lcZ(17,24,s.fileHeaders$)))("itemLabel","Select New Name Column")("ifOutputFirstValue",!0)("formWidth",s.formWidth),e.xp6(3),e.Q6J("note","The default column for new names is the last column"),e.xp6(1),e.Q6J("ngIf",e.lcZ(20,26,s.ifShowSamplesNote$)),e.xp6(3),e.Q6J("ngIf",!e.lcZ(23,28,s.ifShowSamplesNote$)))},directives:[d._Y,d.JL,d.F,x._L,ne,g.O5,te.y,U.m,g.sg],pipes:[g.Ov,wn.B],styles:[""],changeDetection:0}),i})();function Ln(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-rename-samples-manually",8),e.NdJ("rawNames2newNames$",function(n){return e.CHM(t),e.oxw().processRawNames2newNames(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("rawSampleNames",t.rawSampleNames)("resetToDefault",t.resetToDefault)}}function kn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-rename-samples-with-mapping-file",9),e.NdJ("rawNames2newNames$",function(n){return e.CHM(t),e.oxw().processRawNames2newNames(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("rawSampleNames",t.rawSampleNames)("resetToDefault",t.resetToDefault)("formWidth",t.formWidth)}}let Vn=(()=>{class i{constructor(){this.resetToDefault=!1,this.formWidth="15rem",this.skipSignal$=new e.vpe,this.rawNames2newNames$=new e.vpe,this.methodsForRenameSamples=["Rename Manually","Rename Using Mapping File"]}ngOnInit(){}ngOnChanges(t){t.resetToDefault&&t.resetToDefault.isFirstChange()}skipThisStep(t){this.skipSignal$.next(!0)}processSelectedRenameMethod(t){this.renameMethod=t}processRawNames2newNames(t){this.rawNames2newNames$.next(t)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-rename-sample-names"]],inputs:{rawSampleNames:"rawSampleNames",resetToDefault:"resetToDefault",formWidth:"formWidth"},outputs:{skipSignal$:"skipSignal$",rawNames2newNames$:"rawNames2newNames$"},features:[e.TTD],decls:12,vars:8,consts:[[1,"d-flex"],[1,"align-self-center","ml-3","questionMark",3,"note"],["data-testid","renameMethod",3,"inputItems","itemLabel","formInlineOrNot","ifOutputFirstValue","resetToDefault","selectedItem$"],["appTableHeaderBodySameWidth","",3,"rawSampleNames","resetToDefault","rawNames2newNames$",4,"ngIf"],["appTableHeaderBodySameWidth","",3,"rawSampleNames","resetToDefault","formWidth","rawNames2newNames$",4,"ngIf"],[1,"mt-2"],[1,"text-muted"],["type","button","data-testid","skipRenameSamples",1,"btn","btn-expressvis","btn-sm",3,"click"],["appTableHeaderBodySameWidth","",3,"rawSampleNames","resetToDefault","rawNames2newNames$"],["appTableHeaderBodySameWidth","",3,"rawSampleNames","resetToDefault","formWidth","rawNames2newNames$"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"h5"),e._uU(2,"Select Rename Method"),e.qZA(),e._UZ(3,"app-display-notes-with-question-mark",1),e.qZA(),e.TgZ(4,"app-select-between-two-using-radio",2),e.NdJ("selectedItem$",function(o){return s.processSelectedRenameMethod(o)}),e.qZA(),e.YNc(5,Ln,1,2,"app-rename-samples-manually",3),e.YNc(6,kn,1,3,"app-rename-samples-with-mapping-file",4),e.TgZ(7,"div",5),e.TgZ(8,"p",6),e._uU(9,"This step is optional. Click skip button to proceed to the next step"),e.qZA(),e.TgZ(10,"button",7),e.NdJ("click",function(o){return s.skipThisStep(o)}),e._uU(11,"Skip"),e.qZA(),e.qZA()),2&t&&(e.xp6(3),e.Q6J("note","With Rename Manually, we can edit the new names manually. With Reanme Using Mapping File, we should load a mapping file contains old names and new names"),e.xp6(1),e.Q6J("inputItems",s.methodsForRenameSamples)("itemLabel","Rename Method")("formInlineOrNot",!1)("ifOutputFirstValue",!1)("resetToDefault",s.resetToDefault),e.xp6(1),e.Q6J("ngIf","Rename Manually"===s.renameMethod),e.xp6(1),e.Q6J("ngIf","Rename Using Mapping File"===s.renameMethod))},directives:[U.m,Dn.W,g.O5,Nn,Fe.n,jn],styles:[""],changeDetection:0}),i})();var st=m(4816),Oe=m(438);let Hn=(()=>{class i{constructor(){this.formWidth="15rem",this.filterSetting$=new e.vpe,this.percentageSubject=new c.X(null),this.filterTypeSubject=new c.X(null),this.NAsAcceptanceTypes=[{name:"all",description:"Filter NA Based on All Samples"},{name:"alo",description:"NAs of at least 1 Groups Meet Requirement"}],this.destroyNotifier$=new B.xQ}ngOnInit(){(0,b.aj)([this.percentageSubject.pipe((0,r.h)(t=>null!=t)),this.filterTypeSubject.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>({percentage:t,filterType:s})),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.filterSetting$.next(t)})}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSelectedAcceptanceType(t){this.filterTypeSubject.next(t.name)}processNAacceptancePercentage(t){this.percentageSubject.next(t)}}return i.\u0275fac=function(t){return new(t||i)},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-proteomics-preprocess-filter-settings"]],inputs:{formWidth:"formWidth"},outputs:{filterSetting$:"filterSetting$"},decls:2,vars:14,consts:[[3,"minValue","maxValue","defaultValue","formInlineOrNot","step","itemLabel","outputNumber$"],[3,"inputObjects","idAttribute","displayAttribute","itemLabel","formInlineOrNot","ifOutputFirstValue","emptyNote","formWidth","selectedItemObject$"]],template:function(t,s){1&t&&(e.TgZ(0,"app-input-number-with-range",0),e.NdJ("outputNumber$",function(o){return s.processNAacceptancePercentage(o)}),e.qZA(),e.TgZ(1,"app-select-item-object",1),e.NdJ("selectedItemObject$",function(o){return s.processSelectedAcceptanceType(o)}),e.qZA()),2&t&&(e.Q6J("minValue",.05)("maxValue",.95)("defaultValue",.3)("formInlineOrNot",!1)("step",.01)("itemLabel","Max Acceptable Percentage of NAs"),e.xp6(1),e.Q6J("inputObjects",s.NAsAcceptanceTypes)("idAttribute","name")("displayAttribute","description")("itemLabel","NAs Acceptance Type")("formInlineOrNot",!1)("ifOutputFirstValue",!0)("emptyNote","no types is available")("formWidth",s.formWidth))},directives:[st.C,Oe.w],styles:[""],changeDetection:0}),i})();function Yn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-select-item-object",3),e.NdJ("selectedItemObject$",function(n){return e.CHM(t),e.oxw().processMinMethodSubSetting(n)}),e.qZA()}if(2&i){const t=e.oxw();e.Q6J("inputObjects",t.availableMinMethodsSettings)("idAttribute","name")("displayAttribute","description")("itemLabel","Settings for Impute with Minimum Value")("formInlineOrNot",!1)("ifOutputFirstValue",!0)("emptyNote","No MinMethods settings")}}function Wn(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"app-input-number-with-range",4),e.NdJ("outputNumber$",function(n){return e.CHM(t),e.oxw().processRandomMethodPercent(n)}),e.qZA()}2&i&&e.Q6J("minValue",.01)("maxValue",.3)("defaultValue",.05)("formInlineOrNot",!1)("step",.01)("itemLabel","Smallest n Percent Value to Construct Random Number")}let zn=(()=>{class i{constructor(t){this.cdr=t,this.imputeSetting$=new e.vpe,this.availableImputeMethods=[{name:"minimum_value_impute",description:"Impute with Minimum Value"},{name:"random_number_impute",description:"impute with Random Number Based on Distribution"},{name:"knn_impute",description:"kNN Imputation"},{name:"group_minimum_value_impute",description:"Impute with Minimum Value of Protein in Group"}],this.availableMinMethodsSettings=[{name:"global",description:"Min of All Samples"},{name:"sample",description:"Min of Each Sample"}]}ngOnInit(){}processSelectedImputeMethod(t){this.imputeMethod=t,("knn_impute"===t.name||"group_minimum_value_impute"===t.name)&&this.imputeSetting$.next({method:t.name}),this.cdr.detectChanges()}processMinMethodSubSetting(t){this.imputeSetting$.next({method:"minimum_value_impute",subSetting:t.name}),this.cdr.detectChanges()}processRandomMethodPercent(t){this.imputeSetting$.next({method:"random_number_impute",subSetting:.02})}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-proteomics-preprocess-impute-settings"]],outputs:{imputeSetting$:"imputeSetting$"},decls:3,vars:8,consts:[["data-testid","selectImputeMethod",3,"inputObjects","idAttribute","displayAttribute","itemLabel","formInlineOrNot","ifOutputFirstValue","emptyNote","selectedItemObject$"],["class","mt-2","data-testid","selectMinMethodSetting",3,"inputObjects","idAttribute","displayAttribute","itemLabel","formInlineOrNot","ifOutputFirstValue","emptyNote","selectedItemObject$",4,"ngIf"],[3,"minValue","maxValue","defaultValue","formInlineOrNot","step","itemLabel","outputNumber$",4,"ngIf"],["data-testid","selectMinMethodSetting",1,"mt-2",3,"inputObjects","idAttribute","displayAttribute","itemLabel","formInlineOrNot","ifOutputFirstValue","emptyNote","selectedItemObject$"],[3,"minValue","maxValue","defaultValue","formInlineOrNot","step","itemLabel","outputNumber$"]],template:function(t,s){1&t&&(e.TgZ(0,"app-select-item-object",0),e.NdJ("selectedItemObject$",function(o){return s.processSelectedImputeMethod(o)}),e.qZA(),e.YNc(1,Yn,1,7,"app-select-item-object",1),e.YNc(2,Wn,1,6,"app-input-number-with-range",2)),2&t&&(e.Q6J("inputObjects",s.availableImputeMethods)("idAttribute","name")("displayAttribute","description")("itemLabel","Method for Missing Value Impute")("formInlineOrNot",!1)("ifOutputFirstValue",!0),e.xp6(1),e.Q6J("ngIf","minimum_value_impute"===(null==s.imputeMethod?null:s.imputeMethod.name)),e.xp6(1),e.Q6J("ngIf","random_number_impute"===(null==s.imputeMethod?null:s.imputeMethod.name)))},directives:[Oe.w,g.O5,st.C],styles:[""],changeDetection:0}),i})();const Xn=["submitPreButton"];let Kn=(()=>{class i{constructor(t){this.cdr=t,this.formWidth="15rem",this.preProcessSettings$=new e.vpe,this.filterSettingsSubject=new c.X(null),this.imputeSettingsSubject=new c.X(null),this.normalizationMethodSubject=new c.X(null),this.normalizationMethod=[{name:"median_normalize",description:"Median Normalization"},{name:"mean_normalize",description:"Mean Normalization"},{name:"max_normalize",description:"Max Normalization"},{name:"fot_normalize",description:"Fot Normalization"},{name:"quantile_normalize",description:"Quantile Normalization"}],this.destroyNotifier$=new B.xQ}ngOnInit(){this.ifSubmitAvailable$=(0,b.aj)([this.filterSettingsSubject,this.imputeSettingsSubject,this.normalizationMethodSubject]).pipe((0,f.U)(([t,s,n])=>!!(t&&s&&n)),(0,T.O)(!0),(0,_.R)(this.destroyNotifier$))}ngAfterViewInit(){this.cdr.checkNoChanges(),(0,E.R)(this.submitPreButton.nativeElement,"click").pipe((0,r.h)(t=>null!=t),(0,y.M)(this.filterSettingsSubject.pipe((0,r.h)(t=>null!=t)),this.imputeSettingsSubject.pipe((0,r.h)(t=>null!=t)),this.normalizationMethodSubject.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s,n,o])=>({filterSettings:s,imputeSettings:n,normalizationMethod:o})),(0,_.R)(this.destroyNotifier$)).subscribe(t=>{this.preProcessSettings$.next(t)})}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processFilterSettings(t){this.filterSettingsSubject.next(t)}processImputeSettings(t){this.imputeSettingsSubject.next(t)}processNormalizationMethod(t){this.normalizationMethodSubject.next(t.name)}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-proteomics-preprocess-settings"]],viewQuery:function(t,s){if(1&t&&e.Gf(Xn,7),2&t){let n;e.iGM(n=e.CRH())&&(s.submitPreButton=n.first)}},inputs:{resetToDefault:"resetToDefault",formWidth:"formWidth"},outputs:{preProcessSettings$:"preProcessSettings$"},decls:13,vars:10,consts:[[3,"filterSetting$"],[1,"mt-2"],[3,"imputeSetting$"],[3,"inputObjects","idAttribute","displayAttribute","itemLabel","formInlineOrNot","ifOutputFirstValue","emptyNote","selectedItemObject$"],["data-testid","submitPreprocessSettings",1,"mt-2","p-2","btn","btn-sm","btn-expressvis",3,"disabled"],["submitPreButton",""]],template:function(t,s){1&t&&(e.TgZ(0,"h4"),e._uU(1,"Filter Proteins With NAs"),e.qZA(),e.TgZ(2,"app-proteomics-preprocess-filter-settings",0),e.NdJ("filterSetting$",function(o){return s.processFilterSettings(o)}),e.qZA(),e.TgZ(3,"h4",1),e._uU(4,"NAs Impute Settings"),e.qZA(),e.TgZ(5,"app-proteomics-preprocess-impute-settings",2),e.NdJ("imputeSetting$",function(o){return s.processImputeSettings(o)}),e.qZA(),e.TgZ(6,"h4",1),e._uU(7,"Normalization Settings"),e.qZA(),e.TgZ(8,"app-select-item-object",3),e.NdJ("selectedItemObject$",function(o){return s.processNormalizationMethod(o)}),e.qZA(),e.TgZ(9,"button",4,5),e.ALo(11,"async"),e._uU(12,"Submit Pre-process Settings\n"),e.qZA()),2&t&&(e.xp6(8),e.Q6J("inputObjects",s.normalizationMethod)("idAttribute","name")("displayAttribute","description")("itemLabel","Method")("formInlineOrNot",!0)("ifOutputFirstValue",!0)("emptyNote","No Normalization methods is available"),e.xp6(1),e.Q6J("disabled",!e.lcZ(11,8,s.ifSubmitAvailable$)))},directives:[Hn,zn,Oe.w],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();const ea=["displayPCAbefore"],ta=["displayPCAafter"],sa=["submitButton"];function ia(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 2. Import Quantitative Data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-import-proteomics-quantitative-data",15),e.NdJ("matchedIDsPercentageRequest$",function(n){return e.CHM(t),e.oxw().processMatchedIDsPercentageRequest(n)})("quantificationInfo$",function(n){return e.CHM(t),e.oxw().processProteomicsQuantificationData(n)}),e.ALo(6,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","The file required in this step should be directly from the quantitative software, like PANDA and MaxQuant."),e.xp6(1),e.Q6J("formWidth",t.formWidth)("resetSelection",t.ifResetChildrenSteps.importProteomicsQuantitative)("matchedIDsPercentageResult",e.lcZ(6,5,t.matchedIDsPercentageResult$))}}function na(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 3. Rename Samples "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-rename-sample-names",16),e.NdJ("skipSignal$",function(n){return e.CHM(t),e.oxw().skipRenameSampleNames(n)})("rawNames2newNames$",function(n){return e.CHM(t),e.oxw().processRawSampleNames2NewSampleNames(n)}),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","The sample names in the quantification file usually contains prefixes or suffixes. We can rename to make them more readable"),e.xp6(1),e.Q6J("rawSampleNames",t.rawSampleNames)("resetToDefault",t.ifResetChildrenSteps.renameSamples)("formWidth",t.formWidth)}}function aa(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 4. Import Phenotype data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"div"),e.TgZ(6,"app-import-phenotype-data",17),e.NdJ("skipSignal$",function(n){return e.CHM(t),e.oxw().skipImportPhenotypeDataStep(n)})("phenotypeData$",function(n){return e.CHM(t),e.oxw().processPhenotypeData(n)}),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Phenotype Data contains the annotations of the samples. We can use the sample annotations to define groups"),e.xp6(2),e.Q6J("resetToDefault",t.ifResetChildrenSteps.importPhenotypeData)("formWidth",t.formWidth)}}function oa(i,a){1&i&&(e.TgZ(0,"p",18),e._uU(1," Samples from normalized data and phenotype data should be the same. Please change the phenotype data file or select another column as sample column "),e.qZA())}function ra(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"h3",3),e._uU(2," Step 5. Define Groups "),e.qZA(),e.TgZ(3,"app-define-groups",19),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.ALo(4,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(1),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("phenotypeData",e.lcZ(4,5,t.inputPhenotypeData$))("ifShowSubmitButton",!1)("resetToDefault",t.ifResetChildrenSteps.defineGroups)("formWidth",t.formWidth)}}function pa(i,a){if(1&i&&(e.TgZ(0,"p"),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij(" ",t.statusValue," ")}}function la(i,a){if(1&i&&(e.TgZ(0,"div",23),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. The app is pre-processing data"),e.qZA(),e._UZ(3,"fa-icon",24),e.qZA()),2&i){const t=e.oxw(3);e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function ca(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,pa,2,1,"p",21),e.YNc(2,la,4,3,"div",22),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf","error"===(null==t?null:t.status)),e.xp6(1),e.Q6J("ngIf","loading"===(null==t?null:t.status))}}function ua(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step6. Input Preprocess Settings "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-proteomics-preprocess-settings",20),e.NdJ("preProcessSettings$",function(n){return e.CHM(t),e.oxw().processPreProcessSettings(n)}),e.qZA(),e.YNc(6,ca,3,2,"div",21),e.ALo(7,"async"),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","The preprocess settings will be used to preprocess quantitative values"),e.xp6(1),e.Q6J("resetToDefault",t.ifResetChildrenSteps.inputPreprocessSettings)("formWidth",t.formWidth),e.xp6(1),e.Q6J("ngIf",e.lcZ(7,5,t.preprocessProteomicsStatus$))}}function ma(i,a){if(1&i&&(e.TgZ(0,"p"),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij(" ",t.statusValue," ")}}function ha(i,a){if(1&i&&(e.TgZ(0,"div",23),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. The app is removing batch effects"),e.qZA(),e._UZ(3,"fa-icon",24),e.qZA()),2&i){const t=e.oxw(3);e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function da(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,ma,2,1,"p",21),e.YNc(2,ha,4,3,"div",22),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf","error"===(null==t?null:t.status)),e.xp6(1),e.Q6J("ngIf","loading"===(null==t?null:t.status))}}function fa(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",25),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 7. Remove Batch Effects "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"app-remove-batch-settings",26),e.NdJ("displayNormalizedPCAsignal$",function(n){return e.CHM(t),e.oxw().displayNormalizedPCAplot(n)})("displayPCAofValuesAfterRemovingBatchSignal$",function(n){return e.CHM(t),e.oxw().displayPCAofValuesAfterRemovingBatchEffect(n)})("removeBatchInfo$",function(n){return e.CHM(t),e.oxw().processRemoveBatchInfo(n)}),e.qZA(),e.YNc(6,da,3,2,"div",21),e.ALo(7,"async"),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Remove batch effects caused by laboratory conditions, choice of reagent, instruments used etc"),e.xp6(1),e.Q6J("phenotypeDataAttributes",t.phenotypeDataAttributes)("ifBatchEffectHasBeenRemoved",t.ifBatchEffectHasBeenRemoved),e.xp6(1),e.Q6J("ngIf",e.lcZ(7,5,t.proteomicsRemoveBatchStatus$))}}function ga(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",2),e.TgZ(2,"h3",3),e._uU(3," Step 8. Input Dataset name "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(5,"div",27),e.TgZ(6,"app-click-focus-input",28),e.NdJ("inputString$",function(n){return e.CHM(t),e.oxw().processInputDatasetName(n)}),e.qZA(),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","We can import multiple datasets. To distinguish them, each should has a unique name. We can select one in later analsyis"),e.xp6(2),e.Q6J("placeholder","Dataset Name")("resetToDefault",t.ifResetChildrenSteps.inputDatasetName)}}function _a(i,a){1&i&&e._UZ(0,"app-express-data-success-import-note",29)}function Sa(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",30),e.TgZ(1,"button",31),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e.TgZ(2,"span",32),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",33),e._UZ(5,"app-display-pca-plot-before-removing-batch",34),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",35),e.TgZ(8,"button",36),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("phenotypeData",e.lcZ(6,2,t.inputPhenotypeData$))("PCAinfoArray",t.beforeRemovingBatchPCAresult)}}function va(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",30),e.TgZ(1,"button",31),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e.TgZ(2,"span",32),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",33),e._UZ(5,"app-display-pcas-before-and-after-removing-batch",37),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",35),e.TgZ(8,"button",38),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("PCAinfoArrayBefore",t.beforeRemovingBatchPCAresult)("PCAinfoArrayAfter",t.afterRemovingBatchPCAresult)("phenotypeData",e.lcZ(6,3,t.inputPhenotypeData$))}}let ya=(()=>{class i{constructor(t,s,n){this.facade=t,this.modalService=s,this.cdr=n,this.faSpinner=$.LM3,this.quantitativeSoftwares=["MaxQuant","PANDA","Spectronaut","Proteome Discoverer"],this.availableSpecies=V.it,this.formWidth="15rem",this.samplesFilesInfo=[{name:"Proteomics-MaxQuant-human.zip",description:"Proteomics-MaxQuant-human"},{name:"Proteomics-PANDA-mouse.zip",description:"Proteomics-PANDA-mouse"}],this.downloadSampleDataPercentage$=this.facade.downloadSampleDataPercentage$,this.speciesID$=new c.X(null),this.quantitativeInfo$=new c.X(null),this.samples$=new c.X(null),this.inputPhenotypeData$=new c.X(null),this.groups$=new c.X(null),this.preProcessSettings$=new c.X(null),this.removeBatchSetting$=new c.X(null),this.datasetName$=new c.X(null),this.ifBeforeRemoveBatchPCAexist=!1,this.requestBeforeRemovingBatchPCAinfoSubject=new c.X(null),this.ifBatchEffectHasBeenRemoved=!1,this.ifAfterRemoveBatchPCAexist=!1,this.requestAfterRemovingBatchPCAinfoSubject=new c.X(null),this.ifShowUploadNoteFromReset$=new c.X(null),this.matchedIDsPercentageRequestSubject=new c.X(null),this.destroyNotifier$=new B.xQ,this.steps=[{stepName:"defineSpecies",upStep:"",nextStep:"importProteomicsQuantitative",resetChildren:!1},{stepName:"importProteomicsQuantitative",upStep:"defineSpecies",nextStep:"renameSamples",resetChildren:!1},{stepName:"renameSamples",upStep:"importProteomicsQuantitative",nextStep:"importPhenotypeData",resetChildren:!1},{stepName:"importPhenotypeData",upStep:"renameSamples",nextStep:"defineGroups",resetChildren:!1},{stepName:"defineGroups",upStep:"importPhenotypeData",nextStep:"inputPreprocessSettings",resetChildren:!1},{stepName:"inputPreprocessSettings",upStep:"importPhenotypeData",nextStep:"removeBatch",resetChildren:!1},{stepName:"removeBatch",upStep:"inputPreprocessSettings",nextStep:"inputDatasetName",resetChildren:!1},{stepName:"inputDatasetName",upStep:"removeBatch",nextStep:"clickSubmitButton",resetChildren:!1},{stepName:"clickSubmitButton",upStep:"inputDatasetName",nextStep:"",resetChildren:!1}],this.ifResetChildrenSteps={defineSpecies:!1,importProteomicsQuantitative:!1,renameSamples:!1,importPhenotypeData:!1,defineGroups:!1,inputPreprocessSettings:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.ifShowSteps={defineSpecies:!0,importProteomicsQuantitative:!1,renameSamples:!1,importPhenotypeData:!1,defineGroups:!1,inputPreprocessSettings:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.preprocessProteomicsStatus$=this.facade.preprocessProteomicsStatus$,this.proteomicsRemoveBatchStatus$=this.facade.removeBatchStatus$,this.preprocessProteomicsResult$=this.facade.preprocessProteomicsResult$.pipe((0,r.h)(o=>null===o||(null==o?void 0:o.preprocessID)===this.preprocessRequestID)),this.proteomicsRemoveBatchResult$=this.facade.removeBatchResult$.pipe((0,r.h)(o=>null===o||(null==o?void 0:o.removeBatchID)===this.removeBatchID)),this.facade.beforeRemovingBatchPCAresult$.pipe((0,r.h)(o=>null!=o),(0,r.h)(o=>(null==o?void 0:o.pcaRequestID)===this.preprocessRequestID)).subscribe(o=>{this.beforeRemovingBatchPCAresult=[...o.samplesPCAinfos],this.ifBeforeRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.facade.afterRemovingBatchPCAresult$.pipe((0,r.h)(o=>null!=o),(0,r.h)(o=>(null==o?void 0:o.pcaRequestID)===this.removeBatchID)).subscribe(o=>{this.afterRemovingBatchPCAresult=[...o.samplesPCAinfos],this.ifAfterRemoveBatchPCAexist=!0,this.cdr.detectChanges()})}ngOnInit(){this.proteomicsRemoveBatchResult$.pipe((0,r.h)(t=>null!=t)).subscribe(t=>{this.ifBatchEffectHasBeenRemoved=!0,this.cdr.detectChanges()}),this.ifShowSamplesNotSameNote$=(0,b.aj)([this.samples$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>!(0,v.z7)(t,[...s.samples])),(0,T.O)(!1)),this.matchedIDsPercentageResult$=this.facade.matchedIDsPercentageResult$}ngAfterViewInit(){this.ifSubmitAvailable$=(0,b.aj)([this.speciesID$,this.quantitativeInfo$,this.samples$,this.inputPhenotypeData$,this.groups$,this.preprocessProteomicsResult$,this.removeBatchSetting$,this.proteomicsRemoveBatchResult$,this.datasetName$]).pipe((0,f.U)(([t,s,n,o,l,u,h,p,S])=>!(!(t&&s&&n&&o&&l&&u&&h&&S)||h.ifRemoveBatch&&!p)),(0,Ce.b)(()=>{this.cdr.detectChanges()}),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$)),(0,b.aj)([this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t)),this.speciesID$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>Object.assign(Object.assign({},t),{speciesID:s}))).subscribe(t=>{this.facade.requestMatchedIDsPercentage(t)}),this.preProcessSettings$.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.quantitativeInfo$.pipe((0,r.h)(t=>null!=t)),this.samples$.pipe((0,r.h)(t=>null!=t)),this.groups$.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s,n,o])=>{const l=G();return this.preprocessRequestID=l,{preprocessID:l,proteinIDs:s.proteinIDs,samples:n,groups:o,samples2intensities:(0,v.D)(s.intensityInfo.values,n),preprocessSettings:t}})).subscribe(t=>{this.facade.preprocessProteomicsData(t)}),this.requestBeforeRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.preprocessProteomicsResult$),(0,f.U)(([t,s])=>{const n=(0,v.Ci)(s.normalizedValues,s.samples);return{pcaRequestID:this.preprocessRequestID,values:n,samples:s.samples}})).subscribe(t=>{this.beforeRemovingBatchPCAresult=[],this.ifBeforeRemoveBatchPCAexist=!1,this.cdr.detectChanges(),this.facade.requestBeforeRemovingBatchPCAInfo(t)}),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>!0===(null==t?void 0:t.ifRemoveBatch)),(0,y.M)(this.preprocessProteomicsResult$,this.inputPhenotypeData$,this.groups$,this.samples$),(0,f.U)(([t,s,n,o,l])=>{const u=G(),h=(0,v.Ci)(s.normalizedValues,l),S=k(s.samples,n)[t.batchAttribute];return t.ifUseGroupVariable?{removeBatchID:u,values:h,batch:S,group:me(o,l),valueType:"normalized"}:{removeBatchID:u,values:h,batch:S,group:null,valueType:"normalized"}})).subscribe(t=>{this.removeBatchID=t.removeBatchID,this.facade.processRemoveBatchEffectRequest(t)}),this.requestAfterRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.preprocessProteomicsResult$,this.proteomicsRemoveBatchResult$),(0,f.U)(([t,s,n])=>({pcaRequestID:this.removeBatchID,values:n.values,samples:s.samples}))).subscribe(t=>{this.afterRemovingBatchPCAresult=[],this.ifAfterRemoveBatchPCAexist=!1,this.facade.requestAfterRemovingBatchPCAinfo(t)}),(0,b.aj)([this.speciesID$,this.quantitativeInfo$,this.samples$,this.inputPhenotypeData$,this.groups$,this.preprocessProteomicsResult$,this.proteomicsRemoveBatchResult$,this.datasetName$]).subscribe(([t,s,n,o,l,u,h,p])=>{t&&this._displayNextStepOfCurrentStep("defineSpecies"),t&&s&&this._displayNextStepOfCurrentStep("importProteomicsQuantitative"),this.ifCanProceedInRenameSamplesStep&&t&&s&&n&&this._displayNextStepOfCurrentStep("renameSamples"),t&&s&&n&&o&&Object.keys(o).length>1&&(0,v.z7)(n,[...o.samples])&&this._displayNextStepOfCurrentStep("importPhenotypeData"),t&&s&&n&&o&&l&&this._displayNextStepOfCurrentStep("defineGroups"),t&&s&&n&&o&&l&&u&&u.preprocessID===this.preprocessRequestID&&this._displayNextStepOfCurrentStep("inputPreprocessSettings"),t&&s&&n&&o&&l&&u&&h&&h.removeBatchID===this.removeBatchID&&this._displayNextStepOfCurrentStep("removeBatch")}),this.submitStream$=(0,E.R)(this.submitButton.nativeElement,"click"),this.submitStream$.pipe((0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$),ce(500),(0,y.M)(this.speciesID$.pipe((0,r.h)(t=>null!=t)),this.quantitativeInfo$.pipe((0,r.h)(t=>null!=t)),this.samples$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t)),this.groups$.pipe((0,r.h)(t=>null!=t)),this.preprocessProteomicsResult$.pipe((0,r.h)(t=>null!=t)),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t)),this.proteomicsRemoveBatchResult$,this.datasetName$.pipe((0,r.h)(t=>null!=t)))).subscribe(([t,s,n,o,l,u,h,p,S,A])=>{const N={expressLocalOrRemote:K.M.local,datasetID:A,description:A,speciesID:s,experimentType:"Protein profiling by Mass Spec",IDtype:n.proteinIDtype,samples:o,groups:u},F=k(h.samples,l);let re;re=p.ifRemoveBatch?{datasetID:A,geneIDs:h.geneIDs,samples:h.samples,phenotypeData:F,expressionValues:S.values,defaultGroups:u}:{datasetID:A,geneIDs:h.geneIDs,samples:h.samples,phenotypeData:F,expressionValues:(0,v.Ci)(h.normalizedValues,h.samples),defaultGroups:u},this.facade.addDatasetInfo(N),this.facade.addLocalExpressData(re),this.facade.requestDatabaseGeneID2info({speciesID:s,idType:n.proteinIDtype})}),this.ifShowUploadNote$=(0,j.T)(this.facade.theLatestAddedDatasetID$.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.datasetName$.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>t===s),(0,_.R)(this.destroyNotifier$),(0,T.O)(!1)),this.ifShowUploadNoteFromReset$)}ngOnDestroy(){this.destroyNotifier$.next(),this.destroyNotifier$.complete()}processSampleFileNameToBeDownloaded(t){this.facade.downloadSampleFile(t)}processSelectedSpecies(t){t&&(this.speciesID$.next(t.taxID),this._resetFollowingSteps("defineSpecies"))}processMatchedIDsPercentageRequest(t){this.matchedIDsPercentageRequestSubject.next(t)}processProteomicsQuantificationData(t){this.quantitativeInfo$.next(t),t?(this.rawSampleNames=[...t.intensityInfo.intensityHeaders],this.samples$.next(this.rawSampleNames),this.inputPhenotypeData$.next({samples:this.rawSampleNames}),this._resetFollowingSteps("importProteomicsQuantitative"),this.cdr.detectChanges()):(this.rawSampleNames=[],this.samples$.next(this.rawSampleNames),this.inputPhenotypeData$.next(null),this.cdr.detectChanges())}skipRenameSampleNames(t){this._displayNextStepOfCurrentStep("renameSamples")}processRawSampleNames2NewSampleNames(t){if(t){this.ifCanProceedInRenameSamplesStep=!0;const s=this.rawSampleNames.map(n=>t.get(n));this.samples$.next(s),this.inputPhenotypeData$.next({samples:s}),this._resetFollowingSteps("renameSamples")}}skipImportPhenotypeDataStep(t){this._displayNextStepOfCurrentStep("importPhenotypeData")}processPhenotypeData(t){t&&(this.inputPhenotypeData$.next(t),this._resetFollowingSteps("importPhenotypeData"),this.phenotypeDataAttributes=Object.keys(t))}processGroups(t){this.groups$.next(t),this._resetFollowingSteps("defineGroups")}processPreProcessSettings(t){this.preProcessSettings$.next(t),this._resetFollowingSteps("inputPreprocessSettings")}displayNormalizedPCAplot(t){this.displayPCAbeforeModalRef=this.modalService.show(this.displayPCAbefore),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}processRemoveBatchInfo(t){this.removeBatchSetting$.next(t),t&&!t.ifRemoveBatch&&this._displayNextStepOfCurrentStep("removeBatch")}displayPCAofValuesAfterRemovingBatchEffect(t){this.displayPCAafterModalRef=this.modalService.show(this.displayPCAafter,Object.assign({},{class:"modal-xl"})),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0),this.ifAfterRemoveBatchPCAexist||this.requestAfterRemovingBatchPCAinfoSubject.next(!0)}processInputDatasetName(t){this.datasetName$.next(t),this._resetFollowingSteps("inputDatasetName")}_resetFollowingSteps(t){let s=[],n=!1;this.steps.forEach((o,l)=>{o.stepName===t?(s.push(Object.assign(Object.assign({},o),{resetChildren:!1})),this.ifResetChildrenSteps[o.stepName]=!1,n=!0):n?(s.push(Object.assign(Object.assign({},o),{resetChildren:!0})),this.ifResetChildrenSteps[o.stepName]=!0):(s.push(Object.assign(Object.assign({},o),{resetChildren:!1})),this.ifResetChildrenSteps[o.stepName]=!1)}),this.steps=[...s],this.ifResetChildrenSteps=Object.assign({},this.ifResetChildrenSteps),this.ifShowUploadNoteFromReset$.next(!1)}_displayNextStepOfCurrentStep(t){let s;this.steps.forEach((o,l)=>{o.stepName===t&&(s=l+1)}),this.ifShowSteps[this.steps[s].stepName]=!0,this.ifShowSteps=Object.assign({},this.ifShowSteps),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-proteomics"]],viewQuery:function(t,s){if(1&t&&(e.Gf(ea,5,e.Rgc),e.Gf(ta,5,e.Rgc),e.Gf(sa,7)),2&t){let n;e.iGM(n=e.CRH())&&(s.displayPCAbefore=n.first),e.iGM(n=e.CRH())&&(s.displayPCAafter=n.first),e.iGM(n=e.CRH())&&(s.submitButton=n.first)}},decls:29,vars:25,consts:[[1,"mt-4","p-2","border"],[3,"fileInfos","downloadSampleDataPercentage","fileNameToBeDownloaded$"],[1,"d-flex"],[3,"appAddRequiredOptionalSign"],[1,"align-self-center","ml-3","questionMark",3,"note"],[3,"availableSpecies","ifOutputFirstValue","formWidth","selectedSpecies$"],["class","mt-4 p-2 border",4,"ngIf"],[1,"mt-3"],["class","alert alert-warning",4,"ngIf"],["class","mt-4 p-1 border",4,"ngIf"],["data-testid","submitProteomics",1,"mt-4","p-2","btn","btn-expressvis",3,"disabled"],["submitButton",""],["data-testid","uploadNote",4,"ngIf"],["displayPCAbefore",""],["displayPCAafter",""],[3,"formWidth","resetSelection","matchedIDsPercentageResult","matchedIDsPercentageRequest$","quantificationInfo$"],[3,"rawSampleNames","resetToDefault","formWidth","skipSignal$","rawNames2newNames$"],["data-testid","loadPhenotype",3,"resetToDefault","formWidth","skipSignal$","phenotypeData$"],[1,"alert","alert-warning"],[3,"phenotypeData","ifShowSubmitButton","resetToDefault","formWidth","outGroups"],[3,"resetToDefault","formWidth","preProcessSettings$"],[4,"ngIf"],["class","form-inline",4,"ngIf"],[1,"form-inline"],[1,"hoverBigger",3,"icon","size","spin"],[1,"mt-4","p-1","border"],["data-testid","removeBatch",3,"phenotypeDataAttributes","ifBatchEffectHasBeenRemoved","displayNormalizedPCAsignal$","displayPCAofValuesAfterRemovingBatchSignal$","removeBatchInfo$"],[1,"mt-2"],["data-testid","inputDatasetName",3,"placeholder","resetToDefault","inputString$"],["data-testid","uploadNote"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[3,"phenotypeData","PCAinfoArray"],[1,"modal-footer"],["type","button","data-testid","closePCAbeforeRemoveBatch",1,"btn","btn-default",3,"click"],[3,"PCAinfoArrayBefore","PCAinfoArrayAfter","phenotypeData"],["type","button","data-testid","closePCAafterRemoveBatch",1,"btn","btn-default",3,"click"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"app-download-sample-data",1),e.NdJ("fileNameToBeDownloaded$",function(o){return s.processSampleFileNameToBeDownloaded(o)}),e.ALo(2,"async"),e.qZA(),e.qZA(),e.TgZ(3,"div",0),e.TgZ(4,"div",2),e.TgZ(5,"h3",3),e._uU(6," Step 1. Select Species "),e.qZA(),e._UZ(7,"app-display-notes-with-question-mark",4),e.qZA(),e.TgZ(8,"app-select-species",5),e.NdJ("selectedSpecies$",function(o){return s.processSelectedSpecies(o)}),e.qZA(),e.qZA(),e.YNc(9,ia,7,7,"div",6),e.YNc(10,na,6,5,"div",6),e.YNc(11,aa,7,4,"div",6),e.TgZ(12,"div",7),e.YNc(13,oa,2,0,"p",8),e.ALo(14,"async"),e.qZA(),e.YNc(15,ra,5,7,"div",6),e.YNc(16,ua,8,7,"div",6),e.YNc(17,fa,8,7,"div",9),e.YNc(18,ga,7,4,"div",6),e.TgZ(19,"button",10,11),e.ALo(21,"async"),e._uU(22,"Submit Data\n"),e.qZA(),e.YNc(23,_a,1,0,"app-express-data-success-import-note",12),e.ALo(24,"async"),e.YNc(25,Sa,10,4,"ng-template",null,13,e.W1O),e.YNc(27,va,10,5,"ng-template",null,14,e.W1O)),2&t&&(e.xp6(1),e.Q6J("fileInfos",s.samplesFilesInfo)("downloadSampleDataPercentage",e.lcZ(2,17,s.downloadSampleDataPercentage$)),e.xp6(4),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","With the species information, the app can load annotations of the proteins in the dataset"),e.xp6(1),e.Q6J("availableSpecies",s.availableSpecies)("ifOutputFirstValue",!1)("formWidth",s.formWidth),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importProteomicsQuantitative),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.renameSamples),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importPhenotypeData),e.xp6(2),e.Q6J("ngIf",e.lcZ(14,19,s.ifShowSamplesNotSameNote$)),e.xp6(2),e.Q6J("ngIf",s.ifShowSteps.defineGroups),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.inputPreprocessSettings),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.removeBatch),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.inputDatasetName),e.xp6(1),e.Q6J("disabled",!e.lcZ(21,21,s.ifSubmitAvailable$)),e.xp6(4),e.Q6J("ngIf",!!e.lcZ(24,23,s.ifShowUploadNote$)))},directives:[he,de,U.m,fe.a,g.O5,xn,Vn,Se,ye,Kn,P.BN,Ae,be.y,xe,De,Te],pipes:[g.Ov],styles:[""]}),i})();const Aa=["displayPCAbefore"],ba=["displayPCAafter"],xa=["submitbutton"];function Da(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",3),e.TgZ(2,"h3",4),e._uU(3," Step 2. Import normalized data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",5),e.qZA(),e.TgZ(5,"app-import-normalized-values",16),e.NdJ("matchedIDsPercentageRequest$",function(n){return e.CHM(t),e.oxw().processMatchedIDsPercentageRequest(n)})("expressionImported$",function(n){return e.CHM(t),e.oxw().processExpressionValues(n)}),e.ALo(6,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","The expression data should be normalized values. That is to say, the values between different samples should be comparable"),e.xp6(1),e.Q6J("availableIDtypes",t.availableGeneIDtypes)("formWidth",t.formWidth)("fileType","Expression Values File")("IDtypeLabel","Gene/Protein ID Type in the File")("IDcolumnLabel","Select Gene/Protein ID Column")("resetSelection",t.ifResetChildrenSteps.importNormalizedData)("matchedIDsPercentageResult",e.lcZ(6,9,t.matchedIDsPercentageResult$))}}function Ta(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",3),e.TgZ(2,"h3",4),e._uU(3," Step 3. Import Phenotype data "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",5),e.qZA(),e.TgZ(5,"app-import-phenotype-data",17),e.NdJ("phenotypeData$",function(n){return e.CHM(t),e.oxw().processPhenotypeData(n)})("skipSignal$",function(n){return e.CHM(t),e.oxw().skipImportPhenotypeData(n)}),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Phenotype Data contains the annotations of the samples. We can use the sample annotations to define groups"),e.xp6(1),e.Q6J("resetToDefault",t.ifResetChildrenSteps.importPhenotypeData)}}function Ca(i,a){1&i&&(e.TgZ(0,"p",18),e._uU(1," Samples from normalized data and phenotype data should be the same. Please change the phenotype data file or select another column as sample column "),e.qZA())}function Ia(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"h3",4),e._uU(2," Step 4. Define Groups "),e.qZA(),e.TgZ(3,"app-define-groups",19),e.NdJ("outGroups",function(n){return e.CHM(t),e.oxw().processGroups(n)}),e.ALo(4,"async"),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(1),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("phenotypeData",e.lcZ(4,3,t.inputPhenotypeData$))("ifShowSubmitButton",!1)}}function Za(i,a){if(1&i&&(e.TgZ(0,"p"),e._uU(1),e.qZA()),2&i){const t=e.oxw().ngIf;e.xp6(1),e.hij(" ",t.statusValue," ")}}function Pa(i,a){if(1&i&&(e.TgZ(0,"div",24),e.TgZ(1,"p"),e._uU(2,"Wait for a moment. The app is removing batch effects"),e.qZA(),e._UZ(3,"fa-icon",25),e.qZA()),2&i){const t=e.oxw(3);e.xp6(3),e.Q6J("icon",t.faSpinner)("size","2x")("spin",!0)}}function Na(i,a){if(1&i&&(e.TgZ(0,"div"),e.YNc(1,Za,2,1,"p",22),e.YNc(2,Pa,4,3,"div",23),e.qZA()),2&i){const t=a.ngIf;e.xp6(1),e.Q6J("ngIf","error"===(null==t?null:t.status)),e.xp6(1),e.Q6J("ngIf","loading"===(null==t?null:t.status))}}function wa(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",20),e.TgZ(1,"div",3),e.TgZ(2,"h3",4),e._uU(3," Step 5. Remove Batch Effects "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",5),e.qZA(),e.TgZ(5,"app-remove-batch-settings",21),e.NdJ("displayNormalizedPCAsignal$",function(n){return e.CHM(t),e.oxw().displayNormalizedPCAplot(n)})("displayPCAofValuesAfterRemovingBatchSignal$",function(n){return e.CHM(t),e.oxw().displayPCAofValuesAfterRemovingBatchEffect(n)})("removeBatchInfo$",function(n){return e.CHM(t),e.oxw().processRemoveBatchInfo(n)}),e.qZA(),e.YNc(6,Na,3,2,"div",22),e.ALo(7,"async"),e.qZA()}if(2&i){const t=e.oxw();e.xp6(2),e.Q6J("appAddRequiredOptionalSign","optional"),e.xp6(2),e.Q6J("note","Remove batch effects caused by laboratory conditions, choice of reagent, instruments used etc"),e.xp6(1),e.Q6J("phenotypeDataAttributes",t.phenotypeDataAttributes)("ifBatchEffectHasBeenRemoved",t.ifBatchEffectHasBeenRemoved),e.xp6(1),e.Q6J("ngIf",e.lcZ(7,5,t.otherRemoveBatchStatus$))}}function qa(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",0),e.TgZ(1,"div",3),e.TgZ(2,"h3",4),e._uU(3," Step 6. Input Dataset name "),e.qZA(),e._UZ(4,"app-display-notes-with-question-mark",5),e.qZA(),e.TgZ(5,"div",8),e.TgZ(6,"app-click-focus-input",26),e.NdJ("inputString$",function(n){return e.CHM(t),e.oxw().processInputDatasetName(n)}),e.qZA(),e.qZA(),e.qZA()}2&i&&(e.xp6(2),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","We can import multiple datasets. To distinguish them, each should has a unique name. We can select one in later analsyis"),e.xp6(2),e.Q6J("placeholder","Dataset Name"))}function Ra(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pca-plot-before-removing-batch",31),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",33),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAbeforeModalRef?null:n.displayPCAbeforeModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("phenotypeData",e.lcZ(6,2,t.inputPhenotypeData$))("PCAinfoArray",t.beforeRemovingBatchPCAresult)}}function $a(i,a){if(1&i){const t=e.EpF();e.TgZ(0,"div",27),e.TgZ(1,"button",28),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e.TgZ(2,"span",29),e._uU(3,"\xd7"),e.qZA(),e.qZA(),e.qZA(),e.TgZ(4,"div",30),e._UZ(5,"app-display-pcas-before-and-after-removing-batch",34),e.ALo(6,"async"),e.qZA(),e.TgZ(7,"div",32),e.TgZ(8,"button",35),e.NdJ("click",function(){e.CHM(t);const n=e.oxw();return null==n.displayPCAafterModalRef?null:n.displayPCAafterModalRef.hide()}),e._uU(9," close "),e.qZA(),e.qZA()}if(2&i){const t=e.oxw();e.xp6(5),e.Q6J("PCAinfoArrayBefore",t.beforeRemovingBatchPCAresult)("PCAinfoArrayAfter",t.afterRemovingBatchPCAresult)("phenotypeData",e.lcZ(6,3,t.inputPhenotypeData$))}}function Ba(i,a){1&i&&e._UZ(0,"app-express-data-success-import-note",36)}let Fa=(()=>{class i extends ie.e{constructor(t,s,n){super(),this.facade=t,this.modalService=s,this.cdr=n,this.faSpinner=$.LM3,this.formWidth="15rem",this.availableSpecies=V.it,this.availableGeneIDtypes=V.k1,this.samplesFilesInfo=[{name:"Proteomics-preprocessed-Gao.zip",description:"Proteomics-preprocessed-Gao"},{name:"Remove_Batch_Effect_Demo.zip",description:"Proteomics-used-to-test-batch-correction"}],this.downloadSampleDataPercentage$=this.facade.downloadSampleDataPercentage$,this.steps=[{stepName:"defineSpecies",upStep:"",nextStep:"importNormalizedData",resetChildren:!1},{stepName:"importNormalizedData",upStep:"defineSpecies",nextStep:"importPhenotypeData",resetChildren:!1},{stepName:"importPhenotypeData",upStep:"importNormalizedData",nextStep:"defineGroups",resetChildren:!1},{stepName:"defineGroups",upStep:"importPhenotypeData",nextStep:"removeBatch",resetChildren:!1},{stepName:"removeBatch",upStep:"defineGroups",nextStep:"inputDatasetName",resetChildren:!1},{stepName:"inputDatasetName",upStep:"removeBatch",nextStep:"clickSubmitButton",resetChildren:!1},{stepName:"clickSubmitButton",upStep:"inputDatasetName",nextStep:"",resetChildren:!1}],this.ifResetChildrenSteps={defineSpecies:!1,importNormalizedData:!1,importPhenotypeData:!1,defineGroups:!1,inputDatasetName:!1,clickSubmitButton:!1},this.ifShowSteps={defineSpecies:!0,importNormalizedData:!1,importPhenotypeData:!1,defineGroups:!1,removeBatch:!1,inputDatasetName:!1,clickSubmitButton:!1},this.ifShowUploadNoteFromReset$=new c.X(null),this.matchedIDsPercentageRequestSubject=new c.X(null),this.ifBeforeRemoveBatchPCAexist=!1,this.requestBeforeRemovingBatchPCAinfoSubject=new c.X(null),this.ifBatchEffectHasBeenRemoved=!1,this.ifAfterRemoveBatchPCAexist=!1,this.requestAfterRemovingBatchPCAinfoSubject=new c.X(null),this.selectedSpecies$=new c.X(null),this.expressionValuesInfo$=new c.X(null),this.inputPhenotypeData$=new c.X(null),this.groups$=new c.X(null),this.removeBatchSetting$=new c.X(null),this.datasetName$=new c.X(null),this.datasetName=""}ngOnInit(){this.ifShowSamplesNotSameNote$=(0,b.aj)([this.expressionValuesInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t))]).pipe((0,f.U)(([t,s])=>!(0,v.z7)([...t.samples],[...s.samples])),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$)),this.matchedIDsPercentageResult$=this.facade.matchedIDsPercentageResult$,this.facade.beforeRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.pcaRequestID)).subscribe(t=>{this.beforeRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifBeforeRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.otherRemoveBatchResult$=this.facade.removeBatchResult$.pipe((0,r.h)(t=>null===t||(null==t?void 0:t.removeBatchID)===this.pcaRequestID)),this.otherRemoveBatchResult$.pipe((0,r.h)(t=>null!=t)).subscribe(t=>{this.ifBatchEffectHasBeenRemoved=!0,this._displayNextOfCurrentStep("removeBatch"),this.cdr.detectChanges()}),this.facade.afterRemovingBatchPCAresult$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>(null==t?void 0:t.pcaRequestID)===this.pcaRequestID)).subscribe(t=>{this.afterRemovingBatchPCAresult=[...t.samplesPCAinfos],this.ifAfterRemoveBatchPCAexist=!0,this.cdr.detectChanges()}),this.ifSubmitAvailable$=(0,b.aj)([this.selectedSpecies$,this.expressionValuesInfo$,this.inputPhenotypeData$,this.datasetName$,this.groups$]).pipe((0,f.U)(([t,s,n,o,l])=>!!(t&&s&&n&&o&&l)&&!!(0,v.z7)(s.samples,n.samples)),(0,T.O)(!1),(0,_.R)(this.destroyNotifier$))}ngAfterViewInit(){this.submitStream$=(0,E.R)(this.button.nativeElement,"click"),this.submitStream$.pipe((0,r.h)(t=>null!=t),(0,_.R)(this.destroyNotifier$),ce(500),(0,y.M)(this.selectedSpecies$.pipe((0,r.h)(t=>null!=t)),this.expressionValuesInfo$.pipe((0,r.h)(t=>null!=t)),this.inputPhenotypeData$.pipe((0,r.h)(t=>null!=t)),this.groups$.pipe((0,r.h)(t=>null!=t)),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t)),this.otherRemoveBatchResult$,this.datasetName$.pipe((0,r.h)(t=>null!=t)))).subscribe(([t,s,n,o,l,u,h,p])=>{const A=k(n.samples,o);let F;F=u.ifRemoveBatch?{geneIDs:n.IDs,samples:n.samples,expressionValues:h.values,datasetID:p,phenotypeData:A}:{geneIDs:n.IDs,samples:n.samples,expressionValues:n.values,datasetID:p,phenotypeData:A},this.facade.addDatasetInfo({expressLocalOrRemote:K.M.local,datasetID:p,description:p,speciesID:s.taxID,speciesName:s.name,IDtype:n.IDtype,ifLogTransformed:!0,experimentType:"Other",samples:n.samples,groups:l}),this.facade.addLocalExpressData(F),this.facade.requestDatabaseGeneID2info({speciesID:s.taxID,idType:n.IDtype})}),this.matchedIDsPercentageRequestSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.selectedSpecies$),(0,f.U)(([t,s])=>Object.assign(Object.assign({},t),{speciesID:s.taxID}))).subscribe(t=>{this.facade.requestMatchedIDsPercentage(t)}),this.requestBeforeRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.expressionValuesInfo$),(0,f.U)(([t,s])=>{const n=G();return this.pcaRequestID=n,{pcaRequestID:n,values:s.values,samples:s.samples}})).subscribe(t=>{this.beforeRemovingBatchPCAresult=[],this.ifBeforeRemoveBatchPCAexist=!1,this.facade.requestBeforeRemovingBatchPCAInfo(t),this.cdr.detectChanges()}),this.removeBatchSetting$.pipe((0,r.h)(t=>null!=t),(0,r.h)(t=>!0===(null==t?void 0:t.ifRemoveBatch)),(0,y.M)(this.expressionValuesInfo$,this.inputPhenotypeData$,this.groups$),(0,f.U)(([t,s,n,o])=>{const l=s.samples,h=k(l,n)[t.batchAttribute];if(t.ifUseGroupVariable){const p=me(o,l);return{removeBatchID:this.pcaRequestID,values:s.values,batch:h,group:p,valueType:"normalized"}}return{removeBatchID:this.pcaRequestID,values:s.values,batch:h,group:null,valueType:"normalized"}})).subscribe(t=>{this.facade.processRemoveBatchEffectRequest(t)}),this.requestAfterRemovingBatchPCAinfoSubject.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.expressionValuesInfo$,this.otherRemoveBatchResult$),(0,f.U)(([t,s,n])=>({pcaRequestID:this.pcaRequestID,values:n.values,samples:s.samples}))).subscribe(t=>{this.afterRemovingBatchPCAresult=[],this.ifAfterRemoveBatchPCAexist=!1,this.facade.requestAfterRemovingBatchPCAinfo(t)}),this.ifShowUploadNote$=this.facade.theLatestAddedDatasetID$.pipe((0,r.h)(t=>null!=t),(0,y.M)(this.datasetName$.pipe((0,r.h)(t=>null!=t))),(0,f.U)(([t,s])=>t===s),(0,_.R)(this.destroyNotifier$),(0,T.O)(!1))}processSampleFileNameToBeDownloaded(t){this.facade.downloadSampleFile(t)}processSelectedSpecies(t){t?(this.selectedSpecies$.next(t),this._displayNextOfCurrentStep("defineSpecies"),this._resetFollowingSteps("defineSpecies")):this._notDisplayFollowingStepsOfCurrentStep("defineSpecies")}processMatchedIDsPercentageRequest(t){this.matchedIDsPercentageRequestSubject.next(t)}processExpressionValues(t){this.expressionValuesInfo$.next(t),this.ifResetChildrenSteps.importNormalizedData=!1,t?(this.inputPhenotypeData$.next({samples:t.samples}),this._displayNextOfCurrentStep("importNormalizedData"),this._resetFollowingSteps("importNormalizedData"),this.cdr.detectChanges()):(this.inputPhenotypeData$.next(null),this._notDisplayFollowingStepsOfCurrentStep("importNormalizedData"),this.cdr.detectChanges())}processPhenotypeData(t){this.ifResetChildrenSteps.importPhenotypeData=!1,t?(this.inputPhenotypeData$.next(t),this._displayNextOfCurrentStep("importPhenotypeData"),this._resetFollowingSteps("importPhenotypeData"),this.phenotypeDataAttributes=Object.keys(t)):this._notDisplayFollowingStepsOfCurrentStep("importPhenotypeData")}skipImportPhenotypeData(t){this._displayNextOfCurrentStep("importPhenotypeData"),this._resetFollowingSteps("importPhenotypeData")}processGroups(t){this.ifResetChildrenSteps.defineGroups=!1,(null==t?void 0:t.length)>0?(this.groups$.next(t),this._resetFollowingSteps("defineGroups"),this._displayNextOfCurrentStep("defineGroups")):this._notDisplayFollowingStepsOfCurrentStep("defineGroups")}displayNormalizedPCAplot(t){this.displayPCAbeforeModalRef=this.modalService.show(this.displayPCAbefore),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0)}processRemoveBatchInfo(t){this.removeBatchSetting$.next(t),t&&!t.ifRemoveBatch&&this._displayNextOfCurrentStep("removeBatch")}displayPCAofValuesAfterRemovingBatchEffect(t){this.displayPCAafterModalRef=this.modalService.show(this.displayPCAafter,Object.assign({},{class:"modal-xl"})),this.ifBeforeRemoveBatchPCAexist||this.requestBeforeRemovingBatchPCAinfoSubject.next(!0),this.ifAfterRemoveBatchPCAexist||this.requestAfterRemovingBatchPCAinfoSubject.next(!0)}processInputDatasetName(t){this.datasetName$.next(t)}_displayNextOfCurrentStep(t){let s;this.steps.forEach((n,o)=>{n.stepName===t&&(s=o+1)}),this.steps.forEach((n,o)=>{this.ifShowSteps[n.stepName]=o<=s}),this.cdr.detectChanges()}_notDisplayFollowingStepsOfCurrentStep(t){let s;this.steps.forEach((n,o)=>{n.stepName===t&&(s=o)}),this.steps.forEach((n,o)=>{this.ifShowSteps[n.stepName]=o<=s}),this.cdr.detectChanges()}_resetFollowingSteps(t){let s=[],n=!1;this.steps.forEach((o,l)=>{o.stepName===t?(s.push(Object.assign(Object.assign({},o),{resetChildren:!1})),this.ifResetChildrenSteps[o.stepName]=!1,n=!0):n?(s.push(Object.assign(Object.assign({},o),{resetChildren:!0})),this.ifResetChildrenSteps[o.stepName]=!0):(s.push(Object.assign(Object.assign({},o),{resetChildren:!1})),this.ifResetChildrenSteps[o.stepName]=!1)}),this.ifResetChildrenSteps=Object.assign({},this.ifResetChildrenSteps),this.ifShowUploadNoteFromReset$.next(!1),this.cdr.detectChanges()}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W),e.Y36(q.tT),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-expression-other"]],viewQuery:function(t,s){if(1&t&&(e.Gf(Aa,5,e.Rgc),e.Gf(ba,5,e.Rgc),e.Gf(xa,5)),2&t){let n;e.iGM(n=e.CRH())&&(s.displayPCAbefore=n.first),e.iGM(n=e.CRH())&&(s.displayPCAafter=n.first),e.iGM(n=e.CRH())&&(s.button=n.first)}},features:[e.qOj],decls:34,vars:22,consts:[[1,"mt-4","p-2","border"],[1,"mt-2","p-2","border"],[3,"fileInfos","downloadSampleDataPercentage","fileNameToBeDownloaded$"],[1,"d-flex"],[3,"appAddRequiredOptionalSign"],[1,"align-self-center","ml-3","questionMark",3,"note"],[3,"availableSpecies","ifOutputFirstValue","selectedSpecies$"],["class","mt-4 p-2 border",4,"ngIf"],[1,"mt-3"],["class","alert alert-warning",4,"ngIf"],["class","mt-4 p-1 border",4,"ngIf"],["displayPCAbefore",""],["displayPCAafter",""],["data-testid","submitOtherExpressionData",1,"btn","btn-expressvis",3,"disabled"],["submitbutton",""],["data-testid","uploadNote",4,"ngIf"],[3,"availableIDtypes","formWidth","fileType","IDtypeLabel","IDcolumnLabel","resetSelection","matchedIDsPercentageResult","matchedIDsPercentageRequest$","expressionImported$"],["appTableHeaderBodySameWidth","",3,"resetToDefault","phenotypeData$","skipSignal$"],[1,"alert","alert-warning"],[3,"phenotypeData","ifShowSubmitButton","outGroups"],[1,"mt-4","p-1","border"],["data-testid","removeBatch",3,"phenotypeDataAttributes","ifBatchEffectHasBeenRemoved","displayNormalizedPCAsignal$","displayPCAofValuesAfterRemovingBatchSignal$","removeBatchInfo$"],[4,"ngIf"],["class","form-inline",4,"ngIf"],[1,"form-inline"],[1,"hoverBigger",3,"icon","size","spin"],[3,"placeholder","inputString$"],[1,"modal-header"],["type","button","aria-label","Close",1,"btn-close","close","pull-right",3,"click"],["aria-hidden","true",1,"visually-hidden"],[1,"modal-body"],[3,"phenotypeData","PCAinfoArray"],[1,"modal-footer"],["type","button","data-testid","closePCAbeforeRemoveBatch",1,"btn","btn-default",3,"click"],[3,"PCAinfoArrayBefore","PCAinfoArrayAfter","phenotypeData"],["type","button","data-testid","closePCAafterRemoveBatch",1,"btn","btn-default",3,"click"],["data-testid","uploadNote"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"p"),e._uU(2," This module is used to import expression profile data that can not be imported in other modules. The expression values "),e.TgZ(3,"strong"),e._uU(4,"must be normalized values"),e.qZA(),e._uU(5,". That is to say, the values between different samples should be comparable. "),e.qZA(),e.qZA(),e.TgZ(6,"div",1),e.TgZ(7,"app-download-sample-data",2),e.NdJ("fileNameToBeDownloaded$",function(o){return s.processSampleFileNameToBeDownloaded(o)}),e.ALo(8,"async"),e.qZA(),e.qZA(),e.TgZ(9,"div",0),e.TgZ(10,"div",3),e.TgZ(11,"h3",4),e._uU(12," Step 1. Select Species "),e.qZA(),e._UZ(13,"app-display-notes-with-question-mark",5),e.qZA(),e.TgZ(14,"app-select-species",6),e.NdJ("selectedSpecies$",function(o){return s.processSelectedSpecies(o)}),e.qZA(),e.qZA(),e.YNc(15,Da,7,11,"div",7),e.YNc(16,Ta,6,3,"div",7),e.TgZ(17,"div",8),e.YNc(18,Ca,2,0,"p",9),e.ALo(19,"async"),e.qZA(),e.YNc(20,Ia,5,5,"div",7),e.YNc(21,wa,8,7,"div",10),e.YNc(22,qa,7,3,"div",7),e.YNc(23,Ra,10,4,"ng-template",null,11,e.W1O),e.YNc(25,$a,10,5,"ng-template",null,12,e.W1O),e.TgZ(27,"div",0),e.TgZ(28,"button",13,14),e.ALo(30,"async"),e._uU(31," Upload Data "),e.qZA(),e.qZA(),e.YNc(32,Ba,1,0,"app-express-data-success-import-note",15),e.ALo(33,"async")),2&t&&(e.xp6(7),e.Q6J("fileInfos",s.samplesFilesInfo)("downloadSampleDataPercentage",e.lcZ(8,14,s.downloadSampleDataPercentage$)),e.xp6(4),e.Q6J("appAddRequiredOptionalSign","required"),e.xp6(2),e.Q6J("note","With the species information, the app can load annotations of the genes in the dataset"),e.xp6(1),e.Q6J("availableSpecies",s.availableSpecies)("ifOutputFirstValue",!1),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importNormalizedData),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.importPhenotypeData),e.xp6(2),e.Q6J("ngIf",e.lcZ(19,16,s.ifShowSamplesNotSameNote$)),e.xp6(2),e.Q6J("ngIf",s.ifShowSteps.defineGroups),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.removeBatch),e.xp6(1),e.Q6J("ngIf",s.ifShowSteps.inputDatasetName),e.xp6(6),e.Q6J("disabled",!e.lcZ(30,18,s.ifSubmitAvailable$)),e.xp6(4),e.Q6J("ngIf",!!e.lcZ(33,20,s.ifShowUploadNote$)))},directives:[he,de,U.m,fe.a,g.O5,ze,Se,Fe.n,ye,Ae,P.BN,be.y,De,Te,xe],pipes:[g.Ov],styles:[""],changeDetection:0}),i})();function Oa(i,a){1&i&&(e.TgZ(0,"p",4),e._uU(1," Your project have been loaded successfully. Click one app to continue your exploration\n"),e.qZA())}let Ma=(()=>{class i{constructor(t,s){this.facade=t,this.cdr=s,this.datasetIDsToBeLoaded=[],this.ifShowLoadSuccessNote=!1}ngOnInit(){this.facade.allDatasetInfoForDisplay$.pipe((0,r.h)(Boolean)).subscribe(t=>{if((null==t?void 0:t.length)>=1){const s=t.map(n=>n.datasetID);this.ifShowLoadSuccessNote=(0,v.z7)(s,this.datasetIDsToBeLoaded),this.cdr.detectChanges()}})}ngOnChanges(t){}processSelectedProjectJsonFile(t){const s=t.target.files[0],n=new FileReader;n.readAsText(s),n.onload=o=>{const u=JSON.parse(n.result);this.datasetIDsToBeLoaded=u.datasetListState.ids,this.facade.prepareLocalProject(u)}}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W),e.Y36(e.sBO))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-load-existed-project"]],features:[e.TTD],decls:8,vars:1,consts:[["border","",1,"mt-4","p-1"],["data-testid","loadLocalProject",1,"mt-2"],["type","file",1,"form-control","form-control-sm",3,"change"],["class","mt-2 p-1",4,"ngIf"],[1,"mt-2","p-1"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"p"),e._uU(2,"Load the project downloaded before to restore your analysis"),e.qZA(),e.qZA(),e.TgZ(3,"div",1),e.TgZ(4,"h5"),e._uU(5,"Project Json File Downloaded Before:"),e.qZA(),e.TgZ(6,"input",2),e.NdJ("change",function(o){return s.processSelectedProjectJsonFile(o)}),e.qZA(),e.qZA(),e.YNc(7,Oa,2,0,"p",3)),2&t&&(e.xp6(7),e.Q6J("ngIf",s.ifShowLoadSuccessNote))},directives:[g.O5],styles:[""]}),i})();function Ua(i,a){1&i&&e._UZ(0,"app-import-rna-seq")}function Ja(i,a){1&i&&e._UZ(0,"app-import-microarray")}function Qa(i,a){1&i&&e._UZ(0,"app-import-proteomics")}function Ea(i,a){1&i&&e._UZ(0,"app-import-expression-other")}function Ga(i,a){1&i&&e._UZ(0,"app-load-existed-project")}const oe=function(i){return{active:i}},ja=[{path:"",component:(()=>{class i{constructor(t){this.importDataFacade=t,this.dataType="Proteomics"}ngOnInit(){this.allDatasetInfoForDisplay$=this.importDataFacade.allDatasetInfoForDisplay$}setActiveDataType(t){this.dataType=t}processGroups(t){this.importDataFacade.updateDatasetGroups(t)}processDatasetIDforDelete(t){this.importDataFacade.deleteThisDataset(t)}}return i.\u0275fac=function(t){return new(t||i)(e.Y36(W))},i.\u0275cmp=e.Xpm({type:i,selectors:[["app-import-data"]],decls:27,vars:23,consts:[[1,"container-fluid"],[1,"row"],[1,"col-md-3","col-xl-2","border-right","bd-sidebar"],[3,"datasets","selectedDatasetIDforDelete$"],[1,"col-sm-12","col-md-9","col-xl-9","border-right","mb-5","bd-content"],["data-testid","differentDataSource",1,"nav","nav-tabs"],[1,"nav-item"],["href","javascript:void(0)",1,"nav-link",3,"ngClass","click"],[4,"ngIf"]],template:function(t,s){1&t&&(e.TgZ(0,"div",0),e.TgZ(1,"div",1),e.TgZ(2,"div",2),e.TgZ(3,"app-present-datasets",3),e.NdJ("selectedDatasetIDforDelete$",function(o){return s.processDatasetIDforDelete(o)}),e.ALo(4,"async"),e.qZA(),e.qZA(),e.TgZ(5,"div",4),e.TgZ(6,"nav",5),e.TgZ(7,"li",6),e.TgZ(8,"a",7),e.NdJ("click",function(){return s.setActiveDataType("rnaseq")}),e._uU(9,"RNA-seq"),e.qZA(),e.qZA(),e.TgZ(10,"li",6),e.TgZ(11,"a",7),e.NdJ("click",function(){return s.setActiveDataType("microarray")}),e._uU(12,"Microarray"),e.qZA(),e.qZA(),e.TgZ(13,"li",6),e.TgZ(14,"a",7),e.NdJ("click",function(){return s.setActiveDataType("Proteomics")}),e._uU(15,"Proteomics"),e.qZA(),e.qZA(),e.TgZ(16,"li",6),e.TgZ(17,"a",7),e.NdJ("click",function(){return s.setActiveDataType("Other")}),e._uU(18,"Other"),e.qZA(),e.qZA(),e.TgZ(19,"li",6),e.TgZ(20,"a",7),e.NdJ("click",function(){return s.setActiveDataType("localProject")}),e._uU(21,"Local Project"),e.qZA(),e.qZA(),e.qZA(),e.YNc(22,Ua,1,0,"app-import-rna-seq",8),e.YNc(23,Ja,1,0,"app-import-microarray",8),e.YNc(24,Qa,1,0,"app-import-proteomics",8),e.YNc(25,Ea,1,0,"app-import-expression-other",8),e.YNc(26,Ga,1,0,"app-load-existed-project",8),e.qZA(),e.qZA(),e.qZA()),2&t&&(e.xp6(3),e.Q6J("datasets",e.lcZ(4,11,s.allDatasetInfoForDisplay$)),e.xp6(5),e.Q6J("ngClass",e.VKq(13,oe,"rnaseq"===s.dataType)),e.xp6(3),e.Q6J("ngClass",e.VKq(15,oe,"microarray"===s.dataType)),e.xp6(3),e.Q6J("ngClass",e.VKq(17,oe,"Proteomics"===s.dataType)),e.xp6(3),e.Q6J("ngClass",e.VKq(19,oe,"Other"===s.dataType)),e.xp6(3),e.Q6J("ngClass",e.VKq(21,oe,"localProject"===s.dataType)),e.xp6(2),e.Q6J("ngIf","rnaseq"===s.dataType),e.xp6(1),e.Q6J("ngIf","microarray"===s.dataType),e.xp6(1),e.Q6J("ngIf","Proteomics"===s.dataType),e.xp6(1),e.Q6J("ngIf","Other"===s.dataType),e.xp6(1),e.Q6J("ngIf","localProject"===s.dataType))},directives:[Dt,g.mk,g.O5,Ci,Ki,ya,Fa,Ma],pipes:[g.Ov],styles:[""]}),i})(),data:{key:"importData"}}];let La=(()=>{class i{}return i.\u0275fac=function(t){return new(t||i)},i.\u0275mod=e.oAB({type:i}),i.\u0275inj=e.cJS({imports:[[Ue.Bz.forChild(ja)],Ue.Bz]}),i})(),ka=(()=>{class i{}return i.\u0275fac=function(t){return new(t||i)},i.\u0275mod=e.oAB({type:i}),i.\u0275inj=e.cJS({providers:[W],imports:[[g.ez,D.JF,d.u5,d.UX,x.ZQ,x.lQ,x.Oz,x.HK,O._t,P.uH,q.zk.forRoot(),C.ZQ.forRoot(),M.r,L.f,H.m,it.D,La]]}),i})()},7682:(Me,z,m)=>{m.d(z,{n:()=>D});var g=m(3668);let D=(()=>{class d{constructor(x,P){this.renderer=x,this.el=P}ngAfterViewInit(){if(this.el.nativeElement.querySelector("table")){this.makeEquealHeaderBodyWidth(),this.tableObserver=new MutationObserver(()=>{this.makeEquealHeaderBodyWidth()}),this.tableObserver.observe(this.el.nativeElement.querySelector("table"),{subtree:!0,childList:!0});const x=this.el.nativeElement.querySelector("div"),P=this.el.nativeElement.querySelector("table"),O=this.el.nativeElement.querySelector("thead"),C=this.el.nativeElement.querySelector("tbody");this.renderer.listen(x,"scroll",M=>{const H=(M.srcElement.scrollLeft+P.offsetWidth).toString()+"px";this.renderer.setStyle(O,"width",H),this.renderer.setStyle(C,"width",H)})}}ngOnDestroy(){this.tableObserver&&(this.tableObserver.disconnect(),this.tableObserver=void 0)}makeEquealHeaderBodyWidth(){let x=[];x=[...this.el.nativeElement.querySelectorAll("thead tr th")].map(C=>C.offsetWidth),this.el.nativeElement.querySelectorAll("tbody tr").forEach(C=>{C.querySelectorAll("td").forEach((M,L)=>{const H=M.offsetWidth;H>=x[L]&&(x[L]=H)})});const O=x.map(C=>C.toString()+"px");this.el.nativeElement.querySelectorAll("thead tr th").forEach((C,M)=>{C.style.minWidth=O[M],C.style.maxWidth=O[M]}),this.el.nativeElement.querySelectorAll("tbody tr").forEach(C=>{C.querySelectorAll("td").forEach((M,L)=>{M.style.minWidth=O[L],M.style.maxWidth=O[L]})})}}return d.\u0275fac=function(x){return new(x||d)(g.Y36(g.Qsj),g.Y36(g.SBq))},d.\u0275dir=g.lG2({type:d,selectors:[["","appTableHeaderBodySameWidth",""]]}),d})()}}]);