"use strict";(self.webpackChunkclient=self.webpackChunkclient||[]).push([[565],{7565:(g,i,o)=>{o.r(i),o.d(i,{TutorialModule:()=>p});var e=o(3668),l=o(6019),c=o(4522),s=o(5668),r=o(1144);const u=[{path:"",component:(()=>{class t{constructor(){this.tutorialMarkdown="\n# User Manual of ExpressVis\n\nExpressVis is designed as a comprehensive platform to due with multi-omics data. RNA-Seq/MicroArray/Proteomics data are uploaded and then categorized according to the uploader\u2019s experimental design. After preprocessing like missing value imputation and normlization, standard matrix can be used for **Differentially-expressed Anlaysis** and **Clustering Analysis**  (Fig 1). After obtaining the concerned gene/protein list in the above two steps, we can proceed to **Protein-protein Interaction Analysis** and **Enrichment Analysis**. For the KEGG pathway, the significant gene set of enrichment analysis can be further visualized via KeggExp to display concerned gene/protein list. In addition, if the user uploads follow-up information, the platform supports **Dichotomization** and **Survival Analysis** of the detected protein/gene.\n\n![Flowchart](./ExpressVis_Flowchart.png)\n\n<center>Fig1. An overview of ExpressVis platform</center>\n\n# Analysis Process of Proteomics Data\n\n![ImportData_F](./ImportData_F.png)\n\n<center>Fig6.  Handle with NAs</center>\n\n### Step-4: Normalization\n\nEven small changes in experimental conditions may introduce systematic biases in the quantitative results of -omics data, which are often unknown and cannot be easily eliminated by experimental manipulation. Data normalization minimizes the influence of systematic bias on sample alignment by keeping data in each sample at the same level.\n\n![ImportData_G](./ImportData_G.png)\n\n<center>Fig7.  Normalization</center>\n\n- *Median Normalization* (default)\n\n\n\n",this.katexOptions={displayMode:!0}}ngOnInit(){}}return t.\u0275fac=function(n){return new(n||t)},t.\u0275cmp=e.Xpm({type:t,selectors:[["app-tutorial"]],decls:3,vars:2,consts:[[1,"row","flex-xl-nowrap"],[1,"offset-sm-0","col-sm-12","offset-md-1","col-md-9","offset-lg-2","col-lg-8"],["katex","",3,"src","katexOptions"]],template:function(n,f){1&n&&(e.TgZ(0,"div",0),e.TgZ(1,"div",1),e._UZ(2,"markdown",2),e.qZA(),e.qZA()),2&n&&(e.xp6(2),e.Q6J("src","/static/tutorial/UserManual.md")("katexOptions",f.katexOptions))},directives:[s.lF],styles:["img[_ngcontent-%COMP%]{width:100%;height:auto}"]}),t})()}];let m=(()=>{class t{}return t.\u0275fac=function(n){return new(n||t)},t.\u0275mod=e.oAB({type:t}),t.\u0275inj=e.cJS({imports:[[r.Bz.forChild(u)],r.Bz]}),t})(),p=(()=>{class t{}return t.\u0275fac=function(n){return new(n||t)},t.\u0275mod=e.oAB({type:t}),t.\u0275inj=e.cJS({imports:[[l.ez,s.JP.forRoot({loader:c.eN,sanitize:e.q3G.NONE}),m]]}),t})()}}]);